---
title: "tcr resampling"
author: "Kent Riemondy RBI"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(rjson)
library(tidyverse)
source("../../R/globals.R")
```

# Examine TCR resampling metrics

```{r}
barcodes_10x <- c(
  "AAGTCTGAGAAACCGC-1",
  "CTAGCCTGTCTAGCGC-1"
)

pretty_cell_names <- c(
  "AAGTCTGAGAAACCGC-1" = "Cell-1",
  "CTAGCCTGTCTAGCGC-1" = "Cell-2"
)
```

```{r}
fns <- dir(file.path(data_dir,
                     "tcr_resampling"),
           recursive = T,
           full.names = T,
           pattern = "all_contig_annotations.csv$")
names(fns) <- c("original", "resampled")
annotations <- map_dfr(fns, read_csv, .id = "library")

selected_annotations <- filter(annotations, barcode %in% barcodes_10x)

plot_dat <- filter(selected_annotations,
                   chain != "Multi",
                   cdr3 == "CASSFSTCSANYGYTF" | 
                     cdr3 == "CAVSDLEPNSSASKIIF",
                   high_confidence == "True") 

len_plt <- ggplot(plot_dat,
                  aes(barcode, length)) +
  geom_point(aes(color = library)) +
  facet_wrap(~chain) +
  scale_x_discrete(labels = c("Cell-1", "Cell-2")) +
  scale_y_continuous(labels = scales::comma) + 
  scale_color_manual(values = color_palette,
                     name = "") +
  labs(y = "Assembled Contig Length",
       x = "") + 
  theme(legend.position = "none")

read_plt <- ggplot(plot_dat,
                   aes(barcode, reads)) +
  geom_point(aes(color = library)) +
  facet_wrap(~chain) +
  scale_x_discrete(labels = c("Cell-1", "Cell-2")) +
  scale_y_continuous(labels = scales::comma) + 
  scale_color_manual(values = color_palette,
                     name = "") +
  labs(y = "# of Reads",
       x = "") + 
  theme(legend.position = "none")

umi_plt <- ggplot(plot_dat,
                   aes(barcode, umis)) +
  geom_point(aes(color = library)) +
  facet_wrap(~chain) +
  scale_x_discrete(labels = c("Cell-1", "Cell-2")) +
  scale_y_continuous(labels = scales::comma) + 
  scale_color_manual(values = color_palette,
                     name = "") +
  labs(y = "# of UMIs",
       x = "") + 
  theme(legend.position = "none")

plt <- plot_grid(len_plt, read_plt, umi_plt, nrow = 1)
save_plot("contig_length_and_reads.pdf", plt, 
          nrow = 1, ncol = 3)
plt
```

```{r make_tbl}
plt_tbl <- plot_dat %>% 
  mutate(pretty_cell_name = pretty_cell_names[barcode],
         pretty_lib_name = ifelse(library == "original",
                                  "Original Library",
                                  "Resampled Library")) %>% 
  dplyr::select(pretty_cell_name, 
                chain,
                pretty_lib_name, 
                full_length,
                length,
                reads,
                umis) %>% 
  group_by(chain) %>% 
  arrange(pretty_cell_name, .by_group = T)

clipr::write_clip(plt_tbl)
plt_tbl
```

The assembly doesn't use all of the UMIs, so the reported UMIs in the stats table will only show the umis that are used in assembly. To plot the overall umi enrichment per barcode a 10x intermediate summary file will used. 

The file is found here:
`Sample_Name/SC_VDJ_ASSEMBLER_CS/SC_VDJ_ASSEMBLER/ASSEMBLE_VDJ/fork0/files/umi_summary_tsv.tsv`

```{r read_umi_enrichment}
called_cells_info <- read_csv(file.path(data_dir,
                                        "tcr_resampling",
                                        "original",
                                        "VDJ_newcellranger_Tcell",
                                        "outs",
                                        "filtered_contig_annotations.csv"))

fn_og <- dir(file.path(data_dir,
                     "tcr_resampling",
                     "original",
                     "VDJ_newcellranger_Tcell",
                     "SC_VDJ_ASSEMBLER_CS",
                     "SC_VDJ_ASSEMBLER",
                     "ASSEMBLE_VDJ",
                     "fork0",
                     "files"),
           recursive = T,
           full.names = T,
           pattern = "umi_summary_tsv.tsv$")

fn_rs <- dir(file.path(data_dir,
                     "tcr_resampling",
                     "resampled",
                     "VDJ_newcellranger_JH109",
                     "SC_VDJ_ASSEMBLER_CS",
                     "SC_VDJ_ASSEMBLER",
                     "ASSEMBLE_VDJ",
                     "fork0",
                     "files"),
           recursive = T,
           full.names = T,
           pattern = "umi_summary_tsv.tsv$")

fns <- list(fn_og, fn_rs)
names(fns) <- c("Original",
                "Resampled")
umis <- map_dfr(fns, 
                read_tsv,
                .id = "library")

# first line for each barcode is not a umi
# keep info for called cells
umis <- filter(umis, 
               !is.na(umi), 
               barcode %in% unique(called_cells_info$barcode)) 

umi_counts <- group_by(umis, library, barcode) %>% 
  summarize(n_umis = n()) %>%  
  spread(library, n_umis, fill = 0) %>% 
  mutate(
    norm_res =  1e6 * (Resampled / sum(Resampled)),
    norm_og =  1e6 * (Original / sum(Original)),
    enrichment = log2(norm_res + 1) - log2(norm_og + 1),
    abundance = norm_og,
    resampled = barcode %in% barcodes_10x)

plt <- ggplot(umi_counts, aes(abundance, enrichment)) +
  geom_point(aes(color = resampled),
             size = 0.5) +
  geom_abline(aes(intercept = 0, slope = 0), 
              linetype = "dashed",
              colour = "darkgrey") +
  geom_text_repel(data = filter(umi_counts, resampled),
                  aes(label = ".",
                      color = resampled),
                  size = 0,
                  force = 10,
                  min.segment.length = 0,
                  point.padding = 1.0,
                  seed = 42,
                  arrow = arrow(length = unit(0.3,
                                              "line"), 
                                angle = 35,
                                type = "open", 
                                ends = "last")
  ) +
  labs(x = "Abundance in original library\n (UMIs)", 
       y = expression(paste( "Log"[2], " TCR-a/b UMIs ", 
                             frac("Resampled", "Original")))) +
  scale_x_continuous(labels = scales::comma) +      
  scale_colour_manual(name = "resampled:", values = color_palette) +
theme_cowplot(font_size = 16, line_size = 0.5) +
  theme(legend.position = "none",
        legend.text = element_text(size = 12))

save_plot("resampled_barcode_enrichment.pdf", plt)
```

## Coverage

Plot read coverage across the consensus TCR-alpha and beta chains. Fastqs from the resampled cells were realigned to the consensus sequences derived from cellranger VDJ output from the original library to compute coverage. The default output doesn't provide direct read alignments to the consensus sequences. 

For visualization purposes first I will generate annotations of the TCR a/b consensus sequences. There annotations for the `concat_ref.fasta` are buried in the log files in the cellranger output:
`SC_VDJ_ASSEMBLER_CS/SC_VDJ_ASSEMBLER/ASSEMBLE_CONSENSUS/fork0/files/concat_ref_annotations_json.json`

This json file is formatted similarlly to the `consensus_annotations.json` that is included in the outs directory.

```{r}
library(Gviz)
library(rtracklayer)
library(GenomicAlignments)
options(ucscChromosomeNames=FALSE)
  
annot_fn <- file.path(data_dir, 
                    "tcr_resampling",
                    "original",
                    "VDJ_newcellranger_Tcell",
"outs", "consensus_annotations.json")

annots <- jsonlite::read_json(annot_fn)

ids <- map(annots, ~.x[c("cdr3_start", 
                         "cdr3_stop",
                         "clonotype",
                         "contig_name",
                         "productive",
                         "start_codon_pos")] %>% 
  as_data_frame()) %>% 
  bind_rows() %>% 
  mutate(row = as.character(row_number()))

annotations <- suppressWarnings(map(annots, ~.x$annotations) %>% 
                                  map(., ~map_dfr(.x, ~.x$feature %>% 
                                                as.data.frame(), 
                                                .id = "annot_idx")) %>% 
                                  bind_rows(.id = "row"))

regions <- map(annots, ~.x$annotations) %>% 
  map(., ~map_dfr(.x, ~.x[c("contig_match_end", "contig_match_start") ] %>% 
                as.data.frame(), .id = "annot_idx")) %>% 
  bind_rows(.id = "row")


per_feature_annots <- inner_join(regions, annotations, by = c("row", "annot_idx"))
per_feature_annots <- left_join(per_feature_annots, ids, by = "row")
per_feature_annots <- dplyr::rename(per_feature_annots, 
                                    chromosome = contig_name,
                                    id = gene_name,
                                    feature = region_type,
                                    end = contig_match_end, 
                                    start = contig_match_start)

per_feature_annots <- mutate(per_feature_annots, 
                             feature = str_replace_all(feature, "-REGION", ""))

aTrack <- AnnotationTrack(range = per_feature_annots, 
                          genome = "foo", 
                          showFeatureID = T, 
                          featureAnnotation = "feature",
                          showId = T, 
                          groupAnnotation = "id",
                          just.group = "below",
                          fill = color_palette[1],
                          fontcolor.item = "white",
                          fontcolor.group = "black",
                          col.title = "black",
                          col.line = color_palette[1],
                          col = color_palette[1],
                          lwd = 0.5,
                          fontsize = 8,
                          fontsize.group=14)

per_feature_cdr <- per_feature_annots %>% 
  mutate(id = "CDR3") %>% 
  dplyr::select(
    chromosome,
    start = cdr3_start,
    end = cdr3_stop,
    id) %>% 
  unique()

cdrTrack <- AnnotationTrack(range = per_feature_cdr, 
                          name = "",
                          genome = "foo", 
                          stacking = "dense",
                          showFeatureID = T, 
                          featureAnnotation = "id",
                          showId = F,
                          fill = color_palette[1],
                          fontcolor.item = "white",
                          fontcolor.group = "black",
                          col.title = "black",
                          col = color_palette[1],
                          fontsize = 6,
                          lwd = 0.5)
```


```{r gviz}

fa_fn <- file.path(data_dir,
                   "tcr_resampling",
                   "original",
                   "VDJ_newcellranger_Tcell",
                   "outs",
                   "consensus.fasta")
fa <- import(fa_fn)
bam_fns <- dir(file.path(data_dir, 
                     "tcr_resampling"),
               pattern = "bam$",
               recursive = T,
               full.names = T
               ) %>% 
  str_subset("bowtie2")

names(bam_fns) <- str_match(bam_fns,
          str_replace(barcodes_10x, "-1", "")) %>% 
  str_c(., "_", str_match(bam_fns, "original|resampled"))

#reorder to place resampled after original for each cell
bam_fns <- bam_fns[sort(names(bam_fns))]
tcrb <- "clonotype1_consensus_1"
tcra <- "clonotype2_consensus_1"

plot_cov <- function(fasta, tcr, bam_fns, aTrack, cdrTrack){
  axisTrack <- GenomeAxisTrack()
  grange_region <- GRanges(tcr,
                           IRanges(start(fasta[[tcr]])[1], 
                                   end(fasta[[tcr]])[1]))
  
  param <- ScanBamParam(which=grange_region, 
                        flag=scanBamFlag(isDuplicate=F))
  
  bams <- map(bam_fns, ~readGAlignmentPairs(.x, 
                                            param=param))
  names(bams) <- names(bam_fns)
  read_coverage <- map(bams, coverage)
  read_coverage_gr <- map(read_coverage, 
                          function(x){
                            data.frame(
                              chromosome = tcr,
                              start = 1:length(x[[tcr]]),
                              end = 1:length(x[[tcr]]),
                              coverage = as.numeric(x[[tcr]])
                            )})
  

  
 # ylims <- map(read_coverage_gr, ~log10(max(.x$score) + 1))
  
  read_coverage_gr[1:2] %>% 
    bind_rows(., .id = "library") %>% 
    spread(library, coverage) %>%  
    GRanges() -> cell_1_dat
  
  read_coverage_gr[3:4] %>% 
    bind_rows(., .id = "library") %>% 
    spread(library, coverage) %>%  
    GRanges() -> cell_2_dat
  
  dt1 <- DataTrack(range = cell_1_dat, 
                   name = "Cell-1\n(Read Coverage)",
                   groups = c("Original", "Resampled"),
                   col = color_palette,
                   legend = T,
                   cex.axis= 1,
                   genome = "foo",
                   transformation = function(x){
                     log(x + 1)
                   })
  
  dt2 <- DataTrack(range = cell_2_dat, 
                   name = "Cell-2\n(Read Coverage)",
                   groups = c("Original", "Resampled"),
                   col = color_palette,
                   legend = F,
                   genome = "foo",
                   cex.axis= 1,
                   transformation = function(x){
                     log(x + 1)
                   })                       
  
  options(scipen=16)
  pdf(paste0(tcr, "_coverage.pdf"), width = 7, height = 4)
  plotTracks(c(dt1, dt2, list(aTrack), list(cdrTrack)),
             chromosome = tcr,
             from = start(fasta[[tcr]])[1], 
             to = end(fasta[[tcr]])[1],
             type = "a",
             stacking = "squish",
             col.title = "black",
             col.axis = "black",
             background.title ="transparent",
             lwd = 1)
  dev.off()
  
  plotTracks(c(dt1, dt2, list(aTrack), list(cdrTrack)),
             chromosome = tcr,
             from = start(fasta[[tcr]])[1], 
             to = end(fasta[[tcr]])[1],
             type = "a",
             stacking = "squish",
             col.title = "black",
             col.axis = "black",
             background.title ="transparent",
             lwd = 1.5)
  
}

plot_cov(fasta = fa, tcr = tcra, 
         bam_fns = bam_fns, 
         aTrack = aTrack, 
         cdrTrack = cdrTrack)

plot_cov(fasta = fa, tcr = tcrb, 
         bam_fns = bam_fns, 
         aTrack = aTrack, 
         cdrTrack = cdrTrack)
```

## make contig figure

```{r}
annot_fn <- file.path(data_dir, 
                      "tcr_resampling",
                      "original",
                      "VDJ_newcellranger_Tcell",
                      "outs", "consensus_annotations.json")

annots <- jsonlite::read_json(annot_fn)

ids <- map(annots, ~.x["contig_name"] %>% 
  as_data_frame()) %>% 
  bind_rows() %>% 
  mutate(row = as.character(row_number()))

seqs <- map(annots, ~.x["sequence"] %>% 
  as_data_frame()) %>% 
  bind_rows() %>% 
  mutate(row = as.character(row_number())) %>% 
  dplyr::rename(seq = sequence)

per_feature_annots <- left_join(ids, seqs, by = "row")

consensus_alpha <- unique(filter(per_feature_annots, 
                          contig_name == "clonotype2_consensus_1")$seq)
  
consensus_beta <- unique(filter(per_feature_annots, 
                          contig_name == "clonotype1_consensus_1")$seq)

## per_cell_contigs
annot_fn <- file.path(data_dir, 
                    "tcr_resampling",
                    "original",
                    "VDJ_newcellranger_Tcell",
"outs", "all_contig_annotations.json")

annots <- jsonlite::read_json(annot_fn)

get_cell_annots <- map_lgl(annots, ~.x$barcode %in% barcodes_10x)
cell_annots <- annots[get_cell_annots]

barcodes <- map_chr(cell_annots, ~.x$barcode)
cell_chain_annots <- map(cell_annots, 
                         ~.x$annotations) %>% 
  modify_depth(., 2, 
               ~.x$feature) %>% 
  modify_depth(., 2, 
               ~.x$chain) %>% 
  map_chr(., ~paste0(unique(unlist(.x)), collapse = ","))

cell_seqs <- map_chr(cell_annots, ~.x$seq)

og_annots <- data_frame(barcode = barcodes,
                        chain = cell_chain_annots,
                        seq = cell_seqs) %>% 
  mutate(contig_len = nchar(seq)) %>% 
  group_by(barcode, chain) %>% 
  arrange(desc(contig_len), .by_group = T) %>% 
  dplyr::slice(1)

## resampled
annot_fn <- file.path(data_dir, 
                      "tcr_resampling",
                      "resampled",
                      "VDJ_newcellranger_JH109",
                      "outs", "all_contig_annotations.json")

annots <- jsonlite::read_json(annot_fn)

get_cell_annots <- map_lgl(annots, ~.x$barcode %in% barcodes_10x)
cell_annots <- annots[get_cell_annots]

barcodes <- map_chr(cell_annots, ~.x$barcode)
cell_chain_annots <- map(cell_annots, 
                         ~.x$annotations) %>% 
  modify_depth(., 2, 
               ~.x$feature) %>% 
  modify_depth(., 2, 
               ~.x$chain) %>% 
  map_chr(., ~paste0(unique(unlist(.x)), collapse = ","))

cell_seqs <- map_chr(cell_annots, ~.x$seq)

rs_annots <- data_frame(barcode = barcodes,
                        chain = cell_chain_annots,
                        seq = cell_seqs) %>% 
  mutate(contig_len = nchar(seq)) %>% 
  group_by(barcode, chain) %>% 
  arrange(desc(contig_len), .by_group = T) %>% 
  dplyr::slice(1)

per_cell_annots <- bind_rows(list(
  original = og_annots,
  resampled = rs_annots), 
  .id = "library")

per_cell_annots <- filter(per_cell_annots, 
                          chain %in% c("TRA", "TRB")) %>% 
  mutate(header = str_c(library, "_", barcode, "_", chain)) %>% 
  ungroup()

kentr::write_fasta(per_cell_annots, out_path = "cell_seqs.fasta")
out_consensus <- data_frame(seq = c(consensus_alpha, 
                                    consensus_beta),
                            header = c("TCRA", "TCRB"))
kentr::write_fasta(out_consensus[1, ], out_path = "tcra.fasta")
kentr::write_fasta(out_consensus[2, ], out_path = "tcrb.fasta")
```

```{bash run_blast}
blastn -outfmt 6 -strand plus \
  -query cell_seqs.fasta \
  -subject tcra.fasta \
  > tcra_alignments.txt

blastn -outfmt 6 -strand plus \
  -query cell_seqs.fasta \
  -subject tcrb.fasta \
  > tcrb_alignments.txt
```

```{r make_bed_files}
alignments <- map_dfr(c("tcra", "tcrb"), 
        ~read_tsv(paste0(.x, 
                         "_alignments.txt"),
                  col_names = F))
alignments_to_bed <- dplyr::select(alignments, 
                                  query = X1,
                                  ref = X2,
                                  start = X9, 
                                  end = X10, 
                                  bitscore = X12)

bed <- group_by(alignments_to_bed, query, ref) %>% 
  arrange(desc(bitscore), .by_group = T) %>% 
  dplyr::slice(1) %>% 
  ungroup() %>% 
  mutate(chrom = ifelse(ref == "TCRA",
                        "clonotype2_consensus_1",
                        "clonotype1_consensus_1"),
         start = start - 1) %>% 
  dplyr::select(chrom, start, end, query)

og_bed <- filter(bed, str_detect(query, "original"))
rs_bed <- filter(bed, str_detect(query, "resampled"))

write_tsv(og_bed, "original_annotations.bed", col_names = F)
write_tsv(rs_bed, "resampled_annotations.bed", col_names = F)
```