---
title: '2018-02-18'
author: "Kent Riemondy RBI"
date: "`R Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

## Analyses to do:

1) Clean up basic processing and plots into more readable Rmds
2) quantify read-level and umi-level subsampling results
3) examine enriched genes (what are they?)
4) examine novel umi-diversity (off-by-ones?)
5) examine cDNA duplication rates


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(cache = TRUE)
```

```{r, message=F, warning=F, echo=F}
source("../../R/globals.R")
```

## Organize single cell libraries

First designate the libraries and the cells that were subsampled. 

```{r cell_ids}
libs <- list("original",
             "standard", 
             "biotinylated", 
             "phosphorothioate")


cells <- list(
  standard = c(
    "Cell_4Bone_63_70",
    "Cell_4Bone_27_61",
    "Cell_4Bone_34_48",
    "Cell_1Tumor_0_55",
    "Cell_1Tumor_7_50",
    "Cell_4Bone_30_23",
    "Cell_4Bone_61_22",
    "Cell_4Bone_66_9",
    "Cell_3Brain_47_38",
    "Cell_3Brain_14_46"
  ),
  biotin = c(
    "Cell_4Bone_63_70",
    "Cell_4Bone_27_61",
    "Cell_4Bone_34_48",
    "Cell_1Tumor_0_55",
    "Cell_1Tumor_7_50",
    "Cell_3Brain_47_38",
    "Cell_3Brain_14_46",
    "Cell_4Bone_28_54",
    "Cell_4Bone_26_20",
    "Cell_4Bone_60_43"
  ),
  phosphoro = c(
    "Cell_1Tumor_7_50",
    "Cell_3Brain_47_38",
    "Cell_4Bone_28_54",
    "Cell_4Bone_26_20",
    "Cell_4Bone_60_43" 
  )
)
```

Next read in the well data information from wafergen that contains the cell barcode information. 

```{r metadata}
bc_metadat <- file.path(data_dir, 
                        "original",
                        "fastq",
                        "original",
                        "well_data.txt")

bc_metadat <- read_tsv(bc_metadat)

bc_metadat <- select(bc_metadat,
                     Row, 
                     Col, 
                     Sample,
                     Barcode)

bc_metadat <- mutate(bc_metadat,
                     Sample = str_replace_all(Sample, "-", ""),
                     cell_id = str_c("Cell", Sample, Row, Col, sep = "_"))
bc_metadat
```

Load and organize a table for each library of read counts per cell per gene, and a table of umi counts per cell per gene. 

```{r read_counts}

## read in featurecount table with read counts per gene per cell
read_dat <- map(libs, 
                ~read_tsv(file.path(data_dir, 
                          .x,
                          "featurecounts",
                          "assigned_read_counts.txt"),
                          skip = 1))
names(read_dat) <- libs

## read in umi_tools count table with umi counts per gene per cell
umi_dat <- map(libs, 
                ~read_tsv(file.path(data_dir, 
                          .x,
                          "dgematrix",
                          "dge_matrix.txt")))
names(umi_dat) <- libs

## format read table to match umi table
read_dat <- map(read_dat, function(x){
                       x <- select(x, gene = Geneid, 7:ncol(x))
                       cols <- colnames(x)
                       gene_col <- cols[1]
                       cell_cols <- cols[2:length(cols)]
                       new_cell_cols <- str_split(cell_cols, ":", simplify = T)
                       new_cell_cols <- new_cell_cols[, 2]
                       colnames(x) <- c(gene_col, new_cell_cols)
                       x
                     })

all_cells <- c(list(original = bc_metadat$cell_id), 
               cells)

cell_obj_mdata <- map(all_cells, 
                      ~mutate(bc_metadat, 
                              subsampled = ifelse(cell_id %in% .x,
                                                  TRUE,
                                                  FALSE)))

```

Next organize these tables into simple classes called `subsampled-sets` to keep track of each experiments relavant raw, processed, and meta data. 

```{r s3ify}
#' simple class to hold info for each experiment
create_sc_obj <- function(umi_df,
                          read_df,
                          cell_mdata_df){
  x <- list()
  class(x) <- "subsampled-set"
  x$umis <- umi_df
  x$reads <- read_df
  x$meta_dat <- cell_mdata_df
  return(x)
}

sc_objs <- list(umi_dat, read_dat, cell_obj_mdata)
sc_objs <- pmap(sc_objs, create_sc_obj)
```

Next perform basic processing. 
  1) generate separate objects to store sparse matrices of umi and read counts. 
  2) normalize read and umi count data by total library size (sum of all read or umi counts for all cells in the experiment) and report as Reads per million or UMIs per million. 
  3) Compute per cell metrics (read and umi counts, sequencing saturation)

```{r store_matrices}

tidy_to_matrix <- function(df){
   df <- as.data.frame(df)
   rownames(df) <- df[, 1]
   df[, 1] <- NULL
   mat <- as.matrix(df)
   mat <- as(mat, "sparseMatrix")   
   return(mat)
}

#' keep both tidy and matrix objs
generate_matrices <- function(sc_obj){
  sc_obj$umi_matrix <- tidy_to_matrix(sc_obj$umis)
  sc_obj$read_matrix <- tidy_to_matrix(sc_obj$reads)
  sc_obj
}

#' normalize by library size (Reads per Million)
norm_libsize <- function(sc_obj){
  sc_obj$norm_umi <- 1e6 * sweep(sc_obj$umi_matrix, 2, 
                                 sum(as.vector(sc_obj$umi_matrix)), "/")
  sc_obj$norm_reads <- 1e6 * sweep(sc_obj$read_matrix, 2, 
                                   sum(as.vector(sc_obj$read_matrix)), "/")
  sc_obj
}

add_metadata <- function(sc_obj, dat){
  if (is.vector(dat)){
    new_colname <- deparse(substitute(dat))
    df <- data_frame(!!new_colname := dat)
    df[[new_colname]] <- dat
    df[["cell_id"]] <- names(dat)
    sc_obj$meta_dat <- left_join(sc_obj$meta_dat,
                                 df,
                                 by = "cell_id")
    
  } else if (is.data.frame(dat)) {
    sc_obj$meta_dat <- left_join(sc_obj$meta_dat,
                                 dat,
                                 by = "cell_id")
  }
  sc_obj
}

compute_summaries <- function(sc_obj){
  ## raw counts
  total_umis <- colSums(sc_obj$umi_matrix)
  names(total_umis) <- colnames(sc_obj$umi_matrix)
  total_reads <- colSums(sc_obj$read_matrix)
  names(total_reads) <- colnames(sc_obj$read_matrix)
  
  ## norm counts
  norm_total_umis <- colSums(sc_obj$norm_umi)
  names(norm_total_umis) <- colnames(sc_obj$norm_umi)
  norm_total_reads <- colSums(sc_obj$norm_reads)
  names(norm_total_reads) <- colnames(sc_obj$norm_reads)
    
  sc_obj <- add_metadata(sc_obj, total_umis)
  sc_obj <- add_metadata(sc_obj, total_reads)
  sc_obj <- add_metadata(sc_obj, norm_total_umis)
  sc_obj <- add_metadata(sc_obj, norm_total_reads)
  
  ## compute cDNA duplication rate 
  sc_obj$meta_dat$cDNA_duplication <- 1 - (sc_obj$meta_dat$total_umis /
                                             sc_obj$meta_dat$total_reads)
  
  sc_obj
}

sc_objs <- map(sc_objs, generate_matrices)
sc_objs <- map(sc_objs, norm_libsize)
sc_objs <- map(sc_objs, compute_summaries)
```

Compute enrichment of reads/umis over the original library. 

```{r calc_enrichment}

sc_objs <- map(sc_objs,
    function(sub_dat){
      og_counts <- select(sc_objs$original$meta_dat,
                          og_total_reads = total_reads,
                          og_total_umis = total_umis,
                          og_norm_total_umis = norm_total_umis,
                          og_norm_total_reads = norm_total_reads,
                          og_cDNA_duplication = cDNA_duplication,
                          cell_id)
      sub_dat$meta_dat <- left_join(sub_dat$meta_dat,
                         og_counts, 
                         by = "cell_id")
      
      sub_dat$meta_dat <- mutate(sub_dat$meta_dat,
                                 read_proportion = log2( total_reads / og_total_reads),
                                 umi_proportion = log2( total_umis / og_total_umis),
                                 norm_read_proportion = log2( norm_total_reads /
                                                                og_norm_total_reads),
                                 norm_umi_proportion = log2( norm_total_umis /
                                                               og_norm_total_umis))
      sub_dat
    })

```

## plot cDNA duplication rate

```{r cDNA_duplication_rate, fig.cap="Note the high level of sequencing saturation (0 = no-duplication, 1 = all duplicates) in the original library. Also note that the libraries tend to have higher saturatioin rates, after subsampling."}
sc_metadat <- map(sc_objs, ~.x$meta_dat) %>% 
  bind_rows(.id = "library") 

ggplot(sc_metadat, aes(total_reads, cDNA_duplication)) +
  geom_point(aes(color = subsampled)) +
  labs(x = expression(paste("Read count (log"[10],")")),
       y = "Sequencing saturation") + 
  scale_x_log10() +
  scale_color_viridis(discrete = T) +
  facet_wrap(~library)
```

## plot read and umi counts per library

```{r plot_read_and_umi_counts}
color_palette = c("#1F78B4", "#bdbdbd")

global_plot_theme <- theme(
        legend.position = "top",
        legend.text = element_text(size = 10),
        strip.text = element_text(size = 8))

subsampled_metadat <- filter(sc_metadat,
                             library != "original") %>% 
  mutate(library = factor(library, 
                          levels = c(
                               "standard",
                               "biotinylated",
                               "phosphorothioate")))

unnorm_plt <- ggplot(subsampled_metadat, aes(og_total_reads / 1e6, total_reads / 1e6, colour = subsampled)) + 
  geom_point(size = 0.5) + 
  geom_abline(slope = 1) +
  coord_fixed() +
  facet_wrap(~library, nrow = 1) + 
  xlab("original library\nreads count (millions)") +
  ylab("subsampled library\nreads count (millions)") +
  ylim(c(0, 8)) + 
 # ggtitle("Raw reads associated with each cell barcode") +
  scale_colour_manual(name = "subsampled:",
                      values = color_palette) +
  theme_cowplot(font_size = 10, line_size = .5) +
  global_plot_theme

norm_plt <- ggplot(subsampled_metadat, aes(og_norm_total_reads / 1e3, norm_total_reads / 1e3, colour = subsampled)) + 
  geom_point(size = 0.5) + 
  geom_abline(slope = 1) + 
  facet_wrap(~library, nrow = 1) + 
  xlab("original library\nRPM (Thousands)") +
  ylab("subsampled library\nRPM (Thousands)") +
  scale_colour_manual(name = "subsampled:",
                      values = color_palette) +
  theme_cowplot(font_size = 10, line_size = 0.5) +
  theme(aspect.ratio = 1) + 
  global_plot_theme

unnorm_umi_plt <- ggplot(subsampled_metadat, aes(og_total_umis / 1e3, total_umis / 1e3, colour = subsampled)) + 
  geom_point(size = 0.5) + 
  geom_abline(slope = 1) +
  coord_fixed() +
  facet_wrap(~library, nrow = 1) + 
  xlab("original library\nUMI count (Thousands)") +
  ylab("subsampled library\nUMI count (Thousands)") +
  scale_colour_manual(name = "subsampled:",
                      values = color_palette) +
  theme_cowplot(font_size = 10, line_size = .5) +
  global_plot_theme

norm_umi_plt <- ggplot(subsampled_metadat, aes(og_norm_total_umis / 1e3, norm_total_umis / 1e3, colour = subsampled)) + 
  geom_point(size = 0.5) + 
  geom_abline(slope = 1) + 
  facet_wrap(~library, nrow = 1) + 
  xlab("original library\nUMI normalized RPM (Thousands)") +
  ylab("subsampled library\nUMIs per Million (Thousands)") +
  scale_colour_manual(name = "subsampled:",
                      values = color_palette) +
  theme_cowplot(font_size = 10, line_size = 0.5) +
  theme(aspect.ratio = 1) + 
  global_plot_theme

plt <- plot_grid(unnorm_plt, norm_plt, unnorm_umi_plt, norm_umi_plt, 
                 labels = "AUTO",
                 align = 'hv')
plt
save_plot("reads_per_barcode_scatterplots.pdf", plt, base_width = 8 )
```

## Plot enrichment of reads/umis 
```{r plot_ratio_reads_counts, fig.width=6, fig.height=5}

read <- ggplot(subsampled_metadat, 
       aes(cell_id, norm_read_proportion, colour = subsampled)) + 
  geom_point() + 
  labs(x = "Cell", 
       y = expression(paste( " Log"[2], " normalized reads ", frac("subsampled", "original")))) +
  scale_colour_manual(name = "subsampled:", values = color_palette) +
  facet_wrap(~library, nrow = 1) + 
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(axis.text.x = element_blank(),
        legend.position = "top",
        legend.text = element_text(size = 12))

umi <- ggplot(subsampled_metadat, 
       aes(cell_id, norm_umi_proportion, colour = subsampled)) + 
  geom_point() + 
  labs(x = "Cell", 
       y = expression(paste( "Log"[2], " normalized UMIs ", frac("subsampled", "original")))) +
  scale_colour_manual(name = "subsampled:", values = color_palette) +
  facet_wrap(~library, nrow = 1) + 
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(axis.text.x = element_blank(),
        legend.position = "top",
        legend.text = element_text(size = 12))

plt <- plot_grid(read, umi,
                 labels = "AUTO",
                 align = 'hv')
plt

ggsave("reads_umi_ratio_per_barcode_normalized.pdf", width = 6.25, height = 5)
```

  
  
```{r per_cell_ratio_reads_counts}
ggplot(dat, aes(selected_cells, 
                proportion, fill = selected_cells)) + 
  geom_boxplot() + 
  facet_wrap(~library, nrow = 1) +
  xlab("Selected for Reamplification") +
  ylab(expression(paste( "Log"[2], " ", frac("subsampled", "original")))) +
  scale_fill_manual(name = "Cell Selected\nfor reamplification\n",
                    values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 0.5) +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    legend.title = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 12)
  )
ggsave("reads_ratio_per_barcode_boxplot.pdf", width = 6, height = 5)

ggplot(dat, aes(selected_cells, 
                norm_proportion, fill = selected_cells)) + 
  geom_boxplot() + 
  facet_wrap(~library, nrow = 1) +
  xlab("Selected for Reamplification") +
  ylab(expression(paste( "Log"[2], " normalized ", frac("subsampled", "original")))) +
  scale_fill_manual(name = "Cell Selected\nfor reamplification\n",
                    values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 0.5) +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    legend.title = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 12)
  )
ggsave("reads_ratio_per_barcode_boxplot_normalized.pdf", width = 6, height = 5)
```

```{r  total_lib_enrichment}

dat<- group_by(dat, library) %>%
  mutate(total_new = sum(total), 
         total_old = sum(original_total))

dat_group <- group_by(dat, library, selected_cells) %>% 
  summarize(total_new = sum(total) / unique(total_new), 
            total_old = sum(original_total) / unique(total_old)) %>% 
  gather(lib, percent_lib, -library, -selected_cells ) %>%
  mutate(lib = factor(lib, levels = c("total_old", "total_new"), 
                      labels = c("original\nlibrary", "subsampled\nlibrary")))

ggplot(dat_group, aes(lib, percent_lib, fill =  relevel(selected_cells, "Selected for\nreamplification"))) + 
  geom_bar(stat = "identity") + 
  ylab("Fraction of\n Reads Assigned") +
  scale_fill_manual(name = "Cell Selected\nfor reamplification\n",
                    values = color_palette) +
  facet_wrap(~library) + 
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_text(angle = 90),
    legend.title = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 16)
  )
ggsave("proportion_per_barcode_barplot.pdf", width = 7, height = 7)

dat_group %>% rename(Method = library) %>% 
  summarize(
    `Targeted Library Fold-Enrichment` = (nth(percent_lib, 1) / nth(percent_lib, 3))) %>% 
  knitr::kable()
```

## Get the gene count matrices
```{r extract_the_count_data, eval = F}

system("mkdir -p count_matrices/")

clean_up_gene_data <- function(lib){
    in_dir <- paste0("../../data/", lib, "/feature_counts/")
    pattern <- "sorted.tsv$"
    files <- list.files(path = in_dir,
                    pattern = pattern,
                    full.names = T)
    dat <-  lapply(files,  readr::read_tsv,
                   col_names = T, col_types = "cc----i", skip = 1, progress = F)
    mouse_dat <- lapply(dat, function(x) x[grepl("^M_", x$Chr), c(1, 3)])
    human_dat <- lapply(dat, function(x) x[grepl("^H_", x$Chr), c(1, 3)])
    mouse_dat <- Reduce(function(x, y) dplyr::inner_join(x, y, by = "Geneid" ), mouse_dat)
    human_dat <- Reduce(function(x, y) dplyr::inner_join(x, y, by = "Geneid" ), human_dat)
    mouse_dat$Geneid <- str_c("mm38_", mouse_dat$Geneid)
    human_dat$Geneid <- str_c("hg38_",human_dat$Geneid)
    dat <- bind_rows(mouse_dat, human_dat)
    dat <- as.data.frame(dat)
    rownames(dat) <- dat$Geneid
    dat[, 1] <- NULL
    colnames(dat) <- stringr::str_match(colnames(dat), "(Cell_\\w+_\\d+_\\d+)_sorted.bam")[, 2]
    dat <- dat[, !is.na(colnames(dat)) ]
    write.table(dat, paste0("count_matrices/", lib, "_genematrix.txt"), quote = F, sep = "\t")
    dat
}

gene_dat <- map(libs, ~clean_up_gene_data(.x))
og_gene_dat <- map(original, ~clean_up_gene_data(.x))[[1]]
```

## Compute summary stats

### mapped reads comparison
``` {r reload_dat}
matrices <- dir("../01-19-2017/count_matrices/", pattern = "[psb]", full.names = T)
gene_dat <- map(matrices, ~read.table(.x)) 
names(gene_dat) <- basename(matrices)
#reorder matrix to match libs variable

gene_dat <- list(gene_dat$"standard_genematrix.txt", gene_dat$"biotinylated_genematrix.txt", gene_dat$"phosphorothioate_genematrix.txt")
names(gene_dat) <- libs

og_gene_dat <- read.table("../01-19-2017/count_matrices/original_genematrix.txt")
```

```{r number_of_mapped_reads}
get_reads_count <- function(gene_dat, cells, libs){
  gene_dat <- map(gene_dat, ~colSums(.x) ) 
  gene_dat <- map(gene_dat, ~data_frame(cell = names(.x), 
                                      total = .x))
  gene_dat <- map2(gene_dat, cells, 
                          ~mutate(.x, selected_cells = cell %in% .y))
  names(gene_dat) <- libs
  gene_dat
}


bind_data_calc_read_fc <- function(new_dat, old_dat, libs){
  res <- inner_join(new_dat, old_dat, by = "cell")
  #### remove total and unmatched counts
  res <- res %>% dplyr::filter(!grepl("unmatched|total", cell))
  ## log2 proportion
  res <- group_by(res, library) %>% 
    mutate(lib_total = sum(total),
         lib_orig_total = sum(original_total),
         norm_total = 1e6 * (total / lib_total),
         norm_original_total = 1e6 * (original_total / lib_orig_total)) %>% 
    select(-lib_total, -lib_orig_total) %>% 
    ungroup()
  ## log2 proportion
  res <- mutate(res, 
              proportion = log2(total / original_total),
              norm_proportion = log2(norm_total / norm_original_total))
  
  res <- mutate(res, 
                selected_cells = factor(selected_cells, 
                                      levels = c("TRUE", "FALSE"),
                                      labels = c("Selected for\nreamplification",
                                                 "Not Selected for\nreamplification")),
                library = factor(library,
                               levels = libs))
  res
}

get_read_summary <- function(new_dat, old_dat, libs, original) {
  sub_sampled_reads <- get_reads_count(new_dat, cells, libs)
  sub_sampled_reads <-  bind_rows(sub_sampled_reads, .id = "library")
  og_sub_sampled_reads <- get_reads_count(list(old_dat), cells, list(original))
  og_sub_sampled_reads <- rename(og_sub_sampled_reads[[1]], original_total = total) %>% 
  select(-selected_cells)
  res <- bind_data_calc_read_fc(sub_sampled_reads, og_sub_sampled_reads, libs)
  res
}

human_read_dat <- map(gene_dat, ~.x[grepl("^hg38_", rownames(.x)), ])
og_human_read_dat <- og_gene_dat[grepl("^hg38_", rownames(og_gene_dat)), ]

mouse_read_dat <- map(gene_dat, ~.x[grepl("^mm38_", rownames(.x)), ])
og_mouse_read_dat <- og_gene_dat[grepl("^mm38_", rownames(og_gene_dat)), ]

sub_sampled_reads_human <- get_read_summary(human_read_dat, og_human_read_dat, libs, original)
sub_sampled_reads_mouse <- get_read_summary(mouse_read_dat, og_mouse_read_dat, libs, original)

combined_dat <- inner_join(sub_sampled_reads_human, sub_sampled_reads_mouse, 
                by = c("cell", "selected_cells", "library")) %>% 
  select(library, cell, selected_cells, total.x, total.y, original_total.x, original_total.y) %>% 
  rename(total_human = total.x,
         total_mouse = total.y,
         original_total_human = original_total.x,
         original_total_mouse = original_total.y) %>% 
  mutate(purity = total_human / (total_mouse + total_human),
         original_purity = original_total_human / (original_total_human + original_total_mouse),
         `% human reads > 90%` = purity > .90)

subsampled_cells <- combined_dat %>%  
  filter(selected_cells == "Selected for\nreamplification") %>% 
  select(cell) %>% unique() %>%  unlist()


og_dat_to_add <- combined_dat %>% 
  select(cell, selected_cells, original_total_human, original_total_mouse) %>% 
  mutate(total_human = original_total_human, 
         total_mouse = original_total_mouse, 
         selected_cells = ifelse(cell %in% subsampled_cells, "Selected for\nreamplification", "Not Selected for\nreamplification"), 
         purity = total_human / (total_mouse + total_human),
         `% human reads > 90%` = purity > .90,
         library = "original") %>% 
  select(library, cell, selected_cells, total_human, total_mouse, original_total_human, original_total_mouse, purity, `% human reads > 90%`) %>% 
  unique() #remove the duplicates 
combined_dat <- bind_rows(combined_dat, og_dat_to_add)

combined_dat <- mutate(combined_dat,
                       library = factor(library,
                                        levels = c("original", libs)))

```




Here are the number of mapped reads compared:

```{r plot_mapped_reads_counts}
ggplot(combined_dat, aes(total_human / 1e6, total_mouse / 1e6, colour = selected_cells)) +
  geom_point(size = 0.5) + 
  coord_fixed() + 
  facet_wrap(~library, nrow = 2, scales = "free") + 
  xlab("Human reads (Millions)") +
  ylab("Mouse reads (Millions)") +
  scale_colour_manual(name = "Cell Selected\nfor reamplification\n",
                      values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(legend.title = element_blank(),
        legend.position = "top",
        legend.text = element_text(size = 12))
ggsave("mapped_reads_per_barcode_scatterplot__human_mouse.pdf", width = 7, height = 7)


ggplot(sub_sampled_reads_human, aes(norm_original_total / 1e3, norm_total / 1e3, colour = selected_cells)) + 
  geom_point(size = 0.5) + 
  geom_abline(slope = 1) +
  facet_wrap(~library, nrow = 1) + 
  xlab("original library\nRPM (Thousands)") +
  ylab("subsampled library\nRPM (Thousands)") +
  scale_colour_manual(name = "Cell Selected\nfor reamplification\n",
                      values = color_palette) +
  theme_cowplot(font_size = 12, line_size = 0.5) +
  global_plot_theme
ggsave("mapped_reads_per_barcode_scatterplot_normalized_human.pdf")

ggplot(sub_sampled_reads_mouse, aes(norm_original_total / 1e3, norm_total / 1e3, colour = selected_cells)) + 
  geom_point(size = 0.5) + 
  facet_wrap(~library, nrow = 1) + 
  xlab("original library\nRPM (Thousands)") +
  ylab("subsampled library\nRPM (Thousands)") +
  scale_colour_manual(name = "Cell Selected\nfor reamplification\n",
                      values = color_palette) +
  theme_cowplot(font_size = 12, line_size = 0.5) +
  global_plot_theme
ggsave("mapped_reads_per_barcode_scatterplot_normalized_mouse.pdf")

ggplot(sub_sampled_reads_human, aes(cell, norm_proportion, colour = selected_cells)) + 
  geom_point() + 
  ylab(expression(paste( "Log"[2], " normalized human ", frac("subsampled", "original")))) +
  scale_colour_manual(name = "Cell Selected\nfor reamplification\n", values = color_palette) +
  facet_wrap(~library, nrow = 1) + 
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    axis.text.x = element_blank(),
    legend.title = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 16)
  )
ggsave("mapped_reads_ratio_per_barcode_scatterplot_normalized_human.pdf", width = 7, height = 7)

ggplot(sub_sampled_reads_mouse, aes(cell, norm_proportion, colour = selected_cells)) + 
  geom_point() + 
  ylab(expression(paste( "Log"[2], " normalized mouse ", frac("subsampled", "original")))) +
  ggtitle("Enrichment of cell barcodes in subsampled library") +  
  scale_colour_manual(name = "Cell Selected\nfor reamplification\n", values = color_palette) +
  facet_wrap(~library, nrow = 1) + 
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    axis.text.x = element_blank(),
    legend.title = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 16)
  )
ggsave("mapped_reads_ratio_per_barcode_scatterplot_normalized_mouse.pdf", width = 7, height = 7)
```


### number of genes detected comparison {#anchor}


```{r genes_detected}
count_detected_genes <- function(gene_dat, .min_reads = 0, cells, libs){
  gene_dat <- map(gene_dat, ~map(.x, ~sum(.x > .min_reads ) ) %>% unlist() )
  gene_dat <- map(gene_dat, ~data_frame(cell = names(.x), 
                                      n_genes = .x))
  gene_dat <- map2(gene_dat, cells, 
                          ~mutate(.x, selected_cells = cell %in% .y))
  names(gene_dat) <- libs
  gene_dat
}


bind_data_calc_fc <- function(new_dat, old_dat){
  res <- inner_join(new_dat, old_dat, by = "cell")
  #### remove total and unmatched counts
  res <- res %>% dplyr::filter(!grepl("unmatched|total", cell))
  ## log2 proportion
  res <- mutate(res,  
                proportion = log2(n_genes / orig_n_genes))
  
  res <- mutate(res, 
                selected_cells = factor(selected_cells, 
                                      levels = c("TRUE", "FALSE"),
                                      labels = c("Selected for\nreamplification",
                                                 "Not Selected for\nreamplification")),
                library = factor(library,
                               levels = libs))
  res
}

get_gene_summary <- function(new_dat, old_dat, libs, original) {
  sub_sampled_genes <- count_detected_genes(new_dat, .min_reads = 0, cells, libs)
  sub_sampled_genes <-  bind_rows(sub_sampled_genes, .id = "library")
  og_sub_sampled_genes <- count_detected_genes(list(old_dat), .min_reads = 0, cells, list(original))
  og_sub_sampled_genes <- rename(og_sub_sampled_genes[[1]], orig_n_genes = n_genes) %>% 
  select(-selected_cells)
  res <- bind_data_calc_fc(sub_sampled_genes, og_sub_sampled_genes)
  res
}

human_gene_dat <- map(gene_dat, ~.x[grepl("^hg38_", rownames(.x)), ])
og_human_gene_dat <- og_gene_dat[grepl("^hg38_", rownames(og_gene_dat)), ]

mouse_gene_dat <- map(gene_dat, ~.x[grepl("^mm38_", rownames(.x)), ])
og_mouse_gene_dat <- og_gene_dat[grepl("^mm38_", rownames(og_gene_dat)), ]

sub_sampled_genes_human <- get_gene_summary(human_gene_dat, og_human_gene_dat, libs, original)
sub_sampled_genes_mouse <- get_gene_summary(mouse_gene_dat, og_mouse_gene_dat, libs, original)
```


```{r plot_gene_counts}

ggplot(sub_sampled_genes_human, aes(orig_n_genes / 1e3, n_genes / 1e3, colour = selected_cells)) + 
  geom_point(size = 0.5) +
  coord_fixed() +
  xlab("human genes detected\n original lib. (K)") +
  ylab("human genes detected\n subsampled lib. (K)") +
  scale_colour_manual(values = color_palette) +
  facet_wrap(~library, nrow = 1) +
  geom_abline(slope = 1) +
  coord_equal() +
  theme_cowplot(font_size = 16, line_size = 0.5) +
  theme(
    legend.title = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 12)
  )

ggsave("genes_per_barcode_scatterplot_human.pdf", width = 6, height = 5)

ggplot(sub_sampled_genes_mouse, aes(orig_n_genes / 1e3, n_genes / 1e3, colour = selected_cells)) + 
  geom_point(size = 0.5) +
  coord_fixed() +
  xlab("mouse genes detected\n original lib. (K)") +
  ylab("mouse genes detected\n subsampled lib. (K)") +
  scale_colour_manual(values = color_palette) +
  facet_wrap(~library, nrow = 1) +
  geom_abline(slope = 1) +
  coord_equal() +
  theme_cowplot(font_size = 16, line_size = 0.5) +
  theme(
    legend.title = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 12)
  )

ggsave("genes_per_barcode_scatterplot_mouse.pdf", width = 6, height = 5)


```


```{r plot_ratio_gene_counts}
ggplot(sub_sampled_genes_human, aes(cell, proportion, colour = selected_cells)) + 
  geom_point() + 
  ylab(expression(paste("genes ratio ", frac("subsampled library", "original library"), " Log2"))) +
  ggtitle("Enrichment of human genes in subsampled library") +  
  facet_wrap(~library, nrow = 1) + 
  scale_colour_manual(values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    axis.text.x = element_blank()
  )
ggsave("genes_per_barcode_enrichment_human.pdf")

ggplot(sub_sampled_genes_mouse, aes(cell, proportion, colour = selected_cells)) + 
  geom_point() + 
  ylab(expression(paste("genes ratio ", frac("subsampled library", "original library"), " Log2"))) +
  ggtitle("Enrichment of mouse genes in subsampled library") +  
  facet_wrap(~library, nrow = 1) + 
  scale_colour_manual(values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    axis.text.x = element_blank()
  )
ggsave("genes_per_barcode_enrichment_mouse.pdf")
```


```{r n_genes_per_selected_cell}
human_resampled <- filter(sub_sampled_genes_human, 
                          selected_cells == "Selected for\nreamplification") %>% 
  gather(gene_origin, gene_count, -library, -cell, -selected_cells) %>% 
  mutate(gene_origin = ifelse(gene_origin == "n_genes", 
                              "subsampled", "original"))

mouse_resampled <- filter(sub_sampled_genes_mouse, 
                          selected_cells == "Selected for\nreamplification") %>% 
  gather(gene_origin, gene_count, -library, -cell, -selected_cells) %>% 
  mutate(gene_origin = ifelse(gene_origin == "n_genes", 
                              "subsampled", "original"))


ggplot(human_resampled, aes(cell, gene_count, fill = gene_origin)) + 
  geom_bar(stat = "identity", position = "dodge") +
  ylab("Human\nGenes Detected") +
  scale_fill_manual(values = color_palette) +
  facet_wrap(~library, nrow = 1, scales = "free_x") +
  theme_cowplot(font_size = 16, line_size = 0.5) +
  theme(
    legend.title = element_blank(),
    legend.position = "top",
    axis.title.x = element_blank(),
    axis.text.x = element_text(size = 8, angle = 90),
    legend.text = element_text(size = 12)
  )

ggsave("genes_per_cell_barplot_human.pdf", width = 6, height = 5)


ggplot(mouse_resampled, aes(cell, gene_count, fill = gene_origin)) + 
  geom_bar(stat = "identity", position = "dodge") +
  ylab("Mouse\nGenes Detected") +
  scale_fill_manual(values = color_palette) +
  facet_wrap(~library, nrow = 1, scales = "free_x") +
  theme_cowplot(font_size = 16, line_size = 0.5) +
  theme(
    legend.title = element_blank(),
    legend.position = "top",
    axis.title.x = element_blank(),
    axis.text.x = element_text(size = 8, angle = 90),
    legend.text = element_text(size = 12)
  )

ggsave("genes_per_cell_barplot_human.pdf", width = 6, height = 5)
```
# redo next section with 
```{r compare_genesets_all, cache=FALSE}
library(dplyr)
library(stringr)
library(tidyr)
library(ggplot2)
library(ComplexHeatmap)
library(purrr)
old_countdat <- read.table("../01-19-2017/count_matrices/original_genematrix.txt")
countdat <- read.table("../01-19-2017/count_matrices/phosphorothioate_genematrix.txt")

countdat$genes <- rownames(countdat)
old_countdat$genes <- rownames(old_countdat)

countdat <- countdat %>% tbl_df() %>% 
  mutate(library = "phosphorothioate") %>% 
  select(genes, library, everything()) 

old_countdat<- old_countdat %>% 
  tbl_df() %>% 
  mutate(library = "original") %>% 
  select(genes, library, everything())

compare_counts <- function(x, y, x_name = "x_only", y_name = "y_only"){
  out <- ifelse(x > 0 & y > 0, "original genes\nrecovered",
         ifelse(x > 0 & y == 0, x_name,
                ifelse(x == 0 & y > 0, y_name, 
                       "not_detected")))
  out
}

out_dat <- purrr::map2_df(countdat[, 3:ncol(countdat)], 
                          old_countdat[, 3:ncol(old_countdat)], 
                          compare_counts, 
                          x_name = "newly\ndetected genes", 
                          y_name = "original genes\nnot recovered")

out_dat <- bind_cols(countdat[, 1], out_dat)
out_dat <- gather(out_dat, cell, counts, starts_with("Cell_"), -genes)
out_dat <- out_dat %>% group_by(cell, counts) %>% tally()
out_dat <- out_dat %>% filter(counts != "not_detected") 
out_dat <- out_dat %>% group_by(cell) %>% mutate(total_genes = sum(n), 
                                                 proportion = n / total_genes)

#### annotate selected cells
reamplified_cells <- cells$phosphoro

out_dat <- mutate(out_dat, 
                  selected_cells = ifelse(cell %in% reamplified_cells,
                                                   "Selected for\nreamplification",
                                                   "Not\n Selected"))
 

out_dat$cell_type <- str_match(out_dat$cell, "Cell_(\\w+)_\\d+_\\d+")[, 2]

ggplot(out_dat, aes(cell, proportion, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds") +
  facet_wrap( ~ selected_cells, scales = "free")

ggsave("proportion_of_genes_all_libs.pdf", width = 14, height = 7)

all_bg <- filter(out_dat, selected_cells == "Not\n Selected") %>% 
  group_by(counts) %>% 
  summarize(avg_prop = mean(proportion),
            avg_gene_counts = mean(n))

all_bg <- all_bg %>% mutate(cell = "not amplified libaries",
                            proportion = avg_prop,
                            n = avg_gene_counts, 
                            total_genes = 0,
                            selected_cells = "Selected for\nreamplification",
                            cell_type = "average of all non amplified") %>%
  select(cell, counts, n, total_genes, proportion, selected_cells, cell_type)

out_dat <- bind_rows(out_dat, all_bg)

select_dat <- filter(out_dat, selected_cells == "Selected for\nreamplification")
ggplot(select_dat, aes(cell, proportion, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds") +
   ylab("Genes Detected") +
  coord_flip() + 
  theme(
    axis.title.y = element_blank(),
    legend.title = element_blank()
  )
ggsave("proportion_of_genes_all_species.pdf", width = 7, height = 7)

ggplot(select_dat, aes(cell, n, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds") +
   ylab("Genes Detected") +
  coord_flip() + 
  theme(
    axis.title.y = element_blank(),
    legend.title = element_blank()
  )
ggsave("count_of_genes_all_species.pdf", width = 7, height = 7)
```


```{r human_nonly}
countdat_human <- countdat %>% tbl_df() %>% 
  mutate(library = "phosphorothioate") %>% 
  select(genes, library, everything()) %>% 
  filter(!str_detect(genes, "^mm38"))

old_countdat_human <- old_countdat %>% 
  tbl_df() %>% 
  mutate(library = "original") %>% 
  select(genes, library, everything()) %>% 
  filter(!str_detect(genes, "^mm38"))

out_dat <- purrr::map2_df(countdat_human[, 3:ncol(countdat)], 
                          old_countdat_human[, 3:ncol(old_countdat)], 
                          compare_counts, 
                          x_name = "newly\ndetected genes", 
                          y_name = "original genes\nnot recovered")

out_dat <- bind_cols(countdat_human[, 1], out_dat)
out_dat <- gather(out_dat, cell, counts, starts_with("Cell_"), -genes)
out_dat <- out_dat %>% group_by(cell, counts) %>% tally()
out_dat <- out_dat %>% filter(counts != "not_detected") 
out_dat <- out_dat %>% group_by(cell) %>% mutate(total_genes = sum(n), 
                                                 proportion = n / total_genes)

#### annotate selected cells
reamplified_cells <- cells$phosphoro

out_dat <- mutate(out_dat, 
                  selected_cells = ifelse(cell %in% reamplified_cells,
                                                   "Selected for\nreamplification",
                                                   "Not\n Selected"))
 

out_dat$cell_type <- str_match(out_dat$cell, "Cell_(\\w+)_\\d+_\\d+")[, 2]

ggplot(out_dat, aes(cell, proportion, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds") +
  facet_wrap( ~ selected_cells, scales = "free")

ggsave("proportion_of_genes_all_libs.pdf", width = 14, height = 7)

all_bg <- filter(out_dat, selected_cells == "Not\n Selected") %>% 
  group_by(counts) %>% 
  summarize(avg_prop = mean(proportion),
            avg_gene_counts = mean(n))

all_bg <- all_bg %>% mutate(cell = "not amplified libaries",
                            proportion = avg_prop,
                            n = avg_gene_counts, 
                            total_genes = 0,
                            selected_cells = "Selected for\nreamplification",
                            cell_type = "average of all non amplified") %>%
  select(cell, counts, n, total_genes, proportion, selected_cells, cell_type)

out_dat <- bind_rows(out_dat, all_bg)

select_dat <- filter(out_dat, selected_cells == "Selected for\nreamplification")
ggplot(select_dat, aes(cell, proportion, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds") +
   ylab("Genes Detected") +
  coord_flip() + 
  theme(
    axis.title.y = element_blank(),
    legend.title = element_blank()
  )
ggsave("proportion_of_genes_human_only.pdf", width = 7, height = 7)

ggplot(select_dat, aes(cell, n, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds") +
   ylab("Genes Detected") +
  coord_flip() + 
  theme(
    axis.title.y = element_blank(),
    legend.title = element_blank()
  )
ggsave("count_of_genes_human_only.pdf", width = 7, height = 7)
```

```{r mouse_genes}
countdat_mouse<- countdat %>% tbl_df() %>% 
  mutate(library = "phosphorothioate") %>% 
  select(genes, library, everything()) %>% 
  filter(str_detect(genes, "^mm38"))

old_countdat_mouse <- old_countdat %>% 
  tbl_df() %>% 
  mutate(library = "original") %>% 
  select(genes, library, everything()) %>% 
  filter(str_detect(genes, "^mm38"))

out_dat <- purrr::map2_df(countdat_mouse[, 3:ncol(countdat)], 
                          old_countdat_mouse[, 3:ncol(old_countdat)], 
                          compare_counts, 
                          x_name = "newly\ndetected genes", 
                          y_name = "original genes\nnot recovered")

out_dat <- bind_cols(countdat_mouse[, 1], out_dat)
out_dat <- gather(out_dat, cell, counts, starts_with("Cell_"), -genes)
out_dat <- out_dat %>% group_by(cell, counts) %>% tally()
out_dat <- out_dat %>% filter(counts != "not_detected") 
out_dat <- out_dat %>% group_by(cell) %>% mutate(total_genes = sum(n), 
                                                 proportion = n / total_genes)

#### annotate selected cells
reamplified_cells <- cells$phosphoro

out_dat <- mutate(out_dat, 
                  selected_cells = ifelse(cell %in% reamplified_cells,
                                                   "Selected for\nreamplification",
                                                   "Not\n Selected"))
 

out_dat$cell_type <- str_match(out_dat$cell, "Cell_(\\w+)_\\d+_\\d+")[, 2]

ggplot(out_dat, aes(cell, proportion, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds") +
  facet_wrap( ~ selected_cells, scales = "free")

ggsave("proportion_of_genes_all_libs_mouseonly.pdf", width = 14, height = 7)

all_bg <- filter(out_dat, selected_cells == "Not\n Selected") %>% 
  group_by(counts) %>% 
  summarize(avg_prop = mean(proportion),
            avg_gene_counts = mean(n))

all_bg <- all_bg %>% mutate(cell = "not amplified libaries",
                            proportion = avg_prop,
                            n = avg_gene_counts, 
                            total_genes = 0,
                            selected_cells = "Selected for\nreamplification",
                            cell_type = "average of all non amplified") %>%
  select(cell, counts, n, total_genes, proportion, selected_cells, cell_type)

out_dat <- bind_rows(out_dat, all_bg)

select_dat <- filter(out_dat, selected_cells == "Selected for\nreamplification")
ggplot(select_dat, aes(cell, proportion, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds") +
   ylab("Genes Detected") +
  coord_flip() + 
  theme(
    axis.title.y = element_blank(),
    legend.title = element_blank()
  )
ggsave("proportion_of_genes_mouseonly.pdf", width = 7, height = 7)

ggplot(select_dat, aes(cell, n, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds") +
  ylab("Genes Detected") +
  coord_flip() + 
  theme(
    axis.title.y = element_blank(),
    legend.title = element_blank()
  )
ggsave("count_of_genes_mouseonly.pdf", width = 7, height = 7)

```

## Compare correlation coefficiencts

```{r}
#### annotate selected cells


countdat_selected <- countdat[, colnames(countdat) %in% reamplified_cells]
old_countdat_selected <- old_countdat[, colnames(old_countdat) %in% reamplified_cells]

#countdat_selected <- countdat_selected[, !colnames(countdat_selected) %in% filter_bad_libs]
#old_countdat_selected <- old_countdat_selected[, !colnames(old_countdat_selected) %in% filter_bad_libs]
library(ComplexHeatmap)

pdf("heatmap_of_cor_coefficients.pdf", width =7, height = 7)
Heatmap(cor(countdat_selected, old_countdat_selected), 
        cluster_rows = F, 
        cluster_columns = F, heatmap_legend_param = list(title = "", color_bar = "discrete"))
dev.off()

Heatmap(cor(countdat_selected, old_countdat_selected), 
        cluster_rows = F, 
        cluster_columns = F, heatmap_legend_param = list(title = "", color_bar = "discrete"))

d <- cor(countdat[, 3:ncol(countdat)], old_countdat[, 3:ncol(old_countdat)])

vec <- vector(mode = "logical", length(3:ncol(countdat)))
for (i in 1:ncol(d)){
  vec[i] <- d[i,i] == max(d[, i], na.rm = T)
}

cells_cor_Data <- data_frame(cell = colnames(countdat[, 3:ncol(countdat)]), max_cor = vec)
cells_cor_Data <- cells_cor_Data %>% filter(cell %in% reamplified_cells)

knitr::kable(cells_cor_Data, caption = "is the pearson correlation coefficient the highest when comparing the derivative to the new library?")
```
