---
title: "Assessing barcode diversity"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(cache = TRUE)

```

```{r, message=F, warning=F, echo=F}
library(readr)
library(tidyr)
library(dplyr)
library(stringr)
library(ggplot2)
library(cowplot)
library(RColorBrewer)
library(scRNA)
library(ggplot2)
library(purrr)
library(ggrepel)
```


## Analysis goals
  The current results for the sub-sampling of single-cell-libraries has demonstrated that using biotin did not improve specific recovery of targeted transcripomes. Jay has a hypothesis that explanation for the lack of specificity is potentially the 3'-5' exonuclease activity of the Phusion polymerase. If a primer anneals to a barcode site with a few mismatchs, then phusion may try to edit these sites, which would result in amplifying reads with mismatched barcodes.  
  To determine if there is evidence for this phenomenon, I will assess the enrichment of non-targeted barcodes and the position of mismatches between targeted and non-targeted barcodes.
  
## Get Barcode data
 First read in barcode counts for the original library, the reamplified library, and the reamplified library with biotinylated primers. 
  
```{r barcode_counts}

#### biotin derivative library
in_dir <- paste0("../../data/biotinylated_libs/fastq/demultiplexed/R1/")
pattern <- "\\d.txt$"
files <- list.files(path = in_dir,
                       pattern = pattern,
                        full.names = T)
dat <-  lapply(files,  readr::read_tsv,
                 col_names = T, col_types = "ci-", progress = F) 
dat <- Reduce(function(x, y) dplyr::full_join(x, y, by = "Barcode" ), dat)
bio_dat <- dat %>% purrr::by_row(function(x) {rowSums(x[-1])}, .to = "total_biotin_derived_lib", .collate = "cols" ) %>% 
         select(Barcode, total_biotin_derived_lib) 

#### non-biotin derivative library
in_dir <- paste0("../../data/derivative_libs/fastq/demultiplexed/R1/")
pattern <- "\\d.txt$"
files <- list.files(path = in_dir,
                       pattern = pattern,
                        full.names = T)
dat <-  lapply(files,  readr::read_tsv,
                 col_names = T, col_types = "ci-", progress = F) 
dat <- Reduce(function(x, y) dplyr::full_join(x, y, by = "Barcode" ), dat)
dat <- dat %>% purrr::by_row(function(x) {rowSums(x[-1])}, .to = "total_derived_lib", .collate = "cols" ) %>% 
         select(Barcode, total_derived_lib) 

#### original library
in_dir <- paste0("../../../scRNA/data/Project_Bentley_oldchem/fastq/demultiplexed/R1/")
pattern <- "\\d.txt$"
files <- list.files(path = in_dir,
                       pattern = pattern,
                        full.names = T)
old_dat <-  lapply(files,  readr::read_tsv,
                 col_names = T, col_types = "ci-", progress = F) 
old_dat <- Reduce(function(x, y) dplyr::full_join(x, y, by = "Barcode" ), old_dat)
old_dat <- old_dat %>% purrr::by_row(function(x) {rowSums(x[-1])}, .to = "total_old_lib", .collate = "cols" ) %>% 
         select(Barcode, total_old_lib) 

#### combine rows
dat <- inner_join(dat, old_dat, by = "Barcode")
dat <- inner_join(bio_dat, dat, by = "Barcode")

#### get total reads count and normalize reads
#total_counts <- filter(dat, Barcode == "total")

#dat <- mutate(dat,
#              total_derived_lib = 1e6 *(total_derived_lib / total_counts$total_derived_lib),
#              total_old_lib = 1e6 *(total_old_lib / total_counts$total_old_lib)
#              )

#### remove total and unmatched counts
dat <- dat %>% 
  dplyr::filter(!grepl("unmatched|total", Barcode))
```



```{r annotate_selected_cells}
#### annotate selected cells for non-biotin lib
derived_selected_cells <- c(
  "Cell_4Bone_63_70",
  "Cell_4Bone_27_61",
  "Cell_4Bone_34_48",
  "Cell_1Tumor_0_55",
  "Cell_1Tumor_7_50",
  "Cell_4Bone_30_23",
  "Cell_4Bone_61_22",
  "Cell_4Bone_66_9",
  "Cell_3Brain_47_38",
  "Cell_3Brain_14_46"
)

#### annotate selected cells biotin lib
biotin_selected_cells <- c(
  "Cell_4Bone_63_70",
  "Cell_4Bone_27_61",
  "Cell_4Bone_34_48",
  "Cell_1Tumor_0_55",
  "Cell_1Tumor_7_50",
  "Cell_3Brain_47_38",
  "Cell_3Brain_14_46",
  "Cell_4Bone_28_54",
  "Cell_4Bone_26_20",
  "Cell_4Bone_60_43"
)
dat <- mutate(dat, 
              derived_selected_cells = Barcode %in% derived_selected_cells,
              biotin_selected_cells = Barcode %in% biotin_selected_cells,
              shared_cells = derived_selected_cells && biotin_selected_cells)

#### calculate ratio of new to old lib
dat <- mutate(dat, 
              log2ratio_raw_reads_derived = 
                log2(total_derived_lib / total_old_lib)) %>%
  arrange(desc(total_old_lib))

dat <- mutate(dat, 
              log2ratio_raw_reads_biotin = 
                log2(total_biotin_derived_lib / total_old_lib)) %>%
  arrange(desc(total_old_lib))

dat <- mutate(dat, 
              derived_selected_cells = factor(derived_selected_cells, 
                                      levels = c("TRUE", "FALSE"),
                                      labels = c("Selected for\nreamplification",
                                                 "Not Selected for\nreamplification"
                                               )),
             biotin_selected_cells = factor(biotin_selected_cells, 
                                      levels = c("TRUE", "FALSE"),
                                      labels = c("Selected for\nreamplification",
                                                 "Not Selected for\nreamplification"
                                               )))
```   

```{r barcode_reads_plot, fig.width = 8, fig.height = 12}

dat <- mutate(dat,
              total_old_lib_norm = 1e6 * total_old_lib / sum(dat$total_old_lib),
              total_biotin_derived_lib_norm = 1e6 * total_biotin_derived_lib / sum(dat$total_biotin_derived_lib),
              total_derived_lib_norm = 1e6 * total_derived_lib / sum(dat$total_derived_lib))

ggplot(dat, aes(total_old_lib, 
                total_biotin_derived_lib, 
                colour = biotin_selected_cells)) + 
  geom_point() + 
  geom_abline(slope = 1,intercept = 0) +
  xlab("original library reads count") +
  ylab("derivative library reads count") +
  ggtitle("Reads associated with each cell barcode") +
  scale_colour_discrete(name = "Cell Selected\nfor reamplification\n") +
  theme_cowplot(font_size = 16, line_size = 1)

cells_to_label <- filter(dat, total_biotin_derived_lib_norm >=15000) 

ggplot(dat, aes(total_old_lib_norm, 
                total_biotin_derived_lib_norm, 
                colour = biotin_selected_cells)) + 
  geom_point() + 
  geom_abline(slope = 1, intercept = 0) +
  geom_text_repel(data = cells_to_label, aes(x = total_old_lib_norm,
                                             y = total_biotin_derived_lib_norm, label = barcode), size = 3) +
  xlab("Normalized original \nlibrary reads count (cpm)") +
  ylab("Normalized derivative \nlibrary reads count (cpm)") +
  ggtitle("Normalized reads associated \nwith each cell barcode") +
  scale_colour_discrete(name = "Cell Selected\nfor reamplification\n") +
  theme_cowplot(font_size = 16, line_size = 1)



```

```{r extract_barcodes, message=F}
old_bcs <- read_tsv("../../../scRNA/data/Project_Bentley_oldchem/fastq/original/Bentley701_TAAGGCGA_L068_barcode_keys.txt",
                    col_names = c("cell", "barcode"))

dat <- inner_join(old_bcs, dat, by  = c("cell" = "Barcode"))
```

```{r mismatch_summary}
char_diff <- function(char_a, char_b){
  dplyr::if_else(char_a == char_b, 0, 1)
}
get_mismatch_index <- function(str_a, str_b){
  str_a_list <- str_split(str_a, "")
  str_b_list <- str_split(str_b, "")
  diff_list <- purrr::map2(str_a_list, str_b_list, ~char_diff(.x, .y))
  diff_list
}

sum_mismatches <- function(barcode, barcodes){
  cross2(barcode, barcodes) %>% 
  map(lift(get_mismatch_index)) %>%
  flatten() %>% as.data.frame() %>% 
  rowSums(.)
}

plot_mismatches <- function(df, x, y, .xlab = "barcode position", .ylab = "mismatch count",
                            .title = "Mismatch frequencies when comparing \na single barcode to non targeted barcodes in the pool"){
    ggplot(mismatches_normal, aes_string(x, y)) +
    geom_boxplot() +
    xlab(.xlab) +
    ylab(.ylab) + 
    expand_limits(y=0) +
    ggtitle(.title)
}

dat <- dat %>% mutate(frequency_of_mismatches = map(barcode, ~sum_mismatches(.x, dat$barcode)))

mismatches_normal <- dat %>% select(cell, barcode, frequency_of_mismatches) %>%
  unnest(frequency_of_mismatches) %>% 
  group_by(barcode) %>%
  mutate(barcode_idx = as.factor(row_number()))


plot_mismatches(mismatches_normal, 
                "barcode_idx", 
                "frequency_of_mismatches")
```

```{r mismatches_induced libraries}
targeted_cells <- dat %>%  filter(biotin_selected_cells == "Selected for\nreamplification")
not_targeted_cells <- dat %>% filter(biotin_selected_cells != "Selected for\nreamplification")

not_selected_cell_barcodes_top10 <- dat %>% filter(biotin_selected_cells == "Not Selected for\nreamplification") %>%
   mutate(log2fc_rank = row_number(desc(log2ratio_raw_reads_biotin))) %>% 
  filter(log2fc_rank <= 10) %>%  
  select(barcode) %>% unlist()

not_selected_cell_barcodes_top50 <- dat %>% filter(biotin_selected_cells == "Not Selected for\nreamplification") %>%
   mutate(log2fc_rank = row_number(desc(log2ratio_raw_reads_biotin))) %>% 
  filter(log2fc_rank <= 50) %>%  
  select(barcode) %>% unlist()

not_selected_cell_barcodes_all<- dat %>% filter(biotin_selected_cells == "Not Selected for\nreamplification") %>%
  select(barcode) %>%  unlist()

get_top_barcodes <- function(df, col_name, n_top){
   mutate_call <- lazyeval::interp(~row_number(desc(x)), x = as.name(col_name))
   mutate_(df, log2fc_rank = mutate_call) %>% 
    filter(log2fc_rank <= n_top) %>%  
    select(barcode) %>% 
    unlist()
}


not_targeted_barcodes <- map(c(10, 25, 50, 500), 
                           ~get_top_barcodes(not_targeted_cells, 
                             "log2ratio_raw_reads_biotin", 
                             .x))

targeted_cells <- targeted_cells %>% mutate(
    top10 = map(barcode, ~sum_mismatches(.x, not_targeted_barcodes[[1]])),
    top25 = map(barcode, ~sum_mismatches(.x, not_targeted_barcodes[[2]])),
    top50 = map(barcode, ~sum_mismatches(.x, not_targeted_barcodes[[3]])),
    topall = map(barcode, ~sum_mismatches(.x, not_targeted_barcodes[[4]]))
  )


mismatches_biotin <- targeted_cells %>%
  select(cell, barcode, dplyr::matches("top")) %>%
  unnest() %>% 
  group_by(barcode) %>%
  mutate(barcode_idx = as.factor(row_number()))

mismatches_biotin <- tidyr::gather(mismatches_biotin, selected_barcodes, frequency, top10:topall)

ggplot(mismatches_biotin, aes(barcode_idx, frequency)) +
  geom_boxplot() +
  facet_wrap(~selected_barcodes, scales = "free") +
  ggtitle("Mismatch frequencies when comparing \na targeted single barcode to non-targeted barcodes\n binned by fold enrichment") +
  xlab("barcode position (bp)") +
  ylab("mismatch count")

```


