---
title: "cd3d pulldown"
author: "Kent Riemondy RBI"
date: '`R Sys.Date()`'
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

## Examine Lna pulldown data

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

```{r, message=F, warning=F, echo=F}
source("../../R/globals.R")
color_palette <- c("#bdbdbd", "#E41A1C")
data_dir <- file.path(data_dir, "lna_gene")
```

## Organize single cell libraries

First designate the libraries and the cells that were resampled. 

```{r cell_ids}
genes <- c("CD3D")

libs <- c(
  "jurkat_original",
  "jurkat_gene_pulldown_lna",
  "jurkat_gene_pulldown_dna")

bc_metadat <- read_tsv(file.path(data_dir, 
                         "jurkat_gene_pulldown_dna", 
                         "fastq",
                         "original", 
                         "barcodes.tsv"),
                         col_names = c("barcode_10x", "cell_id"))

## original library to compare against
reflib <- "jurkat_original"
resampled_libs <- c("jurkat_gene_pulldown_lna", "jurkat_gene_pulldown_dna")

## reference resampled lib for resampled vs control plots
resampled_lib <- "jurkat_gene_pulldown_dna"

pretty_names_resampled_libs <- c("LNA pulldown", "DNA pulldown")
names(pretty_names_resampled_libs) <- resampled_libs

pretty_names_all_libs <- c( "Original", "LNA pulldown", "DNA pulldown")
names(pretty_names_all_libs) <- libs
```

Load and organize a table for each library of read counts per cell per gene, and a table of umi counts per cell per gene. 

```{r read_counts}
## read in umigroups flat file with read counts per umi per gene per cell
## expand out to a read count matrix
umipaths <- file.path(data_dir, 
                      libs, 
                      "umis",
                      "umigroups.txt.gz")
read_dat <- map(umipaths, 
                ~umis_to_genes(.))
names(read_dat) <- libs

## read in umi_tools count table with umi counts per gene per cell
umi_dat <- map(libs, 
                ~read_tsv(file.path(data_dir, 
                          .x,
                          "dgematrix",
                          "dge_matrix.txt")) %>% 
                 select(-matches("Cell_unmatched")))
names(umi_dat) <- libs

# add in cell info, including info for the original sample
cell_obj_mdata <- map(1:3, ~bc_metadat)
names(cell_obj_mdata) <- libs
```


```{r s3ify}
#' simple class to hold info for each experiment
create_sc_obj <- function(umi_df,
                          read_df,
                          cell_mdata_df){
  x <- list()
  class(x) <- "resampled-set"
  x$umis <- umi_df
  x$reads <- read_df
  x$meta_data <- cell_mdata_df
  return(x)
}

sc_objs <- list(umi_dat, read_dat, cell_obj_mdata)
sc_objs <- pmap(sc_objs, create_sc_obj)

rm(umi_dat)
rm(read_dat)

names(sc_objs) <- libs
  
sc_objs <- map(sc_objs, generate_matrices)
sc_objs <- map(sc_objs, norm_libsize)
sc_objs <- map(sc_objs, compute_summaries)
```

```{r}
dir.create("count_matrices", showWarnings = F)
iwalk(sc_objs, 
     function(x, y) {
       out_fn <- file.path("count_matrices", paste0(y, "_umis.tsv"))
       write_tsv(x$umis, out_fn)
       R.utils::gzip(out_fn, overwrite = TRUE)
     })
```

## tSNE analysis

```{r resampled_tsne, results = 'hide'}
library(Seurat)

mat <- sc_objs[[reflib]]$umi_matrix

resampled_mat <- imap(sc_objs[resampled_libs],
                      function(x, y) {
                        resampled_mat <- x$umi_matrix[genes, , drop = F]
                        rownames(resampled_mat) <- str_c(rownames(resampled_mat), 
                                                          "_", 
                                                          "resampled", 
                                                          "_",
                                                          y)
                        resampled_mat}) %>% 
  do.call(rbind, .)

refcols <- colnames(mat)
combined_mats <- rbind(mat, resampled_mat[, refcols, drop = F]) %>% 
  as.matrix() %>% 
  as(., "sparseMatrix")   

sobj <- CreateSeuratObject(combined_mats)
sobj <- NormalizeData(sobj)
sobj <- ScaleData(sobj)
sobj <- FindVariableGenes(sobj, do.plot = F)
sobj <- RunPCA(sobj, pc.genes = sobj@var.genes, pcs.compute = 20, do.print = F)
sobj <- RunTSNE(sobj, dims.use = 1:18, seed.use = 20180516)

```


```{r resampled_plots}
new_ids <- c(genes, paste0(genes, "_resampled_", resampled_libs))
names(new_ids) <- rep("CD3D", 3)
titles <- c("Original",
            "LNA pulldown",
            "DNA pulldown")

max_cd3_expr <- max(sobj@data[new_ids, ])
plts <- pmap(list(
  new_ids,
  names(new_ids),
  titles),
  function(x, y, z){
    plot_feature(sobj, 
                 gene = x,
                 legend_name = y,
                 pt.alpha = 1,
                 pt.size = 0.1,
                 max_y = c(0, max_cd3_expr)) +
      labs(title = z)})

plt <- plot_grid(plotlist = plts, nrow = 1)
plt

save_plot("resampled_genes.pdf", plt, nrow = 1, ncol = 3,
          base_height = 4.25, base_aspect_ratio = 1.2)

VlnPlot(sobj, new_ids)
```

## Average Gene Abundance

```{r avg_gene_abs}

#' normalize by library size (Reads per Million) and log2 + 1
norm_libsize <- function(sc_obj){
  norm_umi <- 1e6 * sweep(sc_obj$umi_matrix, 2, 
                                 sum(as.vector(sc_obj$umi_matrix)), "/")
  norm_reads <- 1e6 * sweep(sc_obj$read_matrix, 2, 
                                   sum(as.vector(sc_obj$read_matrix)), "/")
  sc_obj$log_norm_umi <- log2(norm_umi + 1)
  sc_obj$log_norm_reads <- log2(norm_reads + 1)
  sc_obj
}

sc_objs <- map(sc_objs, norm_libsize)

avg_expr <- map(sc_objs, ~log2(Matrix::rowMeans(2^.x$log_norm_umi)))
avg_expr <- map(avg_expr, ~data_frame(gene = names(.x),
                                     expr = unname(.x)))
avg_expr_df <- bind_rows(avg_expr, .id = "library")

avg_expr_df <- avg_expr_df %>% 
  spread(library, expr)

avg_expr_df <- avg_expr_df %>% 
  na.omit() %>% 
  gather(library, expr, -gene, -reflib)

avg_expr_df$fc <- log2(avg_expr_df$expr / avg_expr_df[[reflib]])
avg_expr_df$resampled = ifelse(avg_expr_df$gene %in% genes,
                              T,
                              F)
avg_expr_df <-arrange(avg_expr_df, resampled)

resampled_info <- filter(avg_expr_df, resampled)

library(ggrepel)

plt <- ggplot(avg_expr_df, aes_string(reflib, "fc")) +
  geom_point(aes(color = resampled), size = 0.25) +
  scale_color_manual(values = color_palette) +
  facet_grid(~library, labeller = as_labeller(pretty_names_resampled_libs)) +
  geom_text_repel(data = resampled_info,
                  aes(label = gene), 
                  arrow = arrow(length = unit(0.02, "npc"))) +
  labs(x = expression(paste("Abundance in original library (Log"[2], " mean UMIs)")),
       y = expression(paste( " Log"[2], " ", frac("Resampled", "Original")))) +
  theme(legend.position = "none")
  
plt
save_plot("ma_plots_umis.pdf", plt, base_aspect_ratio = 2)
```

```{r reads}

avg_expr <- map(sc_objs, ~log2(Matrix::rowMeans(2^.x$log_norm_reads)))
avg_expr <- map(avg_expr, ~data_frame(gene = names(.x),
                                     expr = unname(.x)))
avg_expr_df <- bind_rows(avg_expr, .id = "library")

avg_expr_df <- avg_expr_df %>% 
  spread(library, expr)

avg_expr_df <- avg_expr_df %>% 
  na.omit() %>% 
  gather(library, expr, -gene, -reflib)

avg_expr_df$fc <- log2(avg_expr_df$expr / avg_expr_df[[reflib]])
avg_expr_df$resampled = ifelse(avg_expr_df$gene %in% genes,
                              T,
                              F)
avg_expr_df <-arrange(avg_expr_df, resampled)


resampled_info <- filter(avg_expr_df, resampled)

plt <- ggplot(avg_expr_df, aes_string(reflib, "fc")) +
  geom_point(aes(color = resampled), size = 0.25) +
  scale_color_manual(values = color_palette) +
  facet_grid(~library, labeller = as_labeller(pretty_names_resampled_libs)) +
  geom_text_repel(data = resampled_info,
                  aes(label = gene), 
                  arrow = arrow(length = unit(0.02, "npc"))) +
  labs(x = "Abundance in original library (log2 reads)",
       y = expression(paste( " Log"[2], " ", frac("Resampled", "Original")))) 
  
plt
save_plot("ma_plots_reads.pdf", plt, base_width = 8)
```


```{r vln_plot}
plt_dat <- map(sc_objs, 
    ~.x$log_norm_umi[c("CD3D", "CD3E", "CD3G"), , drop = F]) %>% 
  map(., 
      ~t(.x) %>% as.matrix(.) %>% as.data.frame() %>% 
        tibble::rownames_to_column("cell")) %>%
  bind_rows(., .id = "library")

plt_dat <- mutate(plt_dat,
                  library = pretty_names_all_libs[library],
                  library = factor(library, 
                                   levels = pretty_names_all_libs))

plt_dat <- gather(plt_dat, gene_name, expr, -cell, -library)

plt <- ggplot(plt_dat,  
       aes(library, expr)) + 
  geom_jitter(size = 0.1, width = 0.25) + 
  geom_violin(aes(fill = library), 
               alpha = 0.75) +
  facet_grid(~gene_name) + 
  scale_fill_manual(values = palette_okabeito,
                    labels = labeller(fill = pretty_names_all_libs)) +
  labs(x = "",
       y = expression(paste("Log"[2],
                            " normalized counts"))) + 
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 90,
                                   hjust = 1,
                                   vjust = 0.5))

save_plot("cd3d_boxplot.pdf", plt,
          base_aspect_ratio = 2)
```

```{r detect_rate}
plt_dat <- map(sc_objs, 
    ~data_frame(gene = rownames(.x$umi_matrix), 
                detection_rate = Matrix::rowSums(.x$umi_matrix > 0) / ncol(.x$umi_matrix))) %>%
  bind_rows(., .id = "library")

plt_dat <- mutate(plt_dat,
                  library = pretty_names_all_libs[library],
                  library = factor(library, 
                                   levels = pretty_names_all_libs))
plt_dat <- mutate(plt_dat, ecd = ecdf(detection_rate)(detection_rate))

plt <- ggplot(plt_dat,  
       aes(detection_rate, ecd)) + 
  geom_line() + 
  facet_wrap(~library) +
 geom_text_repel(data = filter(plt_dat, gene == "CD3D"),
                   aes(label = gene),
                   force = 10,
                  min.segment.length = 0,
                  point.padding = 0.25,
                  seed = 42,
                  arrow = arrow(length = unit(0.3,
                                              "line"), 
                                angle = 35,
                                type = "open", 
                                ends = "last")
  ) +
  labs(x = "Gene Detection Rate",
       y = "eCDF")

save_plot("gene_detection_rate.pdf", plt,
          base_aspect_ratio = 3)

plt_dat %>% 
  filter(gene == "CD3D")
```


```{r coverage}
library("GenomicFeatures")
library(GenomicRanges)
library(GenomicAlignments)
library("Gviz")
library(rtracklayer)
options(ucscChromosomeNames=FALSE)

lna_probe <- "AGTTCGGTGACCTGGCTTTA"
dna_probe <- "CAGGCTGATAGTTCGGTGACCTGGCTTTATCTACTGGATG"

probe_locs <- tribble(
  ~chrom, ~start, ~end, ~strand, ~probe,
  "chr11", 118342633, 118342652, "-", "LNA probe",
  "chr11", 118342622, 118342661, "-", "DNA probe"
)
```

```{r get_gtf}
url_link <- "ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_25/gencode.v25.annotation.gtf.gz"

gtf_fn <- file.path(db_dir, "gencode.v25.annotation.gtf")
if(!file.exists(gtf_fn)){
  dir.create(db_dir, showWarnings = F)
  download.file(url_link, file.path(db_dir, "gencode.v25.annotation.gtf.gz"))
  R.utils::gunzip(file.path(db_dir, "gencode.v25.annotation.gtf.gz"))
}

```

```{r}
bam_fns <- file.path(project_dir, "data", "lna_gene",
                    libs, "star", "alignments", "Aligned_bc_exons.bam")

names(bam_fns) <- libs

gtf <- import(gtf_fn)
gtf_df <- gtf %>% as_data_frame()

txdb_hg <- makeTxDbFromGRanges(gtf)
gtrack <- GenomeAxisTrack(scale=0.10, labelPos="beside")

cd3d_coords <- gtf_df %>% 
  filter(gene_name == "CD3D") %>% 
  summarize(seqnames = unique(seqnames),
            start = min(start),
            end = max(end),
            strand = unique(strand))

grange_region <- GRanges(cd3d_coords$seqnames,
                         IRanges(cd3d_coords$start, cd3d_coords$end))

param <- ScanBamParam(which=grange_region, 
                      flag=scanBamFlag(isDuplicate=F))
bams <- map(bam_fns, ~readGAlignments(.x, 
                                      param=param))
names(bams) <- names(bam_fns)
read_coverage <- map(bams, coverage)
read_coverage_gr <- map(read_coverage, ~as(.x, "GRanges"))

options(ucscChromosomeNames=FALSE)

dtracks <- pmap(list(read_coverage_gr[libs],
                     names(read_coverage_gr[libs]),
                     palette_okabeito[1:3]),
                function(x, y, z){
                DataTrack(range = x, 
                           name =  pretty_names_all_libs[y],
                           chromosome = "chr11", 
                           genome = "hg38",
                           col.line = z
                          )})

grtrack <- GeneRegionTrack(txdb_hg, 
                           genome = "hg38", 
                           chromosome="chr11", 
                           name="CD3D",
                           stacking = "dense",
                           col.title = "black")

probe_feature <- probe_locs %>% 
  dplyr::select(
    chromosome = chrom,
    start,
    end,
    strand,
    id = probe)

probeTrack <- AnnotationTrack(range = probe_feature, 
                          name = "",
                          genome = "hg38", 
                          stacking = "squish",
                          showFeatureID = F, 
                          groupAnnotation = "id",
                          just.group = "right",
                          showId = T,
                          fill = palette_okabeito[5],
                          shape = "fixedArrow",
                          fontcolor.item = "white",
                          fontcolor.group = "black",
                          col.title = "black",
                          col =  palette_okabeito[5])

options(scipen=16)
pdf("cd3d_coverage.pdf", 
    width = 6, 
    height = 4)

plotTracks(c(list(grtrack), 
             probeTrack,
             dtracks),
           chromosome = "chr11",
           from = cd3d_coords$start,
           to = cd3d_coords$end,
           type = "a",
           reverseStrand = TRUE, 
           col.title = "black",
           col.axis = "black",
           background.title ="transparent",
           lwd = 1.5, 
           innerMargin = 10,
           margin = 10)
dev.off()
```