---
title: "09-27-16_derivative_lib_analysis"
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(cache = TRUE)

```

```{r, message=F, warning=F, echo=F}
library(readr)
library(tidyr)
library(dplyr)
library(stringr)
library(ggplot2)
library(cowplot)
library(RColorBrewer)
library(scRNA)
library(ggplot2)
```

## Compare raw read counts  
  First need to read in barcode counts for the `old_chem` original library and the `old_chem` derivative library
  
```{r barcode_counts}
#### derivative library
in_dir <- paste0("../../data/derivative_libs/fastq/demultiplexed/R1/")
pattern <- "\\d.txt$"
files <- list.files(path = in_dir,
                       pattern = pattern,
                        full.names = T)
dat <-  lapply(files,  readr::read_tsv,
                 col_names = T, col_types = "ci-", progress = F) 
dat <- Reduce(function(x, y) dplyr::full_join(x, y, by = "Barcode" ), dat)
dat <- dat %>% purrr::by_row(function(x) {rowSums(x[-1])}, .to = "total_derived_lib", .collate = "cols" ) %>% 
         select(Barcode, total_derived_lib) 

#### original library
in_dir <- paste0("../../../scRNA/data/Project_Bentley_oldchem/fastq/demultiplexed/R1/")
pattern <- "\\d.txt$"
files <- list.files(path = in_dir,
                       pattern = pattern,
                        full.names = T)
old_dat <-  lapply(files,  readr::read_tsv,
                 col_names = T, col_types = "ci-", progress = F) 
old_dat <- Reduce(function(x, y) dplyr::full_join(x, y, by = "Barcode" ), old_dat)
old_dat <- old_dat %>% purrr::by_row(function(x) {rowSums(x[-1])}, .to = "total_old_lib", .collate = "cols" ) %>% 
         select(Barcode, total_old_lib) 

#### combine rows
dat <- inner_join(dat, old_dat, by = "Barcode")

#### get total reads count and normalize reads
total_counts <- filter(dat, Barcode == "total")

dat <- mutate(dat,
              total_derived_lib = 1e6 *(total_derived_lib / total_counts$total_derived_lib),
              total_old_lib = 1e6 *(total_old_lib / total_counts$total_old_lib)
              )

#### remove total and unmatched counts
dat <- dat %>% dplyr::filter(!grepl("unmatched|total", Barcode))

#### annotate selected cells
selected_cells <- c(
  "Cell_4Bone_63_70",
  "Cell_4Bone_27_61",
  "Cell_4Bone_34_48",
  "Cell_1Tumor_0_55",
  "Cell_1Tumor_7_50",
  "Cell_4Bone_30_23",
  "Cell_4Bone_61_22",
  "Cell_4Bone_66_9",
  "Cell_3Brain_47_38",
  "Cell_3Brain_14_46"
)

dat <- mutate(dat, selected_cells = Barcode %in% selected_cells)

#### calculate ratio of new to old lib
dat <- mutate(dat, log2ratio_raw_reads = log2(total_derived_lib / total_old_lib)) %>%
  arrange(desc(total_old_lib))

dat <- mutate(dat, selected_cells = factor(selected_cells, 
                                      levels = c("TRUE", "FALSE"),
                                      labels = c("Selected for\nreamplification",
                                                 "Not Selected for\nreamplification"
                                               )))
```    
  
  
```{r plot_raw_reads_counts}
color_palette = c("#1F78B4", "#bdbdbd")
ggplot(dat, aes(total_old_lib / 1e6, total_derived_lib / 1e6, colour = selected_cells)) + 
  geom_point(size = 2) + 
  xlab("original library\nreads count (millions)") +
  ylab("derivative library\nreads count (millions") +
  ggtitle("Raw reads associated with each cell barcode") +
  scale_colour_manual(name = "Cell Selected\nfor reamplification\n",
                      values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(legend.title = element_blank(),
        legend.position = "top",
        legend.text = element_text(size = 16))
ggsave("reads_per_barcode_scatterplot.pdf")

```


```{r plot_ratio_reads_counts}

ggplot(dat, aes(Barcode, log2ratio_raw_reads, colour = selected_cells)) + 
  geom_point() + 
  ylab(expression(paste("reads ratio ", frac("derivative library", "original library"), " Log2"))) +
  ggtitle("Enrichment of cell barcodes in derivative library") +  
  scale_colour_manual(name = "Cell Selected\nfor reamplification\n", values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    axis.text.x = element_blank(),
    legend.title = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 16)
  )
ggsave("reads_ratio_per_barcode_scatterplot.pdf", width = 7, height = 7)
```

  
  
```{r boxplot_ratio_reads_counts}

ggplot(dat, aes(selected_cells, 
                log2ratio_raw_reads, fill = selected_cells)) + 
  geom_boxplot() + 
  xlab("Selected for Reamplification") +
  ylab(expression(paste("reads ratio ", frac("derivative library", "original library"), " Log2"))) +
  ggtitle("Enrichment of cell barcodes in derivative library") +  
  scale_fill_manual(name = "Cell Selected\nfor reamplification\n",
                    values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    axis.title.x = element_blank(),
    legend.title = element_blank(),
    legend.position = "none",
    legend.text = element_text(size = 16)
  )
ggsave("reads_ratio_per_barcode_boxplot.pdf", width = 7, height = 7)
```

```{r  total_lib_enrichment}
total_new <- sum(dat$total_derived_lib)
total_old <- sum(dat$total_old_lib)
dat_group <- group_by(dat, selected_cells) %>% 
  summarize(total_new = sum(total_derived_lib) / total_new, 
            total_old = sum(total_old_lib) / total_old) %>% 
  gather(lib, percent_lib, -selected_cells ) %>%
  mutate(lib = factor(lib, levels = c("total_new", "total_old"), 
                      labels = c("derivative library", "original library")))

ggplot(dat_group, aes(lib, percent_lib, fill =  relevel(selected_cells, "Selected for\nreamplification"))) + 
  geom_bar(stat = "identity") + 
  xlab("Selected for Reamplification") +
  ylab("Fraction of Reads Assigned") +
  ggtitle("Proportion of selected cell barcodes in each library") +  
  scale_fill_manual(name = "Cell Selected\nfor reamplification\n",
                    values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    axis.title.x = element_blank(),
    legend.title = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 16)
  )
ggsave("proportion_per_barcode_barplot.pdf", width = 7, height = 7)
```

## Extract barcode sequences
  
  Given that it appears that some libraries that were not selected were also enriched, the next question is to ask if these barcodes have similar hamming distances to those selected.
  
```{r extract_barcodes, message=F}
old_bcs <- read_tsv("../../../scRNA/data/Project_Bentley_oldchem/fastq/original/Bentley701_TAAGGCGA_L068_barcode_keys.txt",
                    col_names = c("cell", "barcode"))

dat <- inner_join(dat, old_bcs, by  = c("Barcode" = "cell"))

#### functions for computing hamming distances

char_diff <- function(char_a, char_b){
  dplyr::if_else(char_a == char_b, 0, 1)
}

str_diff <- function(str_a, str_b){
  # strings must be the same length
  if (nchar(str_a) != nchar(str_b)){
    stop(paste("lengths of two strings not identical"))
  } else if (str_a == str_b) {
    h_dist = 0 
    return(h_dist)
  } else {
    strs <- c(str_a, str_b)
    chars <- str_split(strs, "", simplify = T)
    h_dist <- 0
    for (i in 1:ncol(chars)){
      h_dist <- h_dist + char_diff(chars[1, i], chars[2, i])
    }
    return(h_dist)
  }
}
## compute all possible hamming distances
### extract out all minimum hamming dist bcs and cell names as a list

get_min_hamming_barcodes <- function(bcs){
  bcs_matrix <- outer(bcs, bcs, Vectorize(str_diff ))
  tmp <- as_data_frame(bcs_matrix)
  colnames(tmp) <- bcs
  tmp <- tmp %>% mutate(barcode_one = bcs) %>%
    select(barcode_one, everything()) %>%
    gather("barcode", "hamming", one_of(bcs)) %>% 
    filter(hamming > 0) %>%
    group_by(barcode) %>%
    mutate(min_hamming = min(hamming)) %>%
    filter(min_hamming == hamming) %>%
    summarize(min_barcodes = paste(barcode_one, collapse = ","),
              min_hamming = first(min_hamming))
  tmp
}



old_min_bcs <- get_min_hamming_barcodes(old_bcs$barcode)

dat <- inner_join(dat, old_min_bcs, by = c("barcode")) %>%
  rename(cell = Barcode)


```


```{r find_barcodes_close_to_amplified_libs}
similar_seq_barcodes <- dat %>% 
  filter(selected_cells == "Selected for\nreamplification") %>%
    select(min_barcodes) %>% 
  separate_rows(min_barcodes) %>% unlist()

dat <- mutate(dat,
              close_to_amplified_lib = if_else(
                barcode %in% similar_seq_barcodes,
                "similar to amplified library",
                "greater than 3 hamming dist away"
              ))
ggplot(dat, aes(cell, log2ratio_raw_reads, colour = factor(close_to_amplified_lib))) + 
  geom_point() + 
  ylab(expression(paste("reads ratio ", frac("derivative library", "original library"), " Log2"))) +
  ggtitle("Enrichment of cell barcodes in derivative library") +  
  scale_colour_discrete(name = "Cell Selected\nfor reamplification\n") +
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    axis.text.x = element_blank()
  )

``` 


## Get the gene count matrices
```{r extract_the_count_data}

system("mkdir -p ../../data/derivative_libs/count_matrices/")

in_dir <- paste0("../../data/derivative_libs/star/feature_counts/")
pattern <- "\\d+.tsv$"
countdat <- read_counts(in_dir, pattern)
write.table(dat, "../../data/derivative_libs/count_matrices/star_reads_counts.txt", quote = F, sep = "\t")

old_countdat <- read.table("../../../scRNA/results/08-30-16/matrices/Project_Bentley_oldchem_reads.txt", sep = "\t") 

```

## Compute summary stats

### mapped reads comparison
``` {r number_of_mapped_reads}
reads_detected <- colSums(old_countdat)
reads_detected <- data_frame(cell = names(reads_detected), 
                             reads_old = reads_detected)

reads_detected_new <- colSums(countdat)
reads_detected_new <- data_frame(cell = names(reads_detected_new), 
                             reads_new = reads_detected_new)

reads_detected <- inner_join(reads_detected, reads_detected_new, by = "cell")


#### remove total and unmatched counts
reads_detected <- reads_detected %>% dplyr::filter(!grepl("unmatched|total", cell))

#### annotate selected cells
selected_cells <- c(
  "Cell_4Bone_63_70",
  "Cell_4Bone_27_61",
  "Cell_4Bone_34_48",
  "Cell_1Tumor_0_55",
  "Cell_1Tumor_7_50",
  "Cell_4Bone_30_23",
  "Cell_4Bone_61_22",
  "Cell_4Bone_66_9",
  "Cell_3Brain_47_38",
  "Cell_3Brain_14_46"
)

reads_detected <- mutate(reads_detected, selected_cells = cell %in% selected_cells)

#### calculate ratio of new to old lib
reads_detected <- mutate(reads_detected, log2ratio_mapped_reads = log2(reads_new / reads_old)) %>%
  arrange(desc(reads_old))
```

Here are the number of mapped reads compared:

```{r plot_mapped_reads_counts}

ggplot(reads_detected, aes(reads_old, reads_new, colour = selected_cells)) + 
  geom_point() + 
  xlab("original library reads count") +
  ylab("derivative library reads count") +
  ggtitle("Mapped reads associated with each cell barcode") +
  scale_colour_discrete(name = "Cell Selected\nfor reamplification\n") +
  theme_cowplot(font_size = 16, line_size = 1)

```


```{r plot_mapped_ratio_reads_counts}

ggplot(reads_detected, aes(cell, log2ratio_mapped_reads, colour = selected_cells)) + 
  geom_point() + 
  ylab(expression(paste("reads ratio ", frac("derivative library", "original library"), " Log2"))) +
  ggtitle("Enrichment of cell barcodes in derivative library") +  
  scale_colour_discrete(name = "Cell Selected\nfor reamplification\n") +
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    axis.text.x = element_blank()
  )

```


### number of genes detected comparison {#anchor}


```{r genes_detected}

new <- purrr::dmap(countdat, function(x) sum(x > 0 ) ) %>% unlist() 
old <- purrr::dmap(old_countdat, function(x) sum(x > 0) ) %>% unlist() 

new <- data_frame(cell = names(new), 
                             genes_new = new)

old <- data_frame(cell = names(old), 
                             genes_old = old)

genes <- inner_join(new, old, by = "cell")

#### remove total and unmatched counts
genes <- genes %>% dplyr::filter(!grepl("unmatched|total", cell))

#### annotate selected cells
selected_cells <- c(
  "Cell_4Bone_63_70",
  "Cell_4Bone_27_61",
  "Cell_4Bone_34_48",
  "Cell_1Tumor_0_55",
  "Cell_1Tumor_7_50",
  "Cell_4Bone_30_23",
  "Cell_4Bone_61_22",
  "Cell_4Bone_66_9",
  "Cell_3Brain_47_38",
  "Cell_3Brain_14_46"
)

genes <- mutate(genes, selected_cells = cell %in% selected_cells)


#### calculate ratio of new to old lib
genes <- mutate(genes, log2ratio_genes = log2(genes_new / genes_old)) %>%
  arrange(desc(genes_old))

genes <- mutate(genes, selected_cells = factor(selected_cells, 
                                      levels = c("TRUE", "FALSE"),
                                      labels = c("Selected for\nreamplification",
                                                 "Not Selected for\nreamplification"
                                               )))
```


```{r plot_gene_counts}

ggplot(genes, aes(genes_old, genes_new, colour = selected_cells)) + 
  geom_point() + 
  xlab("original library genes count") +
  ylab("derivative library genes count") +
  ggtitle("Mapped genes associated with each cell barcode") +
  scale_colour_manual(values = color_palette) +
  coord_equal() +
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    legend.title = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 16)
  )

ggsave("genes_per_barcode_scatterplot.pdf")
```


```{r plot_ratio_gene_counts}

ggplot(genes, aes(cell, log2ratio_genes, colour = selected_cells)) + 
  geom_point() + 
  ylab(expression(paste("genes ratio ", frac("derivative library", "original library"), " Log2"))) +
  ggtitle("Enrichment of cell barcodes in derivative library") +  
  scale_colour_discrete(name = "Cell Selected\nfor reamplification\n") +
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    axis.text.x = element_blank()
  )

```

## combine all datasets
``` {r combine_all_datasets}
comb_dat <- inner_join(dat, reads_detected, by = "cell") %>%
  inner_join(genes, by = "cell") %>%
  select(-selected_cells.x, -selected_cells.y)

```



## Get the gene count matrices from mouse
### old data
``` {r old_library_mouse_human}
in_dir <- "../../../scRNA/data/Project_Bentley_oldchem/star/feature_counts/mousehuman"
pattern <- "\\d.tsv$"
files <- list.files(path = in_dir,
                        pattern = pattern,
                        full.names = T)
dat <-  lapply(files,  readr::read_tsv,
                   col_names = T, col_types = "cc----i", skip = 1, progress = F) 
mouse_dat <- lapply(dat, function(x) x[grepl("^M_", x$Chr), c(1, 3)])
human_dat <- lapply(dat, function(x) x[grepl("^H_", x$Chr), c(1, 3)])
mouse_dat <- Reduce(function(x, y) dplyr::inner_join(x, y, by = "Geneid" ), mouse_dat)
human_dat <- Reduce(function(x, y) dplyr::inner_join(x, y, by = "Geneid" ), human_dat)
mouse_dat <- as.data.frame(mouse_dat)
human_dat <- as.data.frame(human_dat)

rownames(mouse_dat) <- mouse_dat$Geneid
rownames(human_dat) <- human_dat$Geneid
human_dat[, 1] <- NULL
mouse_dat[, 1] <- NULL

colnames(mouse_dat) <- stringr::str_match(colnames(mouse_dat), "(Cell_\\w+_\\d+_\\d+).*.bam")[, 2]
colnames(human_dat) <- stringr::str_match(colnames(human_dat), "(Cell_\\w+_\\d+_\\d+).*.bam")[, 2]

mouse_dat <- mouse_dat[, !is.na(colnames(mouse_dat)) ]
human_dat <- human_dat[, !is.na(colnames(human_dat)) ]

mouse_dat <- data_frame(cells = colnames(mouse_dat), mouse = colSums(mouse_dat))
human_dat <- data_frame(cells = colnames(human_dat), human = colSums(human_dat))

dat <- dplyr::inner_join(mouse_dat, human_dat, by = "cells")

old_dat <- dat %>% mutate(purity = human / (mouse + human),
                      `% human reads > 90%` = purity > .90)

old_dat$cell_type <- str_match(old_dat$cells, "Cell_(\\w+)_\\d+_\\d+")[, 2]
old_dat_h_filter <- old_dat %>% dplyr::filter(`% human reads > 90%` == TRUE) %>% select(cells) %>% unlist()

ggplot(old_dat, aes(human, mouse)) + 
  geom_point(aes(colour = `% human reads > 90%`)) + 
  ggtitle("Original Library") + 
  scale_color_brewer(palette = "Set2") + 
  xlab("Human Read Counts (log10)") +
  ylab("Mouse Read Counts (log10)") +
  theme(legend.position = "top")
ggsave("original_library_mouse_human.pdf", width = 7, height = 7)
```


### new data
``` {r new_library_mouse_human}
in_dir <- "../../data/derivative_libs/star/feature_counts/mousehuman"
pattern <- "\\d.tsv$"
files <- list.files(path = in_dir,
                        pattern = pattern,
                        full.names = T)
dat <-  lapply(files,  readr::read_tsv,
                   col_names = T, col_types = "cc----i", skip = 1, progress = F) 
mouse_dat <- lapply(dat, function(x) x[grepl("^M_", x$Chr), c(1, 3)])
human_dat <- lapply(dat, function(x) x[grepl("^H_", x$Chr), c(1, 3)])
mouse_dat <- Reduce(function(x, y) dplyr::inner_join(x, y, by = "Geneid" ), mouse_dat)
human_dat <- Reduce(function(x, y) dplyr::inner_join(x, y, by = "Geneid" ), human_dat)
mouse_dat <- as.data.frame(mouse_dat)
human_dat <- as.data.frame(human_dat)

rownames(mouse_dat) <- mouse_dat$Geneid
rownames(human_dat) <- human_dat$Geneid
human_dat[, 1] <- NULL
mouse_dat[, 1] <- NULL

colnames(mouse_dat) <- stringr::str_match(colnames(mouse_dat), "(Cell_\\w+_\\d+_\\d+).*.bam")[, 2]
colnames(human_dat) <- stringr::str_match(colnames(human_dat), "(Cell_\\w+_\\d+_\\d+).*.bam")[, 2]

mouse_dat <- mouse_dat[, !is.na(colnames(mouse_dat)) ]
human_dat <- human_dat[, !is.na(colnames(human_dat)) ]

mouse_dat <- data_frame(cells = colnames(mouse_dat), mouse = colSums(mouse_dat))
human_dat <- data_frame(cells = colnames(human_dat), human = colSums(human_dat))

dat <- dplyr::inner_join(mouse_dat, human_dat, by = "cells")

new_dat <- dat %>% mutate(purity = human / (mouse + human),
                      `% human reads > 90%` = purity > .90)

new_dat$cell_type <- str_match(new_dat$cells, "Cell_(\\w+)_\\d+_\\d+")[, 2]
new_dat_h_filter <- new_dat %>% dplyr::filter(`% human reads > 90%` == TRUE) %>% select(cells) %>% unlist()

ggplot(new_dat, aes(human, mouse)) + 
  geom_point(aes(colour = `% human reads > 90%`)) + 
  ggtitle("Derived Library") + 
  scale_color_brewer(palette = "Set2") + 
  xlab("Human Read Counts") +
  ylab("Mouse Read Counts") +
  theme(legend.position = "top")
ggsave("derived_library_mouse_human.pdf", width = 7, height = 7)
```

### pure mouse and human data

```{r get_pure_human_human_data}
in_dir <- "../../../scRNA/data/Project_Colgan/star/feature_counts/mousehuman"
pattern <- "\\d.tsv$"
files <- list.files(path = in_dir,
                        pattern = pattern,
                        full.names = T)
dat <-  lapply(files,  readr::read_tsv,
                   col_names = T, col_types = "cc----i", skip = 1, progress = F) 
mouse_dat <- lapply(dat, function(x) x[grepl("^M_", x$Chr), c(1, 3)])
human_dat <- lapply(dat, function(x) x[grepl("^H_", x$Chr), c(1, 3)])
mouse_dat <- Reduce(function(x, y) dplyr::inner_join(x, y, by = "Geneid" ), mouse_dat)
human_dat <- Reduce(function(x, y) dplyr::inner_join(x, y, by = "Geneid" ), human_dat)
mouse_dat <- as.data.frame(mouse_dat)
human_dat <- as.data.frame(human_dat)

rownames(mouse_dat) <- mouse_dat$Geneid
rownames(human_dat) <- human_dat$Geneid
human_dat[, 1] <- NULL
mouse_dat[, 1] <- NULL

colnames(mouse_dat) <- stringr::str_match(colnames(mouse_dat), "(Cell_\\w+_\\d+_\\d+).*.bam")[, 2]
colnames(human_dat) <- stringr::str_match(colnames(human_dat), "(Cell_\\w+_\\d+_\\d+).*.bam")[, 2]

mouse_dat <- mouse_dat[, !is.na(colnames(mouse_dat)) ]
human_dat <- human_dat[, !is.na(colnames(human_dat)) ]

mouse_dat <- data_frame(cells = colnames(mouse_dat), mouse = colSums(mouse_dat))
human_dat <- data_frame(cells = colnames(human_dat), human = colSums(human_dat))

dat <- dplyr::inner_join(mouse_dat, human_dat, by = "cells")

colgan_dat <- dat %>% mutate(purity = human / (mouse + human),
                      `% human reads > 90%` = purity > .90)

colgan_dat$cell_type <- str_match(colgan_dat$cells, "Cell_(\\w+)_\\d+_\\d+")[, 2]
colgan_dat_h_filter <- colgan_dat %>% dplyr::filter(`% human reads > 90%` == TRUE) %>% select(cells) %>% unlist()

ggplot(colgan_dat, aes(human, mouse)) + 
  geom_point(aes(colour = `% human reads > 90%`)) + 
  ggtitle("Pure Human Library") + 
  scale_color_brewer(palette = "Set2") + 
  xlab("Human Read Counts") +
  ylab("Mouse Read Counts") +
  theme(legend.position = "top")

ggsave("pure_human_colgan_library_mouse_human.pdf", width = 7, height = 7)

```
  The lowest purity level observed in this human scRNA-Seq dataset is `r min(colgan_dat$purity, na.rm = T)`  
 
```{r get_pure_mouse_human_data}
in_dir <- "../../../scRNA/data/Project_Torchia/star/feature_counts/mousehuman"
pattern <- "\\d.tsv$"
files <- list.files(path = in_dir,
                        pattern = pattern,
                        full.names = T)
dat <-  lapply(files,  readr::read_tsv,
                   col_names = T, col_types = "cc----i", skip = 1, progress = F) 
mouse_dat <- lapply(dat, function(x) x[grepl("^M_", x$Chr), c(1, 3)])
human_dat <- lapply(dat, function(x) x[grepl("^H_", x$Chr), c(1, 3)])
mouse_dat <- Reduce(function(x, y) dplyr::inner_join(x, y, by = "Geneid" ), mouse_dat)
human_dat <- Reduce(function(x, y) dplyr::inner_join(x, y, by = "Geneid" ), human_dat)
mouse_dat <- as.data.frame(mouse_dat)
human_dat <- as.data.frame(human_dat)

rownames(mouse_dat) <- mouse_dat$Geneid
rownames(human_dat) <- human_dat$Geneid
human_dat[, 1] <- NULL
mouse_dat[, 1] <- NULL

colnames(mouse_dat) <- stringr::str_match(colnames(mouse_dat), "(Cell_\\w+_\\d+_\\d+).*.bam")[, 2]
colnames(human_dat) <- stringr::str_match(colnames(human_dat), "(Cell_\\w+_\\d+_\\d+).*.bam")[, 2]

mouse_dat <- mouse_dat[, !is.na(colnames(mouse_dat)) ]
human_dat <- human_dat[, !is.na(colnames(human_dat)) ]

mouse_dat <- data_frame(cells = colnames(mouse_dat), mouse = colSums(mouse_dat))
human_dat <- data_frame(cells = colnames(human_dat), human = colSums(human_dat))

dat <- dplyr::inner_join(mouse_dat, human_dat, by = "cells")

torchia_dat <- dat %>% mutate(purity = mouse / (mouse + human),
                      `% mouse reads > 90%` = purity > .90)

torchia_dat$cell_type <- str_match(torchia_dat$cells, "Cell_(\\w+)_\\d+_\\d+")[, 2]
torchia_dat_h_filter <- torchia_dat %>% dplyr::filter(`% mouse reads > 90%` == TRUE) %>% select(cells) %>% unlist()

ggplot(torchia_dat, aes(human, mouse)) + 
  geom_point(aes(colour = `% mouse reads > 90%`)) + 
  ggtitle("Pure Mouse Library") + 
  scale_color_brewer(palette = "Set2") + 
  xlab("Human Read Counts") +
  ylab("Mouse Read Counts") +
  theme(legend.position = "top")

ggsave("pure_mouse_torchia_library_mouse_human.pdf", width = 7, height = 7)

```
  The lowest purity level observed in this mouse scRNA-Seq dataset is `r min(torchia_dat$purity, na.rm = T)`
  
```{r combine_them}
mh_dat <- inner_join(old_dat, new_dat, by = "cells")
colnames(mh_dat) <-  stringr::str_replace(colnames(mh_dat ), ".x$", "_old") 
colnames(mh_dat) <-  stringr::str_replace(colnames(mh_dat ), ".y$", "_new")

comb_dat <- inner_join(comb_dat, mh_dat, by = c("cell" = "cells"))

```

```{r}
a <- filter(comb_dat, selected_cells  == "Selected for\nreamplification") %>% select( cell, barcode, reads_old, reads_new, log2ratio_mapped_reads, purity_old, `% human reads > 90%_old`, purity_new, `% human reads > 90%_new`)

knitr::kable(a, caption = "Stats on cells selected for reamplification")

```

## Pick new mouse cells that are > 90% mouse   

  The `mouse` cells that I selected were not true `mouse` cells, as they didn't have `>90%` alignmnets to the mouse transcriptome. Instead they ranged from 38-63%, which doesn't make much sense. Perhaps these cell libraries are background noise transcript from repriming across molecules after library pooling. Laternatively they could be mouse/human doublets. Neither explanation makes too much sense. 
  However for the purposes of this experiment, I should instead reamplify libraries that are:  
  
  1)  confidently mouse cells (`> 90% mouse alignment`)
  2)  relatively abundant, so that the alignment frequency is an accurate measurement
  
Shown below are the top three most `mouse` aligned libraries. They are aligned to the mouse geneome with >95% and appear to be very good candidates for reamplification. Also, the fraction of mouse aligned reads is the same in the rederived library. 

```{r new_cells_to_amplify}
filter(comb_dat, selected_cells  == F) %>%
  arrange(`% human reads > 90%_old`) %>% select( cell, barcode, reads_old, reads_new, log2ratio_mapped_reads, purity_old, purity_new) %>% slice(1:3)
```


```{r plot_purities}
plot(comb_dat$purity_old, comb_dat$purity_new, col = ifelse(comb_dat$selected_cells == "Selected for\nreamplification", "red", "gray"))
```

## save intermediat results and frames
```{r save_intermediate_analysis}
write_tsv(comb_dat, "old_vs_new_dat.txt")
write.table(countdat, "derivative_count_matrix.txt", quote = F, sep = "\t", row.names = T)
write.table(old_countdat, "old_count_matrix.txt", quote = F, sep = "\t", row.names = T)

```

## Replot genes detected
  Because their are mouse cells in the datasets, I've removed them from the analysis of genes detected (as it was only done for human genes). Here's the replotted data:
  
```{r plot_gene_counts_no_mouse}
filter(genes, selected_cells == "Selected for\nreamplification" && !(cell %in% old_dat_h_filter))

b <- filter(genes, selected_cells == "Selected for\nreamplification")

filter_bad_libs <- b[!b$cell %in% old_dat_h_filter, ]$cell

genes <- genes[!genes$cell %in% filter_bad_libs, ]

ggplot(genes, aes(genes_old, genes_new, colour = selected_cells)) + 
  geom_point() + 
  xlab("original library genes count") +
  ylab("derivative library genes count") +
  ggtitle("Mapped genes associated with each cell barcode") +
  scale_colour_manual(values = color_palette) +
  coord_equal() +
  geom_abline(intercept = 0) +
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    legend.title = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 16)
  )

ggsave("genes_per_barcode_scatterplot_nomouse.pdf")
```  
## Compare gene identities 

Instead of focusing on the numbers of [genes detected](#anchor), it is more informative to determine how many genes detected are shared or distinct between the two libraries. 


```{r read_in_data}
dat <- read_tsv("old_vs_new_dat.txt")
countdat <- read.table("derivative_count_matrix.txt")
old_countdat <- read.table("old_count_matrix.txt")
```

To do this i'll need to compare the two count matrices. First i'd like to be sure that teh rowname order (genes) and colname order(cells) are identical, so that iteration is correct. 

```{r sanity_checks}

all(colnames(countdat) == colnames(old_countdat))

all(rownames(countdat) == rownames(old_countdat))
```

Next i need to design a function to compare the two data frames, returning a df with the following info:

1) Cell 
2) number of shared genes
3) number of genes in original only
4) number of genes in new only

```{r compare_genesets, cache = FALSE}
library(dplyr)
library(tidyr)
library(stringr)
library(ggplot2)
countdat$genes <- rownames(countdat)
old_countdat$genes <- rownames(old_countdat)
countdat <- countdat %>% tbl_df() %>% mutate(library = "derived") %>% select(genes, library, everything())
old_countdat <- old_countdat %>% tbl_df() %>% mutate(library = "original") %>% select(genes, library, everything())

dat <- bind_rows(countdat, old_countdat) %>% gather(cell, counts, starts_with("Cell_"), -genes, -library)
dat <- dat %>% mutate(detected = counts > 0)

compare_counts <- function(x, y, x_name = "x_only", y_name = "y_only"){
  out <- ifelse(x > 0 & y > 0, "original genes\nrecovered",
         ifelse(x > 0 & y == 0, x_name,
                ifelse(x == 0 & y > 0, y_name, "not_detected")))
  out
}

out_dat <- purrr::map2_df(countdat[, 3:ncol(countdat)], 
                          old_countdat[, 3:ncol(old_countdat)], 
                          compare_counts, x_name = "newly\ndetected genes", y_name = "original genes\nnot recovered")
out_dat <- bind_cols(countdat[, 1], out_dat)
out_dat <- gather(out_dat, cell, counts, starts_with("Cell_"), -genes)
out_dat <- out_dat %>% group_by(cell, counts) %>% tally()
out_dat <- out_dat %>% filter(counts != "not_detected") 
out_dat <- out_dat %>% group_by(cell) %>% mutate(total_genes = sum(n), 
                                                 proportion = n / total_genes)

#### annotate selected cells
selected_cells <- c(
  "Cell_4Bone_63_70",
  "Cell_4Bone_27_61",
  "Cell_4Bone_34_48",
  "Cell_1Tumor_0_55",
  "Cell_1Tumor_7_50",
  "Cell_4Bone_30_23",
  "Cell_4Bone_61_22",
  "Cell_4Bone_66_9",
  "Cell_3Brain_47_38",
  "Cell_3Brain_14_46"
)

out_dat <- mutate(out_dat, selected_cells = cell %in% selected_cells)

### exclude odd mouse libs 

out_dat <- out_dat[!out_dat$cell %in% filter_bad_libs, ]
 

out_dat$cell_type <- str_match(out_dat$cell, "Cell_(\\w+)_\\d+_\\d+")[, 2]

ggplot(out_dat, aes(cell, proportion, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds") +
  facet_wrap( ~ selected_cells, scales = "free")
ggsave("proportion_of_genes_all_libs.pdf", width = 14, height = 7)
all_bg <- filter(out_dat, selected_cells == FALSE) %>% 
  group_by(counts) %>% summarize(avg = mean(proportion))

all_bg <- all_bg %>% mutate(cell = "not amplified libaries",
                            proportion = avg,
                            n = 0, 
                            total_genes = 0,
                            selected_cells = TRUE,
                            cell_type = "average of all non amplified") %>%
  select(cell, counts, n, total_genes, proportion, selected_cells, cell_type)

out_dat <- bind_rows(out_dat, all_bg)

select_dat <- filter(out_dat, selected_cells == T)
ggplot(select_dat, aes(cell, proportion, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds") +
  coord_flip() + 
  theme(
    axis.title.y = element_blank(),
    legend.title = element_blank()
  )
ggsave("proportion_of_genes.pdf", width = 7, height = 7)
```


## Compare correlation coefficiencts

```{r}
#### annotate selected cells
selected_cells <- c(
  "Cell_4Bone_63_70",
  "Cell_4Bone_27_61",
  "Cell_4Bone_34_48",
  "Cell_1Tumor_0_55",
  "Cell_1Tumor_7_50",
  "Cell_4Bone_30_23",
  "Cell_4Bone_61_22",
  "Cell_4Bone_66_9",
  "Cell_3Brain_47_38",
  "Cell_3Brain_14_46"
)

countdat_selected <- countdat[, colnames(countdat) %in% selected_cells]
old_countdat_selected <- old_countdat[, colnames(old_countdat) %in% selected_cells]

countdat_selected <- countdat_selected[, !colnames(countdat_selected) %in% filter_bad_libs]
old_countdat_selected <- old_countdat_selected[, !colnames(old_countdat_selected) %in% filter_bad_libs]
library(ComplexHeatmap)

pdf("heatmap_of_cor_coefficients.pdf", width =7, height = 7)
Heatmap(cor(countdat_selected, old_countdat_selected), 
        cluster_rows = F, 
        cluster_columns = F, heatmap_legend_param = list(title = "", color_bar = "discrete"))
dev.off()

Heatmap(cor(countdat_selected, old_countdat_selected), 
        cluster_rows = F, 
        cluster_columns = F, heatmap_legend_param = list(title = "", color_bar = "discrete"))

d <- cor(countdat[, 3:ncol(countdat)], old_countdat[, 3:ncol(old_countdat)])

vec <- vector(mode = "logical", length(3:ncol(countdat)))
for (i in 1:ncol(d)){
  vec[i] <- d[i,i] == max(d[, i])
}

cells_cor_Data <- data_frame(cell = colnames(countdat[, 3:ncol(countdat)]), max_cor = vec)
cells_cor_Data <- cells_cor_Data %>% filter(cell %in% selected_cells)

knitr::kable(cells_cor_Data, caption = "is the pearson correlation coefficient the highest when comparing the derivative to the new library?")
```


```{r compare_umis, cache = FALSE, eval = F}
library(dplyr)
library(tidyr)
library(stringr)
library(ggplot2)

in_dir <- "../../data/derivative_libs/star"
pattern <- "dedup.bam.featureCounts$"
files <- list.files(path = in_dir,
                    pattern = pattern,
                    recursive = T,
                    full.names = T)
dat <-  lapply(files,  read.table,
                 fill = T, stringsAsFactors = F)  
names(dat) <- stringr::str_split(basename(files), "_Aligned", 4, simplify = T)[,1]

dat <- Reduce(function(x, y) dplyr::bind_rows(x, y, .id = "cell"), dat)

dat$sample <- stringr::str_split(basename(files), "_Aligned", 4, simplify = T)[,1]

countdat$genes <- rownames(countdat)
old_countdat$genes <- rownames(old_countdat)
countdat <- countdat %>% tbl_df() %>% mutate(library = "derived") %>% select(genes, library, everything())
old_countdat <- old_countdat %>% tbl_df() %>% mutate(library = "original") %>% select(genes, library, everything())

dat <- bind_rows(countdat, old_countdat) %>% gather(cell, counts, starts_with("Cell_"), -genes, -library)
dat <- dat %>% mutate(detected = counts > 0)

compare_counts <- function(x, y, x_name = "x_only", y_name = "y_only"){
  out <- ifelse(x > 0 & y > 0, "original genes\nrecovered",
         ifelse(x > 0 & y == 0, x_name,
                ifelse(x == 0 & y > 0, y_name, "not_detected")))
  out
}

out_dat <- purrr::map2_df(countdat[, 3:ncol(countdat)], 
                          old_countdat[, 3:ncol(old_countdat)], 
                          compare_counts, x_name = "newly\ndetected genes", y_name = "original genes\nnot recovered")
out_dat <- bind_cols(countdat[, 1], out_dat)
out_dat <- gather(out_dat, cell, counts, starts_with("Cell_"), -genes)
out_dat <- out_dat %>% group_by(cell, counts) %>% tally()
out_dat <- out_dat %>% filter(counts != "not_detected") 
out_dat <- out_dat %>% group_by(cell) %>% mutate(total_genes = sum(n), 
                                                 proportion = n / total_genes)

#### annotate selected cells
selected_cells <- c(
  "Cell_4Bone_63_70",
  "Cell_4Bone_27_61",
  "Cell_4Bone_34_48",
  "Cell_1Tumor_0_55",
  "Cell_1Tumor_7_50",
  "Cell_4Bone_30_23",
  "Cell_4Bone_61_22",
  "Cell_4Bone_66_9",
  "Cell_3Brain_47_38",
  "Cell_3Brain_14_46"
)

out_dat <- mutate(out_dat, selected_cells = cell %in% selected_cells)

out_dat$cell_type <- str_match(out_dat$cell, "Cell_(\\w+)_\\d+_\\d+")[, 2]

ggplot(out_dat, aes(cell, proportion, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds") +
  facet_wrap( ~ selected_cells, scales = "free")

all_bg <- filter(out_dat, selected_cells == FALSE) %>% 
  group_by(counts) %>% summarize(avg = mean(proportion))

all_bg <- all_bg %>% mutate(cell = "not amplified libaries",
                            proportion = avg,
                            n = 0, 
                            total_genes = 0,
                            selected_cells = TRUE,
                            cell_type = "average of all non amplified") %>%
  select(cell, counts, n, total_genes, proportion, selected_cells, cell_type)

out_dat <- bind_rows(out_dat, all_bg)

select_dat <- filter(out_dat, selected_cells == T)
ggplot(select_dat, aes(cell, proportion, fill = counts)) + 
  geom_bar(stat = "identity") + 
  scale_fill_brewer(palette = "Reds")

```
