---
title: "Pick genes to resample"
author: "Kent Riemondy RBI"
date: "3/12/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

Megakaryocytes | # of 3' UTR peaks
--------------- | ----------------
PF4 | 2 (canonical PAS)
TUBA4A | 1 (canonical PAS)
TGFBI | 1 (canonical PAS)
FLI1 | 2 (canonical PAS)
TUBB1 |  3+ (some noncanonical)
STX11 | 3+ (1 canonical)

B cells | # of 3' UTR peaks
--------------- | ----------------
PAX5 | 3 (2 canonical)
CD79A | 1 (canonical)
CD19 | 1 (canonical)
CD22 | 1 (canonical)

## Examine TSNEs

```{r}
library(tidyverse)
library(Seurat)
library(RColorBrewer)
dat <- readRDS("~/Projects/10x_data/runs/20180215_kirkpatrick/pbmcTSNE")

#' set up defaults for tsne plot
plot_feature <- function(seurat_obj,
                      ident = NULL,
                      gene = NULL,
                      plot_dat = NULL,
                      plot_col = NULL,
                      pt.size = 0.001,
                      pt.alpha = 0.25,
                      label_text = FALSE,
                      label.size = 6,
                      .cols = NULL,
                      legend_names = NULL,
                      cell_filter = NULL,
                      palette_type = "cloupe",
                      col_pal = "Reds",
                      max_y = NULL,
                      ...){
  
  mdata <- seurat_obj@meta.data %>% tibble::rownames_to_column("cell")
  
  tsne_dat <- seurat_obj@dr$tsne@cell.embeddings %>% 
    as.data.frame() %>% 
    tibble::rownames_to_column("cell")
  
  tsne_dat <- left_join(mdata, tsne_dat, by = "cell")
  
  if (!is.null(cell_filter)){
    tsne_dat <- dplyr::filter(tsne_dat,
                                cell %in% cell_filter)
  }
  
  if (!is.null(gene)) {
    gene_dat <- FetchData(seurat_obj, gene) %>% 
      as.data.frame() %>% 
      tibble::rownames_to_column("cell")
    tsne_dat <- left_join(tsne_dat, gene_dat, by = "cell")
  }
  
  if (!is.null(plot_dat)){
    tsne_dat <- left_join(tsne_dat, plot_dat, by = "cell")
  }
  
  if (!is.null(ident)){
    color_aes <- ident
  } else if (!is.null(plot_dat)){
    color_aes <- plot_col
  } else {
    color_aes <- gene
  }
  
  p <- ggplot(tsne_dat, 
         aes(tSNE_1, tSNE_2)) +
    geom_point(aes_string(color = color_aes),
               size = pt.size,
               alpha = pt.alpha)
  
  ## handle legend limit 
  
  if (is.null(max_y) & is.null(ident)) {
    max_y <- c(0, max(tsne_dat[[color_aes]]))
  } else if (!is.null(ident)){
    max_y <- c(NA, NA)
  }
  
  ## handle colors
  if (is.null(.cols)){
    if (palette_type == "viridis") {
        p <- p + scale_color_viridis(discrete = F,
                        direction = -1,
                        option = col_pal,
                        limits = max_y)
    } else if (palette_type == "brewer") {
        p <- p + scale_color_distiller(limits = max_y,
                                       palette = col_pal,
                                       direction = 1)
    } else if (palette_type == "cloupe") {
        cols <- brewer.pal(11, "RdGy")[c(1:5, 7)]
        
        p <- p + scale_color_gradientn(limits = max_y,
                                       colors = rev(cols))
    }
  } #else {
  #    if (is.null(legend_names)){
  #       p <- p + scale_color_manual(values = .cols)
  #    } else {
  #       p <- p + scale_color_manual(labels = legend_names, values = .cols)
  #    }
  #}
  p
}

genes <- list(
  megakaryocytes = c(
    "PF4",
    "TUBA4A",
    "TGFBI",
    "FLI1",
    "TUBB1",  
    "STX11"),
  bcells = c(
    "PAX5",
    "CD79A",
    "CD19",
    "CD22"
  )
)


map(genes$megakaryocytes, 
    ~plot_feature(dat, gene = .x, pt.alpha = 1))

map(genes$bcells, 
    ~plot_feature(dat, gene = .x, pt.alpha = 1))
```


## genes to resample

PF4
TUBB1
PAX5
CD19

```{r get_pas}
library(valr)
library(kentr)

pas <- read_tsv("~/Projects/10x_3p/dbases/polya_sites/bed/Kirkpatrick.merged_peaks.bed",
                col_names = F)

pas <- pas[, 1:6]
colnames(pas) <- c("chrom", "start", "end", "name", "score" , "strand")

genes <- c(
  "PF4",
  "TUBB1",
  "PAX5",
  "CD19"
)

pas_genes <- separate(pas, name,
                      into = c("class", 
                               "gene", 
                               "region", 
                               "chr", 
                               "pos",
                               "strand"),
                      sep = ":",
                      remove = F)

pas_sites <- c(
  TUBB1 = "1:TUBB1:E:chr20:59026590:+",
  PF4 = "1:PF4:E:chr4:73981078:-",
  CD19 = "1:CD19:E:chr16:28939334:+",
  PAX5 = "1:PAX5:E:chr9:36833277:-")

targeted_pas <- dplyr::filter(pas_genes,
                              name %in% pas_sites)

gnome_idx <- "~/Projects/shared_dbases/genomes/GRCh38.primary_assembly.genome.fa.fai"
gnome <- "~/Projects/shared_dbases/genomes/GRCh38.primary_assembly.genome.fa"
targeted_pas_upstream <- bed_slop(targeted_pas, 
              genome = read_genome(gnome_idx),
         left = 175,
         strand = T)


write_tsv(targeted_pas_upstream, "selected_pas_sites.txt")
seqs <- get_sequences(targeted_pas_upstream, gnome)
write_fasta(seqs, "selected_pas_sites.fa", header_col = "gene")
```

