---
title: "Barcode enrichment"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

```{r, message=F, warning=F, echo=F}
source("../../R/globals.R")
library(kentr)
library(ggrepel)
```

## 10x genomics control mouse/human cell pulldown
```{r get_bcs}
cells <- c("GACGTTAGTGCCTGTG","CTGATCCCATGACGGA")

libs <- c(
  "original_10x",
  "mouse_human_cell_pulldown")

reflib <- "original_10x"
resampled_lib <- "mouse_human_cell_pulldown"

lib_data_dir <- file.path(data_dir, 
                          "lna_cell",
                          "mh_mix")

bc_metadat <- read_tsv(file.path(lib_data_dir, 
                                 reflib, 
                                 "fastq",
                                 "original", 
                                 "barcodes_from_10x_run.txt"),
                       col_names = c("cell_id", "barcode_10x")) 
cell_names = c(
  "GACGTTAGTGCCTGTG" = "Mouse Cell",
  "CTGATCCCATGACGGA" = "Human Cell")

p5_primer_seqs <- "AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT"

plot_distances <- function(query_seqs,
                           libs,
                           reflib,
                           resampled_lib,
                           lib_data_dir,
                           bc_metadat,
                           cell_names,
                           .ncol = 1, 
                           probe_prefix = str_sub(p5_primer_seqs, -4, -1),
                           barcode_prefix = str_sub(p5_primer_seqs, -16, -1)
){
  bc_counts <- map(libs,
                   ~read_tsv(file.path(lib_data_dir,
                                       .x,
                                       "fastq",
                                       "logs", 
                                       "seen_barcodes.txt"),
                             col_names = c("barcode", "count"))) 
  names(bc_counts) <- libs
  bc_counts <- bind_rows(bc_counts, .id = "library")
  bc_counts <- spread(bc_counts, library, count) 
  bc_counts_norm <- mutate(bc_counts,
                           norm_og_bcs =  UQ(sym(reflib)) /
                             sum(UQ(sym(reflib)), na.rm = T),
                           norm_rs_bcs = UQ(sym(resampled_lib)) /
                             sum(UQ(sym(resampled_lib)), na.rm = T)) %>% 
    na.omit()
  
  bc_counts_norm <- mutate(bc_counts_norm, 
                           bc_enrichment = log2(norm_rs_bcs) - log2(norm_og_bcs))
  
  cellbc_norm <- semi_join(bc_counts_norm,
                           bc_metadat, 
                           by = c("barcode" = "barcode_10x")) %>% 
    dplyr::rename(cell = barcode)
  
  probe_seqs <- str_c(probe_prefix, cells)
  ref_seqs <- str_c(barcode_prefix, unique(bc_counts$barcode))
  
  sw_res <- map_dfr(probe_seqs, 
                    function(x){
                      res <- get_sw(x, 
                                    ref_seqs)
                      res$query <- x
                      res$reference <- ref_seqs
                      res <- mutate(res, 
                                    cell = str_replace(query, 
                                                       probe_prefix, 
                                                       ""),
                                    barcode = str_replace(reference,
                                                          barcode_prefix, 
                                                          ""))
                      res
                      
                    }) %>% 
    tbl_df()
  
  # for hammming distance only directly compare the 20nt hyb region
  ref_seqs <- str_c(probe_prefix, unique(bc_counts$barcode))
  h_res <- map_dfr(probe_seqs, 
                   function(x){
                     res <- data_frame(hdist = get_hamming(x, 
                                                           ref_seqs),
                                       query = x,
                                       reference = ref_seqs)
                     res <- mutate(res, 
                                   cell = str_replace(query, probe_prefix, ""),
                                   barcode = str_replace(reference,
                                                         probe_prefix, 
                                                         ""))
                     res
                   }) %>% 
    tbl_df()
  
  
  cellbc_norm_dist <- left_join(cellbc_norm,
                                sw_res, 
                                by = c("cell" = "barcode"))
  
  cellbc_norm_dist <-  left_join(cellbc_norm_dist,
                                 h_res, 
                                 by = c("cell" = "barcode", 
                                        "cell.y" = "cell"))
  
  cellbc_norm_dist <- mutate(cellbc_norm_dist,
                             resampling_probe = cell %in% cells)
  
  ## set aside df for labeling cells with enrichment > 2 fold
  
  enriched_bcs <- cellbc_norm_dist %>% 
    filter(bc_enrichment > 1)
  
  sw_plots <- map2(probe_seqs,
                   cells,
                   function(probe_seq, cell_seq){
                     tmp <- filter(cellbc_norm_dist,
                                    query.x == probe_seq,
                                   !(resampling_probe == T & 
                                       cell != cell_seq))
                     enriched_bcs <- tmp %>% 
                       filter(bc_enrichment > 1)
                     
                     ggplot(tmp,
                            aes(sw_score,
                                bc_enrichment)) +
                       geom_point(aes(color = resampling_probe)) +
                       scale_color_manual(values = color_palette) + 
                       facet_wrap(~cell.y, labeller = as_labeller(cell_names)) + 
                       geom_text_repel(data = enriched_bcs,
                                       aes(label = cell),
                                       force = 10, 
                                       arrow = arrow(length = unit(0.03,
                                                                   "npc"), 
                                                     type = "closed", 
                                                     ends = "first"),
                                       size = 2) +
                       labs(x = "Alignment Score",
                            y = expression(paste("Barcode enrichment log"[2]))) +
                       theme(legend.pos = "none")
                   })
  sw_plot <- plot_grid(plotlist = sw_plots, nrow = 1)
  
  hdist_plots <- map2(probe_seqs, cells,
                      function(probe_seq, cell_seq){
                        tmp <- filter(cellbc_norm_dist,
                                    query.x == probe_seq,
                                   !(resampling_probe == T & 
                                       cell != cell_seq))
                       enriched_bcs <- tmp %>% 
                         filter(bc_enrichment > 1)
                        
                        ggplot(tmp,
                               aes(hdist,
                                   bc_enrichment)) +
                          geom_point(aes(color = resampling_probe)) +
                          scale_color_manual(values = color_palette) + 
                          facet_wrap(~cell.y, labeller = as_labeller(cell_names)) + 
                          geom_text_repel(data = enriched_bcs,
                                          aes(label = cell),
                                          force = 10, 
                                          arrow = arrow(length = unit(0.03, "npc"), 
                                                        type = "closed", ends = "first"),
                                          size = 2) +
                          labs(x = "Hamming Distance",
                               y = expression(paste("Barcode enrichment log"[2]))) +
                          theme(legend.pos = "none")
                      })
  hdist_plot <- plot_grid(plotlist = hdist_plots, nrow = 1)
  
  plt <- plot_grid(sw_plot, hdist_plot, ncol = .ncol)
  plt
}

plt <- plot_distances(cells,
               libs,
               reflib,
               resampled_lib,
               lib_data_dir,
               bc_metadat,
               cell_names,
               .ncol = 2)

save_plot(str_c(resampled_lib, "_bc_dist.pdf"),
          plt, 
          nrow = 1, 
          ncol = 2,
          base_height = 4,
          base_aspect_ratio = 2)
plt
  
```

## PBMC Mk cell pulldown
```{r}

cells <- c("TGCGCAGCAGGTCGTC",
           "ACTTGTTAGGACCACA",
           "CCATTCGTCCCTGACT",
           "TGTCCCAGTAAACACA")

cell_names = c(
  "TGCGCAGCAGGTCGTC" = "MK-cell-1",
  "ACTTGTTAGGACCACA" = "MK-cell-2",
  "CCATTCGTCCCTGACT" = "MK-cell-3",
  "TGTCCCAGTAAACACA" = "MK-cell-4")

libs <- c(
  "kirkpatrick",
  "mkcell_pulldown")

lib_data_dir <- file.path(data_dir, 
                          "lna_cell",
                          "pbmc_expt/")

bc_metadat <- read_tsv(file.path(lib_data_dir, 
                                 "kirkpatrick", 
                                 "fastq",
                                 "original", 
                                 "barcodes_from_10x_run.txt"),
                       col_names = c("cell_id", "barcode_10x"))

## original library to compare against
reflib <- "kirkpatrick"

## reference resampled lib for resampled vs control plots
resampled_lib <- "mkcell_pulldown"

plt <- plot_distances(cells,
                      libs,
                      reflib,
                      resampled_lib,
                      lib_data_dir,
                      bc_metadat,
                      cell_names,
                      .ncol = 1)


save_plot(str_c(resampled_lib, "_bc_dist.pdf"),
          plt, 
          nrow = 2, 
          ncol = 1,
          base_height = 4,
          base_aspect_ratio = 4)
plt
```

