---
title: "PBMC gene pulldown analysis"
author: "Kent Riemondy RBI"
date: '`R Sys.Date()`'
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

## Examine Lna pulldown data

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

```{r, message=F, warning=F, echo=F}
source("../../R/globals.R")
color_palette = c("#bdbdbd", "#1F78B4")
```

## Organize single cell libraries

First designate the libraries and the cells that were subsampled. 

```{r cell_ids}
genes <- c("PF4", "PAX5", "CD19")

libs <- c(
  "kirkpatrick",
  "gene_pulldown")

bc_metadat <- read_tsv(file.path(data_dir, 
                         "kirkpatrick", 
                         "fastq",
                         "original", 
                         "barcodes_from_10x_run.txt"),
                         col_names = c("barcode_10x", "cell_id"))

## original library to compare against
reflib <- "kirkpatrick"
subsampled_libs <- "gene_pulldown"
```

Load and organize a table for each library of read counts per cell per gene, and a table of umi counts per cell per gene. 

```{r read_counts}

## read in featurecount table with read counts per gene per cell
read_dat <- map(libs, 
                ~read_tsv(file.path(data_dir, 
                          .x,
                          "featurecounts",
                          "assigned_read_counts.txt"),
                          skip = 1))
names(read_dat) <- libs

## read in umi_tools count table with umi counts per gene per cell
umi_dat <- map(libs, 
                ~read_tsv(file.path(data_dir, 
                          .x,
                          "dgematrix",
                          "dge_matrix.txt")))
names(umi_dat) <- libs

## format read table to match umi table
read_dat <- map(read_dat, function(x){
                       x <- select(x, gene = Geneid, 7:ncol(x))
                       cols <- colnames(x)
                       gene_col <- cols[1]
                       cell_cols <- cols[2:length(cols)]
                       new_cell_cols <- str_split(cell_cols, ":", simplify = T)
                       new_cell_cols <- new_cell_cols[, 2]
                       new_cell_cols <- str_split(new_cell_cols, "-", simplify = T)
                       new_cell_cols <- new_cell_cols[, 1]
                       colnames(x) <- c(gene_col, new_cell_cols)
                       x
                     })

## add in all missing gene features to dge_matrix and make colnames and rows match (for easy comparison)
umi_dat <- map2(umi_dat, read_dat, 
          function(x, y){
            genes_to_add <- y$gene[!y$gene %in% x$gene]
            ycols <- ncol(y)
            xcols <- ncol(x)
            if(xcols != ycols){
              stop("check ncols")
            }
            x <- x[, colnames(y)]
            zeros <- matrix(rep(0L, length(genes_to_add) * (xcols - 1)),
                            nrow = length(genes_to_add))
            zero_df <- as_data_frame(zeros) %>% 
              mutate(gene = genes_to_add) %>% 
              dplyr::select(gene, everything())
            colnames(zero_df) <- colnames(x)
            x <- bind_rows(x, zero_df)
            x <- x[rownames(y), ]
            if(!all(rownames(x) == rownames(y))){
              stop("check new rownames")
            }
            x
          })

```


```{r s3ify}
#' simple class to hold info for each experiment
create_sc_obj <- function(umi_df,
                          read_df,
                          cell_mdata_df){
  x <- list()
  class(x) <- "subsampled-set"
  x$umis <- umi_df
  x$reads <- read_df
  return(x)
}

sc_objs <- list(umi_dat, read_dat)
sc_objs <- pmap(sc_objs, create_sc_obj)

rm(umi_dat)
rm(read_dat)

tidy_to_matrix <- function(df){
   df <- as.data.frame(df)
   rownames(df) <- df[, 1]
   df[, 1] <- NULL
   mat <- as.matrix(df)
   mat <- as(mat, "sparseMatrix")   
   return(mat)
}

#' keep both tidy and matrix objs
generate_matrices <- function(sc_obj){
  sc_obj$umi_matrix <- tidy_to_matrix(sc_obj$umis)
  sc_obj$read_matrix <- tidy_to_matrix(sc_obj$reads)
  sc_obj
}

sc_objs <- map(sc_objs, generate_matrices)
```

## tSNE analysis

```{r subsampled_tsne, results = 'hide'}
library(Seurat)

mat <- sc_objs[[reflib]]$umi_matrix

subsampled_mat <- sc_objs[[subsampled_libs]]$umi_matrix[genes, ]
rownames(subsampled_mat) <- str_c(rownames(subsampled_mat), "::", "subsampled")

refcols <- colnames(mat)
combined_mats <- rbind(mat, subsampled_mat[, refcols]) %>% 
  as.matrix() %>% 
  as(., "sparseMatrix")   

sobj <- CreateSeuratObject(combined_mats)
sobj <- NormalizeData(sobj)
sobj <- ScaleData(sobj, vars.to.regress = "nUMI")
sobj <- FindVariableGenes(sobj, do.plot = F)
sobj <- RunPCA(sobj, pc.genes = sobj@var.genes, pcs.compute = 20, do.print = F)
sobj <- RunTSNE(sobj, dims.use = 1:18, seed.use = 20180516)
sobj <- FindClusters(sobj, 
                     dims.use = 1:18, 
                     resolution = 0.6, 
                     print.output = F, 
                     random.seed = 20180516)
```


```{r subsampled_plots}

plts <- FeaturePlot(sobj, 
                   c(genes, paste0(genes, "::subsampled")),
                   no.legend = F,
                   do.return = T,
                   cols.use = c("grey", "purple")) %>% 
  map(., ~.x + theme(legend.position = "bottom"))

plt <- plot_grid(plotlist = plts, nrow = 2)

save_plot("resampled_genes.pdf", plt, nrow = 2, ncol = 3,
          base_height = 4.25, base_width = 4.25)

plts <- map(c(genes, paste0(genes, "\nsubsampled")), 
          ~plot_feature(dat, gene = .x, pt.alpha = 1))

plt <- plot_grid(plotlist = plts, nrow = 2)

save_plot("resampled_genes_newcols.pdf", plt, nrow = 2, ncol = 3,
          base_height = 4.25, base_width = 5.5)
```

## Average Gene Abundance

```{r avg_gene_abs}

#' normalize by library size (Reads per Million) and log2 + 1
norm_libsize <- function(sc_obj){
  norm_umi <- 1e6 * sweep(sc_obj$umi_matrix, 2, 
                                 sum(as.vector(sc_obj$umi_matrix)), "/")
  norm_reads <- 1e6 * sweep(sc_obj$read_matrix, 2, 
                                   sum(as.vector(sc_obj$read_matrix)), "/")
  sc_obj$log_norm_umi <- log2(norm_umi + 1)
  sc_obj$log_norm_reads <- log2(norm_reads + 1)
  sc_obj
}

sc_objs <- map(sc_objs, norm_libsize)

avg_expr <- map(sc_objs, ~log2(rowMeans(2^.x$log_norm_umi)))
avg_expr <- map(avg_expr, ~data_frame(gene = names(.x),
                                     expr = unname(.x)))
avg_expr_df <- bind_rows(avg_expr, .id = "library")

avg_expr_df <- avg_expr_df %>% 
  spread(library, expr)

avg_expr_df <- avg_expr_df %>% 
  filter(!(gene_pulldown == 0 & kirkpatrick == 0)) %>% 
  mutate(fc = log2(gene_pulldown / kirkpatrick),
         subsampled = ifelse(gene %in% genes,
                              T,
                              F))  %>% 
  arrange(subsampled)

subsampled_info <- filter(avg_expr_df, subsampled)

library(ggrepel)

ggplot(avg_expr_df, aes(kirkpatrick, fc)) +
  geom_point(aes(color = subsampled), size = 0.25) +
  scale_color_manual(values = color_palette) +
  geom_text_repel(data = subsampled_info,
                  aes(label = gene), 
                  arrow = arrow(length = unit(0.02, "npc"))) +
  labs(x = "Abundance in original library UMIs (log2)",
       y = expression(paste( " Log"[2], " ", frac("subsampled", "original")))) 
  
```

```{r reads}

avg_expr <- map(sc_objs, ~log2(rowMeans(2^.x$log_norm_reads)))
avg_expr <- map(avg_expr, ~data_frame(gene = names(.x),
                                     expr = unname(.x)))
avg_expr_df <- bind_rows(avg_expr, .id = "library")

avg_expr_df <- avg_expr_df %>% 
  spread(library, expr)

avg_expr_df <- avg_expr_df %>% 
  filter(!(gene_pulldown == 0 & kirkpatrick == 0)) %>% 
  mutate(fc = log2(gene_pulldown / kirkpatrick),
         subsampled = ifelse(gene %in% genes,
                              T,
                              F))  %>% 
  arrange(subsampled)

subsampled_info <- filter(avg_expr_df, subsampled)

library(ggrepel)

ggplot(avg_expr_df, aes(kirkpatrick, fc)) +
  geom_point(aes(color = subsampled), size = 0.25) +
  scale_color_manual(values = color_palette) +
  geom_text_repel(data = subsampled_info,
                  aes(label = gene), 
                  arrow = arrow(length = unit(0.02, "npc"))) +
  labs(x = "Abundance in original library reads (log2)",
       y = expression(paste( " Log"[2], " ", frac("subsampled", "original")))) 

```
