---
title: "plots for megakaryocyte pulldown"
author: "Kent Riemondy RBI"
date: '`r Sys.Date()`'
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

## Megakaryocyte targeting via LNA pulldown

Megakaryocytes represent a small proportion of the cells in a typical PBMC sample and therefore were selected as an additional test of the LNA pulldown techinique. Four megakaryocyte cells were reamplified after LNA pulldown. 

Following reamplification the 4 cell libraries were pooled and resequenced together. The raw fastqs were then processed using a Snakemake [pipeline](../../pipeline/Snakefile), to produce two processed data files:

1. A matrix with UMIs per cell (column) per gene (rows) (dge_matrix.txt)
1. A flatfile with per UMI information (umigroups.txt.gz)

This RMarkdown document will produce the following figures:

1. ...
1. ...
1. ...

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

```{r, message=F, warning=F, echo=F}
source("../../R/globals.R")
library(Seurat)
```

## Organize single cell data

First designate the libraries and the cells that were resampled. 

```{r cell_ids}
cells <- list(
  mkcell_pulldown = c("TGCGCAGCAGGTCGTC",
                      "ACTTGTTAGGACCACA",
                      "CCATTCGTCCCTGACT",
                      "TGTCCCAGTAAACACA"))

libs <- c(
  "kirkpatrick",
  "mkcell_pulldown")


lib_data_dir <- file.path(data_dir, 
                          "lna_cell",
                          "pbmc_expt")

bc_metadat <- read_tsv(file.path(lib_data_dir, 
                         "kirkpatrick", 
                         "fastq",
                         "original", 
                         "barcodes_from_10x_run.txt"),
                         col_names = c("cell_id", "barcode_10x"))

## original library to compare against
reflib <- "kirkpatrick"
## all resampled libs to plot
resampled_libs <- "mkcell_pulldown"
## reference resampled lib for resampled vs control plots
resampled_lib <- "mkcell_pulldown"

## pretty name for libraries
lib_names = c(
  kirkpatrick = "Original Library",
  mkcell_pulldown = "Resampled Library"
)

## pretty names for cells
cell_names = c(
  "TGCGCAGCAGGTCGTC-1" = "MK-cell-1",
  "ACTTGTTAGGACCACA-1" = "MK-cell-2",
  "CCATTCGTCCCTGACT-1" = "MK-cell-3",
  "TGTCCCAGTAAACACA-1" = "MK-cell-4")

sc_objs <- readRDS("processed_data.rds")

sc_metadat <- map(sc_objs, ~.x$meta_dat) %>% 
  bind_rows(.id = "library") %>% 
  mutate(library = factor(library, levels = libs)) %>% 
  arrange(resampled)

resampled_metadat <- filter(sc_metadat,
                             library != reflib) %>% 
  mutate(library = factor(library, 
                          levels = resampled_libs))
```

Load and organize a table for each library of read counts per cell per gene, and a table of umi counts per cell per gene. 

## Various stats

```{r stats}
sc_metadat %>% 
  filter(resampled) %>% 
  dplyr::select(library,
                cell_id, 
                total_reads,
                total_umis,
                n_genes) %>% 
  group_by(cell_id) %>% 
  arrange(library, .by_group = T) %>% 
  mutate(read_enrichment = nth(total_reads, 2) / nth(total_reads, 1)) %>% 
  group_by(library) %>% 
  summarize(median_genes = median(n_genes),
            max_read_enrichment = max(read_enrichment))

sc_metadat %>% 
  filter(resampled) %>% 
  group_by(library) %>% 
  summarize(mean_sat = mean(cDNA_duplication),
            sd_sat = sd(cDNA_duplication))


sc_metadat %>% 
  filter(resampled) %>% 
  dplyr::select(library,
                cell_id, 
                total_reads,
                total_umis,
                n_genes) %>% 
  group_by(cell_id) %>% 
  arrange(library, .by_group = T) %>% 
  summarize(more_genes = nth(n_genes, 2) - nth(n_genes, 1)) %>% 
  summarize(mean(more_genes), sd(more_genes))


 sc_metadat %>%
   group_by(library, resampled) %>%
   count()
 
```

## cDNA duplication rate

```{r cDNA_duplication_rate, fig.cap="Note the high level of sequencing saturation (0 = no-duplication, 1 = all duplicates) in the original library. Also note that the libraries tend to have higher saturation rates, after subsampling."}
sc_metadat <- map(sc_objs, ~.x$meta_dat) %>% 
  bind_rows(.id = "library") %>% 
  mutate(library = factor(library, levels = libs)) %>% 
  arrange(resampled)

plt <- ggplot(sc_metadat, aes(total_umis, cDNA_duplication)) +
  geom_point(aes(color = resampled), size = 0.5) +
  labs(x = expression(paste("# of UMIs")),
       y = "Sequencing saturation") + 
  scale_x_continuous(labels = scales::comma) +
  scale_color_manual(values = color_palette) +
  geom_text_repel(data = filter(sc_metadat, resampled),
                                     aes(label = ".",
                                         color = resampled),
                                     size = 0,
                                     force = 10,
                                     min.segment.length = 0,
                                     point.padding = 1.0,
                                     arrow = arrow(length = unit(0.3,
                                                                 "line"), 
                                                   angle = 35,
                                                   type = "open", 
                                                   ends = "last")
                     ) +
 # guides(colour = guide_legend(override.aes = list(size = 5))) + 
  facet_wrap(~library,
             labeller = labeller(library = lib_names)) +
  theme_cowplot(font_size = 16, line_size = 1)  +
  theme(legend.position = "none",
        plot.margin = unit(c(5.5, 20.5, 5.5, 5.5), 
                           "points"),
        axis.text.x = element_text(angle = 90, 
                                   hjust = 1,
                                   vjust = 0.5)) 

plt
save_plot("cDNA_duplication.pdf", plt,
          base_height = 4,
          base_aspect_ratio = 1.5)
```

## Read and umi counts per library

```{r plot_read_and_umi_counts,}

global_plot_theme <- theme(
        legend.position = "top",
        legend.text = element_text(size = 10),
        strip.text = element_text(size = 8))

resampled_metadat <- sc_metadat[sc_metadat$library != reflib, ] %>% 
  mutate(library = factor(library, 
                          levels = c(resampled_libs)))

unnorm_plt <- ggplot(resampled_metadat, 
                     aes(og_total_reads / 3, total_reads / 3, 
                         colour = resampled)) + 
  geom_point(size = 0.5) + 
  geom_abline(slope = 1) +
  facet_wrap(~library, nrow = 1) + 
#  coord_fixed() +
  xlab("original library\nreads count (Thousands)") +
  ylab("resampled library\nreads count (Thousands)") +
 # ggtitle("Raw reads associated with each cell barcode") +
  scale_colour_manual(name = "resampled:",
                      values = color_palette) +
  theme_cowplot(font_size = 10, line_size = .5) +
  theme(aspect.ratio = 1) + 
  global_plot_theme

norm_plt <- ggplot(resampled_metadat, aes(og_norm_total_reads / 1e3, 
                                           norm_total_reads / 1e3, 
                                           colour = resampled)) + 
  geom_point(size = 0.5) + 
  geom_abline(slope = 1) + 
  facet_wrap(~library, nrow = 1) + 
  xlab("original library\nRPM (Thousands)") +
  ylab("resampled library\nRPM (Thousands)") +
  scale_colour_manual(name = "resampled:",
                      values = color_palette) +
  theme_cowplot(font_size = 10, line_size = 0.5) +
  theme(aspect.ratio = 1) + 
  global_plot_theme

unnorm_umi_plt <- ggplot(resampled_metadat, 
                         aes(og_total_umis / 1e3, 
                             total_umis / 1e3, colour = resampled)) + 
  geom_point(size = 0.5) + 
  geom_abline(slope = 1) +
 # coord_fixed() +
  facet_wrap(~library, nrow = 1) + 
  xlab("original library\nUMI count (Thousands)") +
  ylab("resampled library\nUMI count (Thousands)") +
  scale_colour_manual(name = "resampled:",
                      values = color_palette) +
  theme_cowplot(font_size = 10, line_size = .5) +
    theme(aspect.ratio = 1) + 
  global_plot_theme

norm_umi_plt <- ggplot(resampled_metadat, 
                       aes(og_norm_total_umis / 1e3, 
                           norm_total_umis / 1e3,
                           colour = resampled)) + 
  geom_point(size = 0.5) + 
  geom_abline(slope = 1) + 
  facet_wrap(~library, nrow = 1) + 
  xlab("original library\nUMI normalized RPM (Thousands)") +
  ylab("resampled library\nUMIs per Million (Thousands)") +
  scale_colour_manual(name = "resampled:",
                      values = color_palette) +
  theme_cowplot(font_size = 10, line_size = 0.5) +
  theme(aspect.ratio = 1) + 
  global_plot_theme

plt <- plot_grid(unnorm_plt, norm_plt, unnorm_umi_plt, norm_umi_plt, 
                 labels = "AUTO",
                 align = 'hv')
plt
save_plot("reads_per_barcode_scatterplots.pdf", plt, 
          nrow = 2, ncol = 2, base_width = 8 )
```

## Enrichment of reads/umis 

```{r plot_ratio_reads_counts, fig.width=6, fig.height=5}

read <- ggplot(resampled_metadat, 
       aes(cell_id, norm_read_proportion, colour = resampled)) + 
  geom_point() + 
  labs(x = "Cell", 
       y = expression(paste( " Log"[2], " normalized reads ", 
                             frac("Resampled", "Original")))) +
  scale_colour_manual(name = "resampled:", values = color_palette) +
  facet_wrap(~library, nrow = 1) + 
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(axis.text.x = element_blank(),
        legend.position = "top",
        legend.text = element_text(size = 12))

umi <- ggplot(resampled_metadat, 
       aes(cell_id, norm_umi_proportion, colour = resampled)) + 
  geom_point() + 
  labs(x = "Cell", 
       y = expression(paste( "Log"[2], " normalized UMIs ", frac("Resampled", "Original")))) +
  scale_colour_manual(name = "resampled:", values = color_palette) +
  facet_wrap(~library, nrow = 1) + 
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(axis.text.x = element_blank(),
        legend.position = "top",
        legend.text = element_text(size = 12))

plt <- plot_grid(read, umi,
                 labels = "AUTO",
                 align = 'hv')
plt

ggsave("reads_umi_ratio_per_barcode_normalized.pdf", width = 8, height = 5)

umi <- ggplot(filter(resampled_metadat, library %in% resampled_libs), 
       aes(cell_id, norm_umi_proportion, colour = resampled)) + 
  geom_point() + 
  labs(x = "Cell", 
       y = expression(paste( "Log"[2], " normalized UMIs ", frac("Resampled", "Original")))) +
  scale_colour_manual(name = "resampled:", values = color_palette) +
  facet_wrap(~library, nrow = 1) + 
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(axis.text.x = element_blank(),
        legend.position = "top",
        legend.text = element_text(size = 12))

umi
ggsave("umi_ratio_per_cell.pdf", umi, width =5, height = 5)

umi_plots <- map(split(resampled_metadat, 
                       resampled_metadat$library),
                 function(x){
                   set.seed(42)
                   ggplot(x, 
                          aes(og_total_umis, 
                              norm_umi_proportion, 
                              colour = resampled)) + 
                     geom_point(size = 0.5) + 
                     geom_hline(aes(yintercept = 0), 
                                linetype ="dashed", 
                                color = "darkgrey") + 
                     geom_text_repel(data = filter(x, resampled),
                                     aes(label = "."),
                                     size = 0,
                                     force = 10,
                                     min.segment.length = 0,
                                     point.padding = 1.0,
                                     arrow = arrow(length = unit(0.3,
                                                                 "line"), 
                                                   angle = 35,
                                                   type = "open", 
                                                   ends = "last")
                     ) + 
                     labs(x = "Abundance in original library\n (UMIs)", 
                          
                          y = expression(paste( "Log"[2], " UMIs ", 
                                                frac("Resampled",
                                                     "Original")))) +
                     scale_x_continuous(labels = scales::comma) +      
                     scale_colour_manual(name = "resampled:", 
                                         values = color_palette) +
                    # guides(colour = guide_legend(override.aes = list(size = 5))) + 
                     theme_cowplot(font_size = 16, line_size = 1) +
                     theme(legend.position = "none",
                           # legend.text = element_text(size = 12),
                           plot.margin = unit(c(5.5, 20.5, 5.5, 5.5),
                                              "points"))})

plt <- plot_grid(plotlist = umi_plots, nrow = 1)
save_plot("umi_ratio_MA.pdf", plt)
```

```{r per_cell_ratio_reads_counts}
ggplot(resampled_metadat, aes(resampled, 
                read_proportion, fill = resampled)) + 
  geom_boxplot(coef = Inf) + 
  facet_wrap(~library, nrow = 1) +
  xlab("Selected for Reamplification") +
  ylab(expression(paste( "Log"[2], " Reads ", frac("Resampled", "Original")))) +
  scale_fill_manual(name = "resampled:",
                    values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 0.5) +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 12)
  )
ggsave("reads_ratio_per_barcode_boxplot.pdf", width = 6, height = 5)

ggplot(resampled_metadat, aes(resampled, 
                umi_proportion, fill = resampled)) + 
  geom_boxplot(coef = Inf) + 
  facet_wrap(~library, nrow = 1) +
  xlab("Selected for Reamplification") +
  ylab(expression(paste( "Log"[2], " UMIs ", frac("Resampled", "Original")))) +
  scale_fill_manual(name = "resampled:",
                    values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 0.5) +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 12)
  )
ggsave("umi_ratio_per_barcode_boxplot.pdf", width = 6, height = 5)

ggplot(resampled_metadat, aes(resampled, 
                norm_read_proportion, fill = resampled)) + 
  geom_boxplot(coef = Inf) + 
  facet_wrap(~library, nrow = 1) +
  xlab("Selected for Reamplification") +
  ylab(expression(paste( "Log"[2], " normalized Reads ", frac("Resampled", "Original")))) +
  scale_fill_manual(name = "resampled:",
                    values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 0.5) +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 12)
  )
ggsave("norm_reads_ratio_per_barcode_boxplot.pdf", width = 6, height = 5)

ggplot(resampled_metadat, aes(resampled, 
                norm_umi_proportion, fill = resampled)) + 
  geom_boxplot(coef = Inf) + 
  facet_wrap(~library, nrow = 1) +
  xlab("Selected for Reamplification") +
  ylab(expression(paste( "Log"[2], " normalized UMIs ", frac("Resampled", "Original")))) +
  scale_fill_manual(name = "resampled:",
                    values = color_palette) +
  theme_cowplot(font_size = 16, line_size = 0.5) +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_blank(),
    legend.position = "top",
    legend.text = element_text(size = 12)
  )
ggsave("norm_umi_ratio_per_barcode_boxplot.pdf", width = 6, height = 5)

```

```{r  total_read_enrichment}

dat <- group_by(resampled_metadat, library) %>%
  filter(library != reflib) %>% 
  mutate(total_new = sum(total_reads, na.rm = T), 
         total_old = sum(og_total_reads, na.rm = T))

dat_group <- group_by(dat, library, resampled) %>% 
  summarize(total_new = sum(total_reads, 
                            na.rm = T) / unique(total_new), 
            total_old = sum(og_total_reads, 
                            na.rm = T) / unique(total_old)) %>% 
  gather(lib, percent_lib, -library, -resampled ) %>%
  mutate(lib = factor(lib, levels = c("total_old", "total_new"), 
                      labels = c("original\nlibrary", "resampled\nlibrary")))

ggplot(dat_group, aes(lib, percent_lib, fill = resampled)) + 
  geom_bar(stat = "identity") + 
  ylab("Fraction of\n Reads Assigned") +
  scale_fill_manual(name = "resampled:",
                    values = color_palette) +
  facet_wrap(~library) + 
  theme_cowplot(font_size = 16, line_size = 1) +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 0.5, vjust = 0.5),
    legend.position = "top",
    legend.text = element_text(size = 16)
  )

ggsave("proportion_reads_all_barcode_barplot.pdf", width = 7, height = 7)

dat_group %>% 
 dplyr::rename(Method = library) %>% 
  spread(lib, percent_lib) %>% 
  mutate(`Targeted Library Read Fold-Enrichment` = `resampled\nlibrary` / `original\nlibrary`) %>%
  filter(resampled == T) %>% 
  dplyr::select(Method, `Targeted Library Read Fold-Enrichment`)
```

## Genes detected

```{r gene_plot}

og_genes <- filter(sc_metadat, 
                   library == reflib) %>% 
  dplyr::select(barcode_10x, 
                og_genes = n_genes)

resampled_metadat_mod <- left_join(resampled_metadat, 
                                og_genes,
                                by = "barcode_10x") %>% 
  filter(library != reflib)

plt <- ggplot(resampled_metadat_mod, aes(og_genes, 
                               n_genes, colour = resampled)) + 
  geom_point(size = 0.5) + 
  ylab("Genes detected\n(resampled library)") +
  xlab("Genes detected\n(original library)") + 
  scale_colour_manual(name =  "resampled:", values = color_palette) +
  facet_wrap(~library, nrow = 1) + 
  guides(colour = guide_legend(override.aes = list(size = 3))) +
  theme(
    legend.position = "top",
    legend.text = element_text(size = 10)
  )
plt
save_plot("genes_detected_scatterplot.pdf", plt)
```

### Parse out new versus previously identified genes

```{r per_cell_gene_recovery}

genes_recovered <- resampled_metadat %>% 
  dplyr::filter(library != reflib) %>% 
  dplyr::select(cell_id, 
                library,
                resampled,
                shared_genes,
                not_recovered_genes, 
                new_genes)

genes_recovered <- gather(genes_recovered, 
                          key = type, value = count, 
                          -cell_id, -resampled, -library)
genes_recovered <- mutate(genes_recovered,
                          type = str_replace_all(type, "_", "\n"))

plt <- ggplot(genes_recovered, 
       aes(cell_id, count)) +
  geom_point(aes(color = resampled),
             size = 0.6,
             alpha = 0.75) +
  facet_grid(type ~ library) +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text.y = element_text(size = 12,
                                    margin = margin(0,0.2,0,0.2, "cm"))) +
  scale_color_manual(values = color_palette)

plt 

save_plot("new_genes_detected.pdf", plt, base_width = 8, base_height = 8)

targeted <- genes_recovered %>% 
  dplyr::filter(library %in% resampled_libs,
                resampled) 
targeted

plt_dat <- genes_recovered %>% 
  dplyr::filter(resampled) %>% 
  mutate(type = ifelse(type == "new\ngenes",
                        "Newly\ndetected\ngenes",
                        ifelse(type == "shared\ngenes",
                               "Previously\ndetected\ngenes",
                               ifelse(type == "not\nrecovered\ngenes",
                                      "Previously\ndetected\ngenes\nnot recovered",
                                      NA))),
         cell_id = factor(cell_id, 
                          levels = c(str_c(cells[[resampled_lib]],
                                           "-1"), 
                          "Not targeted\nbarcodes\n(mean)")))


plt <- ggplot(plt_dat, 
       aes(cell_id, count)) +
  geom_bar(aes(fill = type),
           stat = "identity") +
  labs(y = "# of genes") + 
  scale_x_discrete(labels = cell_names) + 
  scale_y_continuous(labels = scales::comma) + 
  scale_fill_manual(values = palette_okabeito, name = "") +
  guides(fill = guide_legend(override.aes = list(size = 0.25))) +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45,
                                   hjust = 1),
        legend.position = "top",
        plot.margin = unit(c(5.5, 50.5, 5.5, 5.5), 
                           "points"))
      #  legend.key.size = unit(0.25, "pt")) 

plt
save_plot("new_genes_barplot.pdf", plt, 
          base_width = 4, base_height = 5)

```

```{r}
plt_dat <- genes_recovered %>% 
  mutate(type = ifelse(type == "new\ngenes",
                        "Newly\ndetected\ngenes",
                        ifelse(type == "shared\ngenes",
                               "Previously\ndetected\ngenes",
                               ifelse(type == "not\nrecovered\ngenes",
                                      "Previously\ndetected\ngenes\nnot recovered",
                                      NA))))

plt_dat <- mutate(plt_dat, 
                  resampled = ifelse(resampled,
                                     "Targeted barcodes",
                                     "Not targeted barcodes"))
plt <- ggplot(plt_dat, 
       aes(type, count, 
           group = resampled)) +
  geom_jitter(aes(color = resampled),
              size = 0.25,
              position = position_jitterdodge(dodge.width=1)) +
  stat_summary(fun.y = "mean", 
               fun.ymin = "mean", 
               fun.ymax = "mean", 
               size= 0.3, 
               geom = "errorbar",
               position = "dodge",
               linetype = "dashed") +
  labs(y = "# of genes") + 
  scale_x_discrete(labels = cell_names) + 
  scale_y_continuous(labels = scales::comma) + 
  scale_color_manual(values = palette_okabeito, name = "") +
  guides(colour = guide_legend(override.aes = list(size = 2))) +
  theme(axis.title.x = element_blank(),
        legend.position = "top",
        plot.margin = unit(c(5.5, 50.5, 5.5, 5.5), 
                           "points"),
        legend.key.size = unit(10, "pt")) 

save_plot("new_genes_boxplots.pdf", plt, 
          base_aspect_ratio = 1.1, base_height = 4) 

```


```{r shared_new_genes}
new_genes <- sc_objs$mkcell_pulldown$new_genes[cells$mkcell_pulldown]

all_genes <- unique(unlist(new_genes))

```

```{r eval = F}
library(UpSetR)
pdf("new_genes_shared.pdf")
  upset(fromList(new_genes), order.by = "freq")
dev.off()

upset(fromList(new_genes), order.by = "freq")
```

Compare new genes to previously detected genes in the cluster

### Plot expression per cell 

#### MA plots

```{r ma_per_cell}

calc_ma <- function(xmat, ymat, cell = "GCAGTTAAGTGTCCAT"){
  x_rn <- rownames(xmat)
  y_rn <- rownames(ymat)
  xmat <- log2(xmat + 1)
  ymat <- log2(ymat + 1)
  
  rownames(xmat) <- x_rn
  rownames(ymat) <- y_rn
  m <- Matrix::rowMeans(log2(((2^ymat + 2^xmat) / 2) + 1))
  a <- xmat[, cell] - ymat[, cell]
  data_frame(gene = names(a),
             mean_expression_log2 = m,
             log2_diff = a)
}


genes_to_plot <- rownames(sc_objs[[resampled_lib]]$umi_matrix)
cols <- colnames(sc_objs[[resampled_lib]]$norm_umi)
cell_ids <- cells[[resampled_lib]]

## append genes to reference library if necessary
ref_mat <- standardize_rows(sc_objs[[resampled_lib]]$umi_matrix[, cols],
                            sc_objs[[reflib]]$umi_matrix[, cols])

ma_dat <- map(cell_ids,
    ~calc_ma(sc_objs[[resampled_lib]]$umi_matrix[genes_to_plot, cols], 
             ref_mat[genes_to_plot, cols],
             cell = .x))

names(ma_dat) <- names(cell_names)
ma_dat <- bind_rows(ma_dat, .id = "cell")

plot_ma <- function(df){
  n_up <- filter(df, log2_diff > 0) %>% 
    group_by(cell) %>% 
    summarize(n = n(), n = paste0("up = ", n))
  
  n_down <- filter(df, log2_diff < 0) %>% 
    group_by(cell) %>% 
    summarize(n = n(), n = paste0("down = ", n))
  
  if (nrow(n_down) == 0) {
    n_down = data_frame(cell = df$cell %>% unique(),
                        n = "down = 0")
  }
  plt <- ggplot(df,
         aes(mean_expression_log2,
             log2_diff)) +
    geom_hline(aes(yintercept = 0), linetype = "dashed", colour = "grey") + 
    geom_point(size = 0.25) +
    geom_text(data = n_up, aes(x = max(ma_dat$mean_expression_log2) * 0.9, 
                               y = max(ma_dat$log2_diff) * 1.2, 
                               label = n)) +
    geom_text(data = n_down, aes(x = max(ma_dat$mean_expression_log2) * 0.9, 
                                 y = min(ma_dat$log2_diff) * 1.2, 
                                 label = n)) + 
    facet_wrap(~cell, nrow = 1) +
    labs(x = expression(paste("Abundance (log"[2], ")")),
         y = expression(paste(frac("Resampled","Original"), " (log"[2], ")")))
  plt
}

plt <- plot_ma(ma_dat)
plt
save_plot("per_cell_MA_plot_all_genes.pdf", plt, base_height = 6)

## Shared genes
ma_dat <- map(cell_ids,
  function(x){
    genes_to_plot <- sc_objs[[resampled_lib]]$shared_genes[[x]]
    calc_ma(sc_objs[[resampled_lib]]$umi_matrix[genes_to_plot,
                                                              cols],
            ref_mat[genes_to_plot, cols],
             cell = x)
})

names(ma_dat) <- names(cell_names)
ma_dat <- bind_rows(ma_dat, .id = "cell")
plt <- plot_ma(ma_dat)
plt
save_plot("per_cell_MA_plot_shared_genes.pdf", plt, base_height = 6)


## New genes
ma_dat <- map(cell_ids,
  function(x){
    genes_to_plot <- sc_objs[[resampled_lib]]$new_genes[[x]]
    calc_ma(sc_objs[[resampled_lib]]$umi_matrix[genes_to_plot,
                                                              cols],
            ref_mat[genes_to_plot, cols],
             cell = x)
})

names(ma_dat) <- names(cell_names)
ma_dat <- bind_rows(ma_dat, .id = "cell")
plt <- plot_ma(ma_dat)
plt
save_plot("per_cell_MA_plot_new_genes.pdf", plt, base_height = 6)

```



#### Histograms

```{r get_expr, fig.height = 12}
get_expr <- function(xmat, ymat, cell = "GCAGTTAAGTGTCCAT"){
  xrows <- rownames(xmat)
  xmat <- log2(xmat[, cell] + 1)
  ymat <- log2(ymat[xrows, cell] + 1)
  data_frame(
    gene = xrows,
    resampled = xmat,
    original = ymat) %>% 
    gather(library, 
           Expression, -gene)
}

plot_histogram <- function(df){
  ggplot(df, 
         aes_string("Expression")) +
    geom_density(aes_string(fill = "library"),
                 alpha = 0.66) +
    scale_fill_viridis_d(name = "") +
    facet_wrap(~cell, nrow = 1) +
    theme(legend.position = "top",
          strip.text = element_text(size = 8)) 
}

expr_dat <- map(cell_ids,
  function(x){
    genes_to_plot <- names(sc_objs[[resampled_lib]]$genes_detected[[x]])
    get_expr(sc_objs[[resampled_lib]]$umi_matrix[genes_to_plot,
                                                              cols],
            ref_mat[genes_to_plot, cols],
             cell = x)
})

names(expr_dat) <- cell_ids
expr_dat <- bind_rows(expr_dat, .id = "cell")
expressed_in_resampled_plt <- plot_histogram(expr_dat)
expressed_in_resampled_plt

expr_dat <- map(cell_ids,
  function(x){
    genes_to_plot <- sc_objs[[resampled_lib]]$shared_genes[[x]]
    get_expr(sc_objs[[resampled_lib]]$umi_matrix[genes_to_plot,
                                                              cols],
            ref_mat[genes_to_plot, cols],
             cell = x)
})

names(expr_dat) <- cell_ids
expr_dat <- bind_rows(expr_dat, .id = "cell")
share_genes_plt <- plot_histogram(expr_dat)
share_genes_plt

expr_dat <- map(cell_ids,
  function(x){
    genes_to_plot <- sc_objs[[resampled_lib]]$new_genes[[x]]
    get_expr(sc_objs[[resampled_lib]]$umi_matrix[genes_to_plot,
                                                              cols],
            ref_mat[genes_to_plot, cols],
             cell = x)
})

names(expr_dat) <- cell_ids
expr_dat <- bind_rows(expr_dat, .id = "cell")
new_gene_plt <- plot_histogram(expr_dat)


plts <- list(
  expressed_in_resampled_plt,
  share_genes_plt,
  new_gene_plt
)

plt <- plot_grid(plotlist = plts, ncol = 1)
plt

save_plot("expression_histograms.pdf", plt, 
          ncol = 1, nrow = 3,
          base_height = 4,
          base_aspect_ratio = 2)
```


#### scatterplots

```{r get_expr_scat}
get_paired_expr <- function(xmat, ymat, cell = "GCAGTTAAGTGTCCAT"){
  xrows <- rownames(xmat)
  xmat <- log2(xmat[, cell] + 1)
  ymat <- log2(ymat[xrows, cell] + 1)
  data_frame(
    gene = xrows,
    resampled = xmat,
    original = ymat)
}

plot_scatter <- function(df){
  ggplot(df, 
         aes_string("original", "resampled")) +
    geom_point(size = 0.5) + 
    geom_abline(aes(slope = 1, intercept = 0)) +
    facet_wrap(~cell, nrow = 1) +
    coord_fixed() + 
    theme(legend.position = "top",
          strip.text = element_text(size = 8)) 
}

expr_dat <- map(cell_ids,
  function(x){
    genes_to_plot <- names(sc_objs[[resampled_lib]]$genes_detected[[x]])
    get_paired_expr(sc_objs[[resampled_lib]]$umi_matrix[genes_to_plot,
                                                              cols],
            ref_mat[genes_to_plot, cols],
             cell = x)
})

names(expr_dat) <- cell_ids
expr_dat <- bind_rows(expr_dat, .id = "cell")
expressed_in_resampled_plt <- plot_scatter(expr_dat)
expressed_in_resampled_plt

expr_dat <- map(cell_ids,
  function(x){
    genes_to_plot <- sc_objs[[resampled_lib]]$shared_genes[[x]]
    get_paired_expr(sc_objs[[resampled_lib]]$umi_matrix[genes_to_plot,
                                                              cols],
            ref_mat[genes_to_plot, cols],
             cell = x)
})

names(expr_dat) <- cell_ids
expr_dat <- bind_rows(expr_dat, .id = "cell")
share_genes_plt <- plot_scatter(expr_dat)
share_genes_plt

expr_dat <- map(cell_ids,
  function(x){
    genes_to_plot <- sc_objs[[resampled_lib]]$new_genes[[x]]
    get_paired_expr(sc_objs[[resampled_lib]]$umi_matrix[genes_to_plot,
                                                              cols],
            ref_mat[genes_to_plot, cols],
             cell = x)
})

names(expr_dat) <- cell_ids
expr_dat <- bind_rows(expr_dat, .id = "cell")
new_gene_plt <- ggplot(expr_dat, aes(cell,resampled)) +
  geom_jitter(alpha = 0.55) +
  geom_violin(aes(fill = cell)) +
  ylim(0, max(expr_dat$resampled) * 1.10) + 
  scale_fill_brewer(palette = "Set1") +
    theme(axis.text.x = element_text(angle = 90, 
                                     hjust = 1, vjust = 0.5),
          legend.pos = "none",
          axis.title.x = element_blank()) 

new_gene_plt

plts <- list(
  expressed_in_resampled_plt,
  share_genes_plt
)

plt <- plot_grid(plotlist = plts, ncol = 1)
plt

save_plot("expression_scatterplots.pdf", plt, 
          ncol = 1, nrow = 3,
          base_height = 4,
          base_aspect_ratio = 2)

save_plot("expression_newgenes_violinplots.pdf", new_gene_plt, 
          base_height = 6,
          base_aspect_ratio = 0.5)
```

## UMIs detected

### Parse out new verses old UMIs
```{r read_umi_seqs}

umi_files <- file.path(lib_data_dir, libs, "umis", "umigroups.txt.gz")

umi_summaries <- map(umi_files[2],
                   ~compare_umis(umi_files[1], .x, return_summary = T))

names(umi_summaries) <- libs[2] 

umi_summary <- bind_rows(umi_summaries, .id = "library")

umis_recovered <- umi_summary %>% 
  gather(class, count, -barcode_10x, -library) 

## annotate with resampled or not
cell_annot <- data_frame(barcode_10x = cells,
                         library = libs,
                         resampled = T) %>% 
  unnest()

umis_recovered <- umis_recovered %>% 
  mutate(resampled = ifelse(barcode_10x %in% unlist(cells),
                             T,
                             F)) %>% 
  arrange(resampled)

plt <- ggplot(umis_recovered, 
       aes(barcode_10x, count)) +
  geom_point(aes(colour = resampled),
             size = 0.6,
             alpha = 0.75) +
  facet_grid(library ~ class) +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        strip.text.y = element_text(size = 12,
                                    margin = margin(0,0.2,0,0.2, "cm"))) +
  scale_color_manual(values = color_palette)

plt

targeted <- umis_recovered %>% 
  dplyr::filter(resampled) %>% 
  mutate()

plt_dat <- umis_recovered %>% 
  dplyr::filter(resampled) %>% 
  mutate(cell_id = str_c(barcode_10x, "-1"),
         class = ifelse(class == "new_umis",
                        "Newly\ndetected\nUMIs",
                        ifelse(class == "shared_umis",
                               "Previously\ndetected\nUMIs",
                               ifelse(class == "not_detected_umis",
                                      "Previously\ndetected\nUMIs\nnot recovered",
                                      NA))))

plt_dat <- mutate(plt_dat, 
                  cell_id = cell_names[cell_id],
                  cell_id = ifelse(is.na(cell_id),
                                   "Not targeted\nbarcodes\n(mean)",
                                   cell_id),
                  cell_id = factor(cell_id, 
                                   levels = c(
                                     cell_names,
                                     "Not targeted\nbarcodes\n(mean)"
                                   )))
plt <- ggplot(plt_dat, 
       aes(cell_id, count)) +
  geom_bar(aes(fill = class),
           stat = "identity") +
  labs(y = "# of UMIs") + 
  scale_y_continuous(labels = scales::comma) + 
  scale_fill_manual(values = palette_okabeito, name = "") +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45,
                                   hjust = 1),
        legend.position = "top",
        plot.margin = unit(c(5.5, 50.5, 5.5, 5.5), 
                           "points"),
        legend.key.size = unit(10, "pt")) 

save_plot("new_umis_barplot.pdf", plt,
          base_width = 4, base_height = 5)


plt_dat <- umis_recovered %>% 
  mutate(cell_id = str_c(barcode_10x, "-1")) %>% 
  mutate(class = ifelse(class == "new_umis",
                        "Newly\ndetected\nUMIs",
                        ifelse(class == "shared_umis",
                               "Previously\ndetected\nUMIs",
                               ifelse(class == "not_detected_umis",
                                      "Previously\ndetected\nUMIs\nnot recovered",
                                      NA))))

plt_dat <- mutate(plt_dat, 
                  resampled = ifelse(resampled,
                                     "Targeted barcodes",
                                     "Not targeted barcodes"))

plt <- ggplot(plt_dat, 
       aes(class, count, 
           group = resampled)) +
  geom_jitter(aes(color = resampled),
              size = 0.25,
              position = position_jitterdodge(dodge.width=1)) +
  stat_summary(fun.y = "mean", 
               fun.ymin = "mean", 
               fun.ymax = "mean", 
               size= 0.3, 
               geom = "errorbar",
               position = "dodge",
               linetype = "dashed") +
  labs(y = "# of UMIs") + 
  scale_x_discrete(labels = cell_names) + 
  scale_y_continuous(labels = scales::comma) + 
  scale_color_manual(values = palette_okabeito, name = "") +
  guides(colour = guide_legend(override.aes = list(size = 2))) +
  theme(axis.title.x = element_blank(),
        legend.position = "top",
        plot.margin = unit(c(5.5, 50.5, 5.5, 5.5), 
                           "points"),
        legend.key.size = unit(10, "pt")) 

save_plot("new_umis_boxplots.pdf", plt, 
           base_aspect_ratio = 1.1, base_height = 4) 

```

```{r}
umi_seqs <- map(umi_files[2],
                ~compare_umis(umi_files[1], .x, return_summary = F))

names(umi_seqs) <- libs[2] 

new_umis <- map(umi_seqs, 
    ~filter(.x, 
       barcode_10x %in% unlist(cells), 
       !is.na(count.y), 
       is.na(count.x))  %>% 
  separate(umi_molecule, c("seq", "gene"), sep = "::") %>% 
    select(-starts_with("count"))) 

old_umis <- map(umi_seqs, 
    ~filter(.x, 
       barcode_10x %in% cells[[resampled_lib]], 
       !is.na(count.x),
       !is.na(count.y))  %>% 
  separate(umi_molecule, c("seq", "gene"), sep = "::") %>% 
    select(-starts_with("count"))) 


umi_edit_dist <- map2(new_umis,
                      old_umis,
                     ~left_join(.x, .y, 
               by = c("barcode_10x", "gene")) %>% 
  na.omit() %>% 
  mutate(ed = kentr::get_hamming_pairs(seq.x, seq.y)) %>% 
  group_by(barcode_10x, seq.y, gene) %>% 
  summarize(min_ed = as.integer(min(ed))) %>% 
  ungroup())

umi_edit_dist <- bind_rows(umi_edit_dist, 
                           .id = "library")

umi_edit_dist <-  mutate(umi_edit_dist, 
                         cells = str_c(barcode_10x, "-1"))

plt <- ggplot(umi_edit_dist, aes(cells, 
                          min_ed)) + 
  geom_boxplot(coef = Inf) +
  scale_x_discrete(labels = cell_names) +
  #facet_wrap(~library, scales = "free_x", 
  #           labeller = labeller(library = lib_names)) +
  scale_y_continuous(breaks= scales::pretty_breaks()) +
  labs(y = "Minimum edit distance\noriginal vs. new UMIs") +
  theme(axis.text.x = element_text(angle = 90,
                                   hjust = 1, 
                                   vjust = 0.5),
        axis.title.x = element_blank())
save_plot("umi_edit_distance.pdf", plt)
```


## tSNE analysis

### original library tSNE

```{r plt_tsne}
sobj <- readRDS("original_sobj.rds")
plt <- TSNEPlot(sobj, 
                colors.use = c(brewer.pal(12, "Paired"), 
                               brewer.pal(9, "Set1")),
                do.label = T) + 
  labs(title = "PBMCs") +
  theme(legend.position = "none")

save_plot("original_mkcells_tsne.pdf", plt, 
          base_height = 4.25, base_width = 4.25)

sobj <- SetAllIdent(sobj, "resampled")

plt <- TSNEPlot(sobj, 
                colors.use = c(brewer.pal(12, "Paired"), 
                               brewer.pal(9, "Set1")),
                do.label = F, 
                pt.size = 0.5) + 
  theme(legend.position = "none")
save_plot("original_selected_mkcells_tsne.pdf", plt, 
          base_height = 4.25, base_width = 4.25)
```

```{r pbmc_markers}
immune_markers <- c(t_cells = "CD3E", 
                    cd8_t ="CD8A", 
                    cytotoxic_t = "NKG7", 
                    dendritic = "FCER1A", 
                    megakaryocyte = "PF4", 
                    b_cell = "CD79A", 
                    cd4_4 = "IL7R", 
                    monocyte = "CD14",
                    RBC = "HBB",
                    NK = "NCAM1",
                    pDC = "LILRA4",
                    monocyte = "FCGR3A")

plts <- map(immune_markers, ~plot_feature(sobj, gene = .x, 
                                          pt.alpha = 1, 
                                          legend_name = .x))
plt <- plot_grid(plotlist = plts,  nrow = 4)
save_plot("og_tsne_markers.pdf", plt,
          nrow = 4, ncol = 3)

sobj <- SetAllIdent(sobj, "res.1.2")

```

```{r cell_percentages}
cell_percents <- group_by(sobj@meta.data, cell_labels) %>% 
  summarize(n_cells = n()) %>% 
  mutate(total_cells = sum(n_cells), 
         percentage = 100 * (n_cells / total_cells)) %>% 
  select(-total_cells)

cell_percents
```

```{r}
# find region around the megakaryocyte cells to highlight
tsne_dat <- GetCellEmbeddings(sobj, "tsne") %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell") %>% 
  left_join(sobj@meta.data %>%
              as.data.frame() %>% 
              tibble::rownames_to_column("cell"),
            ., by = "cell") %>% 
  left_join(., cell_percents, by = "cell_labels") %>% 
  mutate(cell_labels_annotated = str_c(cell_labels, 
                            " (",
                            signif(percentage, 3),
                            "%)"))
dat <- filter(tsne_dat, cell_labels == "Megakaryocytes")

x_min = min(dat[, c("tSNE_1")]) - 1.5
x_max = max(dat[, c("tSNE_1")]) + 1.5
y_min = min(dat[, c("tSNE_2")]) - 1.5
y_max = max(dat[, c("tSNE_2")]) + 1.5

labeled_plt <- ggplot(tsne_dat, 
  aes(tSNE_1, tSNE_2)) + 
  geom_point(aes(color = cell_labels_annotated), size = 0.1) + 
  scale_color_manual(values = brewer.pal(10, "Paired"),
                     name = "") +
  labs(title = "",
       x = "tSNE 1",
       y = "tSNE 2") +
  geom_rect(aes(xmin = x_min,
                xmax = x_max,
                ymin = y_min,
                ymax = y_max), alpha = 0, color = "black", 
            size = 0.5) +
   theme_cowplot() + 
  guides(color = guide_legend(override.aes = list(size = 4))) + 
  theme(legend.pos = "right",
        aspect.ratio = 1) 


sub_plt <- ggplot(arrange(tsne_dat, resampled), 
  aes(tSNE_1, tSNE_2)) + 
  geom_point(aes(color = resampled), size = 2) + 
  scale_color_manual(values = c(brewer.pal(10, 
                                           "Paired")[7],
                                color_palette[2]),
                     name = "Resampled") +
  coord_cartesian(xlim = c(x_min, x_max),
                  ylim = c(y_min, y_max)) +
  guides(color = guide_legend(override.aes = list(size = 5))) + 
  labs(x = "tSNE 1",
       y = "tSNE 2")  +
  theme(aspect.ratio = 1)

plt <- plot_grid(labeled_plt, sub_plt, align = 'hv', 
                 nrow = 1)
save_plot("original_pbmc_annotated_tsne.pdf", plt, 
          base_width = 5.5, 
          base_aspect_ratio = 1.5, nrow = 1, ncol = 2)


arrow_plt <- ggplot(tsne_dat, 
  aes(tSNE_1, tSNE_2)) + 
  geom_point(aes(color = cell_labels_annotated), size = 0.1) + 
  scale_color_manual(values = brewer.pal(10, "Paired"),
                     name = "") +
  labs(title = "",
       x = "tSNE 1",
       y = "tSNE 2") +
    geom_text_repel(data = filter(tsne_dat, resampled),
                                     aes(label = "."),
                                     size = 0,
                                     force = 1,
                                     min.segment.length = 0,
                    box.padding = 0.5,
                    segment.size = 0.35,
                                     point.padding = 0.05,
                                     arrow = arrow(length = unit(0.1,
                                                                 "line"), 
                                                   angle = 35,
                                                   type = "open", 
                                                   ends = "last"),
                    seed = 42
                     ) +
   theme_cowplot() + 
  guides(color = guide_legend(override.aes = list(size = 4))) + 
  theme(legend.pos = "right",
        aspect.ratio = 1,
        legend.text = element_text(size = 10),
        axis.text = element_text(size = 10),
        axis.title = element_text(size = 10)) 

save_plot("original_pbmc_annotated_tsne_arrows.pdf", arrow_plt, 
          base_width = 5.5, 
          base_aspect_ratio = 1.5)
```


```{r find_mk_markers}
all_markers <- read_tsv("original_pbmc_markers.txt")
cell_mdata <- sobj@meta.data %>% 
  tibble::rownames_to_column("cell") 
```

### original library tSNE supplemented with resampled barcodes

```{r resampled_plots}
sobj <- readRDS("rs_sobj.rds")
plt <- TSNEPlot(sobj, 
                colors.use = c(brewer.pal(12, "Paired"), 
                               brewer.pal(9, "Set1")),
                do.label = T) + 
  theme(legend.position = "none")
plt
save_plot("original_with_resampled_mkcells_tsne.pdf", plt, 
          base_height = 4.25, base_width = 4.25)

### Plot original and resampled close-up
pf4 <- plot_feature(sobj, gene = "PF4", legend_name = "PF4",
                    pt.alpha = 1, pt.size = 0.1) +
  labs(title = "",
       x = "tSNE 1",
       y = "tSNE 2") + 
  theme_cowplot() +
  theme(legend.pos = "right")

# find region around the resampled cells to highlight

tsne_dat <- GetCellEmbeddings(sobj, "tsne") %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column("cell") %>% 
  left_join(sobj@meta.data %>%
              as.data.frame() %>% 
              tibble::rownames_to_column("cell"),
            ., by = "cell")

og_dat <- filter(tsne_dat, resampled == "original cell")

sub_dat <-  filter(tsne_dat, resampled == "resampled") %>% 
  mutate(cell = str_replace(cell, "::resampled", "")) %>% 
  select(cell, tSNE_1_resampled = tSNE_1, tSNE_2_resampled = tSNE_2)

dat <- left_join(og_dat, sub_dat, by = "cell")

xpos <- colMeans(as.matrix(dat[, c("tSNE_1", "tSNE_2")]) )[1]
ypos <- colMeans(as.matrix(dat[, c("tSNE_1", "tSNE_2")]) )[2]
x_min = xpos - 1.5
x_max = xpos + 1.5
y_min = ypos - 1.5
y_max = ypos + 1.5

resampled_cells <- ggplot(tsne_dat, 
  aes(tSNE_1, tSNE_2)) + 
  geom_point(aes(color = resampled), size = 0.1) + 
  scale_color_manual(values = c("lightgrey",
                                color_palette),
                     name = "",
                     labels = c("not resampled" = "Not Resampled",
                                "original cell" = "Original Cell",
                                "resampled" = "Resampled Cell")) +
  geom_rect(aes(xmin = x_min,
                xmax = x_max,
                ymin = y_min,
                ymax = y_max), alpha = 0, color = "black",
            size = 0.5) +
  labs(x = "tSNE 1", y = "tSNE 2") +
  theme_cowplot() + 
  guides(ncol = 1, 
         color = guide_legend(override.aes = list(size = 4))) + 
  theme(legend.pos = "none") 


close_up <- ggplot(tsne_dat, 
  aes(tSNE_1, tSNE_2)) + 
  geom_point(aes(color = resampled), size = 2) + 
  scale_color_manual(values = c("lightgrey",
                                color_palette),
                     name = "",
                     labels = c("not resampled" = "Not Resampled",
                                "original cell" = "Original Cell",
                                "resampled" = "Resampled Cell")) +
   coord_cartesian(xlim = c(x_min, x_max),
                  ylim = c(y_min, y_max)) +
  geom_segment(data = dat, 
                               aes(x = tSNE_1,
                                   y = tSNE_2, 
                                   xend = tSNE_1_resampled,
                                   yend = tSNE_2_resampled),
                               linejoin = "mitre",
                               arrow = arrow(length = unit(0.02,
                                                           "npc"))) + 
   labs(x = "tSNE 1", y = "tSNE 2") +
  theme_cowplot() + 
  guides(ncol = 1, 
         color = guide_legend(override.aes = list(size = 4))) + 
  theme(legend.pos = "none")
  
plt <- plot_grid(pf4, resampled_cells, 
          close_up, nrow = 1, rel_widths = c(1.1, 1, 1))

save_plot("resampled_vs_original_closeup.pdf", 
          plt, nrow = 1, ncol = 3, base_aspect_ratio = 1)
```

```{r lib_size_normalized_cell_plots}
og_cells <- rownames(sobj@meta.data)[sobj@meta.data$resampled == "original cell"]
rs_cells <- str_c(og_cells, "::resampled")
names(og_cells) <- og_cells
expr <- map2_dfr(og_cells, rs_cells, function(x, y){
  xdat <- sobj@data[, x, drop = T]
  ydat <- sobj@data[, y, drop = T]
  expr <- data_frame(gene = rownames(sobj@data),
                   Original = xdat,
                   Resampled = ydat)
  expr
}, 
.id = "cell") %>% 
  mutate(cell = str_c(cell, "-1"),
         cell = factor(cell, levels = names(cell_names)))

norm <- ggplot(expr, aes(Original, Resampled)) +
  geom_point(size = 0.25) + 
  geom_abline(aes(intercept = 0, slope = 1)) +
  coord_fixed() +
  facet_grid(~cell, 
             labeller = labeller(cell = cell_names)) +
  labs(x = "Original Library\nlog(Norm. UMI + 1)",
       y = "Resampled Library\nlog(Norm. UMI + 1)")

unnorm_expr <- map2_dfr(og_cells, rs_cells, function(x, y){
  xdat <- log1p(sobj@raw.data[, x, drop = T])
  ydat <- log1p(sobj@raw.data[, y, drop = T])
  expr <- data_frame(gene = rownames(sobj@data),
                   Original = xdat,
                   Resampled = ydat)
  expr
}, 
.id = "cell") %>% 
  mutate(cell = str_c(cell, "-1"),
         cell = factor(cell, levels = names(cell_names)))

unnorm <- ggplot(unnorm_expr, aes(Original, Resampled)) +
  geom_point(size = 0.25) + 
  geom_abline(aes(intercept = 0, slope = 1)) +
  coord_fixed() +
  facet_grid(~cell, 
             labeller = labeller(cell = cell_names)) +
  labs(x = "Original Library\nlog(UMI + 1)",
       y = "Resampled Library\nlog(UMI + 1)")
unnorm


save_plot("norm_expression.pdf", norm, 
          base_aspect_ratio = 1.8)

save_plot("unnorm_expression.pdf", unnorm, 
          base_aspect_ratio = 1.8)
```


Plot average megakaryocyte expression, the original megakaroytes, and the resampled megakaryocytes. Plot markers of megakaryocytes, defined above using the original data.

```{r, eval = T}
og_sobj <- readRDS("original_sobj.rds")
```

## Combined resampled and original cells

Calculate markers for MKs after merging resampling expression into original cells. 

```{r merge}

mat <- sc_objs[[reflib]]$umi_matrix

resampled_mat <- sc_objs[[resampled_lib]]$umi_matrix[,
                                                     cells[[resampled_lib]]]

not_resampled_cell_ids <- colnames(mat)[!colnames(mat) %in% cells[[resampled_lib]]]
no_resampled_mat <- mat[, not_resampled_cell_ids]

## add additional genes (original)
new_genes <- setdiff(rownames(resampled_mat), rownames(no_resampled_mat))
zero_mat <- matrix(0L, 
                   ncol = ncol(no_resampled_mat), 
                   nrow = length(new_genes),
                   dimnames = list(new_genes, colnames(no_resampled_mat)))
no_resampled_mat <- rbind(no_resampled_mat, zero_mat)

## add additional genes (resampled)
new_genes <- setdiff(rownames(no_resampled_mat), rownames(resampled_mat))
zero_mat <- matrix(0L, 
                   ncol = ncol(resampled_mat), 
                   nrow = length(new_genes),
                   dimnames = list(new_genes, colnames(resampled_mat)))
resampled_mat <- rbind(resampled_mat, 
      zero_mat)
## match original matrix roworder
resampled_mat <- resampled_mat[rownames(no_resampled_mat), ]

combined_mat <- cbind(no_resampled_mat, resampled_mat)

sobj <- CreateSeuratObject(combined_mat, min.genes = 200)

new_ids <- sobj@meta.data %>% 
  rownames_to_column("cell") %>% 
  mutate(resampled = ifelse(cell %in% cells[[resampled_lib]],
                             "resampled",
                             "not resampled"))

resampled_cell_ids <- new_ids[new_ids$resampled == "resampled", 
                               "cell"] %>% 
  str_replace("::resampled", "")
 
new_ids <- mutate(new_ids, 
                  resampled = ifelse(cell %in% resampled_cell_ids, 
                                      "original cell",
                                      resampled)) %>% 
  select(cell, resampled) %>% 
  as.data.frame(.) %>% 
  column_to_rownames("cell")

sobj <- AddMetaData(sobj, new_ids)

cell_ids <- select(cell_mdata, cell, cell_labels) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("cell")

sobj <- AddMetaData(sobj, cell_ids)
sobj <- NormalizeData(sobj)
sobj <- ScaleData(sobj)
sobj <- SetAllIdent(sobj,
                    "cell_labels") 
new_markers <- FindMarkers(sobj, 
                           ident.1 = "Megakaryocytes")

new_mk_markers <- tibble::rownames_to_column(new_markers, "gene") %>% 
  filter(p_val_adj < 0.01) %>% 
  tbl_df()

mk_markers <- read_tsv("original_pbmc_markers.txt") %>% 
  filter(cluster == "Megakaryocytes", p_val_adj < 0.01)

shared_mk_markers <- inner_join(new_mk_markers,
                                mk_markers, by = "gene",
                                suffix = c("_new", "_old"))

ggplot(shared_mk_markers,
       aes(p_val_old, p_val_new)) +
  geom_point() +
  scale_x_log10() +
  scale_y_log10() +
  geom_abline()

ggplot(shared_mk_markers,
       aes(avg_logFC_old, avg_logFC_new)) +
  geom_point() + 
  geom_abline()

plot_expr_raw <- data_frame(og = og_sobj@raw.data[new_mk_markers$gene,
                                              cells$mkcell_pulldown[1]],
                        resampled = sobj@raw.data[new_mk_markers$gene,
                                                  cells$mkcell_pulldown[1]])

ggplot(plot_expr_raw, 
       aes(og, resampled)) + 
  geom_point(size = 0.5) +
  geom_abline()

plot_expr_norm <- data_frame(og = og_sobj@data[new_mk_markers$gene,
                                              cells$mkcell_pulldown[1]],
                        resampled = sobj@data[new_mk_markers$gene,cells$mkcell_pulldown[1]])

ggplot(plot_expr_norm, 
       aes(og, resampled)) + 
  geom_point(size = 0.5) +
  geom_abline()

```


## information about genes recovered

```{r}
mk_markers <- filter(all_markers, 
                     cluster == "Megakaryocytes", 
                     p_val_adj < 0.01)

genes_to_plot <- rownames(sc_objs[[resampled_lib]]$umi_matrix)
cols <- colnames(sc_objs[[resampled_lib]]$norm_umi)
cell_ids <- cells[[resampled_lib]]

## append genes to reference library if necessary
ref_mat <- standardize_rows(sc_objs[[resampled_lib]]$umi_matrix[,
                                                                 cols],
                            sc_objs[[reflib]]$umi_matrix[, cols])

ma_dat <- map(cell_ids,
    ~calc_ma(sc_objs[[resampled_lib]]$umi_matrix[genes_to_plot, cols], 
             ref_mat[genes_to_plot, cols],
             cell = .x))

names(ma_dat) <- names(cell_names)
ma_dat <- bind_rows(ma_dat, .id = "cell") %>% 
  mutate(cell = factor(cell,
                       levels = names(cell_names)))

df <- mutate(ma_dat, 
               marker = ifelse(gene %in%  mk_markers$gene,
                                  "Megakaryocyte marker gene",
                                  "Other gene"))
n_up <- filter(df, log2_diff > 0) %>% 
    group_by(cell) %>% 
    summarize(n = n(), n = paste0("up = ", n))
  
n_down <- filter(df, log2_diff < 0) %>% 
    group_by(cell) %>% 
    summarize(n = n(), n = paste0("down = ", n))
  

marker_plt <- ggplot(df,
         aes(mean_expression_log2,
             log2_diff)) +
    geom_hline(aes(yintercept = 0), 
               linetype = "dashed", 
               colour = "grey") + 
    geom_point(aes(color = marker), size = 0.25) +
    geom_text(data = n_up, aes(x = max(ma_dat$mean_expression_log2) * 0.7, 
                               y = max(ma_dat$log2_diff) * 0.9, 
                               label = n)) +
    geom_text(data = n_down, aes(x = max(ma_dat$mean_expression_log2) * 0.7, 
                                 y = min(ma_dat$log2_diff) * 1.2, 
                                 label = n)) + 
    facet_wrap(~cell, nrow = 1, labeller = labeller(cell = cell_names)) +
    labs(x = expression(paste("Abundance (log"[2], ")")),
         y = expression(paste(frac("Resampled","Original"), 
                              " (log"[2], ")"))) +
    scale_color_manual(values = color_palette, name = "") +
    guides(colour = guide_legend(override.aes = list(size = 4))) + 
    theme(legend.position = "top") +
  theme(plot.margin = unit(rep(0, 4), "lines"))

marker_plt
save_plot("per_cell_MA_mk_markers.pdf", marker_plt, 
          base_height = 4, base_aspect_ratio = 2.5)

# umi_seqs defined above
umi_seqs_split <- umi_seqs[[resampled_lib]] %>% 
  filter(!is.na(count.x)) %>% 
  separate(umi_molecule, 
           into = c("umi", 
                    "gene"), 
           sep = "::")

mks <- rownames(sobj@meta.data)[sobj@meta.data$cell_labels == "Megakaryocytes"]

umi_seqs_split <- mutate(umi_seqs_split, 
                         mk = ifelse(barcode_10x %in% mks, 
                                     T,
                                     F)) %>% 
  filter(mk)

umi_summary <- group_by(umi_seqs_split,
                        barcode_10x, 
                        gene) %>% 
  summarize(n_umis = n(),
         n_reads = sum(count.x),
         read_saturation = 1 - (n_umis / n_reads),
         umi_saturation = n_umis / 4^10)

umi_summary <- mutate(umi_summary, 
                      cell = str_c(barcode_10x, "-1"))

plt_dat <- left_join(ma_dat, umi_summary, 
                     by = c("cell", "gene")) %>% 
  
 # filter(!is.na(barcode_10x)) %>% 
  mutate(cell = factor(cell_names[cell],
                       levels = cell_names))

seq_sat_plt <- ggplot(plt_dat,
                      aes(mean_expression_log2,
                          log2_diff)) +
  geom_hline(aes(yintercept = 0), 
             linetype = "dashed", 
             colour = "grey") + 
  geom_point(aes(color = read_saturation), 
               size = 0.25) +
  scale_colour_gradientn(colours = rev(brewer.pal(9, "Reds")),
                         name = "Sequencing\nSaturation",
                         na.value = "lightgrey") +
  facet_wrap(~cell, nrow = 1) +
  labs(x = expression(paste("Abundance (log"[2], ")")),
       y = expression(paste(frac("Resampled","Original"), 
                            " (log"[2], ")"))) +
  theme(legend.position = "top",
        legend.key.width = unit(7.5, 
                                "mm"),
        plot.margin = unit(rep(0, 4), "lines"))
#  legend.text = element_text(size = 10),
# legend.title = element_text(size = 10))

save_plot("per_cell_MA_seq_saturation.pdf", seq_sat_plt, 
          base_height = 4, base_aspect_ratio = 2.5)

plt <- plot_grid(seq_sat_plt, marker_plt, nrow = 2, 
                 align = "hv")
save_plot("per_cell_MA_combined_plots.pdf", plt, nrow = 2, ncol = 1,
          base_height = 4, base_aspect_ratio = 2.5)


# set seq saturation to median value for previously undetected genes
plt_dat_hex <- mutate(plt_dat, 
                      read_saturation = ifelse(is.na(read_saturation),
                                               median(read_saturation,
                                                      na.rm = T),
                                               read_saturation))
hex_plt <- ggplot(plt_dat_hex,
                      aes(mean_expression_log2,
                          log2_diff)) +
  geom_hline(aes(yintercept = 0), 
             linetype = "dashed", 
             colour = "grey") + 
  stat_summary_hex(aes(z = read_saturation),
           bins = 25,
           size = 0.25,
           fun = function(x) mean(x, na.rm = T)) +
  scale_fill_gradientn(colours = rev(brewer.pal(9, "Reds")),
                         name = "Sequencing\nSaturation",
                         na.value = "lightgrey") +
  facet_wrap(~cell, nrow = 1) +
  labs(x = expression(paste("Abundance (log"[2], ")")),
       y = expression(paste(frac("Resampled","Original"), 
                            " (log"[2], ")"))) +
  theme(legend.position = "top",
        legend.key.width = unit(7.5, 
                                "mm"),
        plot.margin = unit(rep(0, 4), "lines"))

save_plot("per_cell_MA_seq_saturation_hexplot_with_medians.pdf", hex_plt, 
          base_height = 4, base_aspect_ratio = 2.5)

# dont plot undetected
hex_plt <- ggplot(plt_dat,
                      aes(mean_expression_log2,
                          log2_diff)) +
  geom_hline(aes(yintercept = 0), 
             linetype = "dashed", 
             colour = "grey") + 
  stat_summary_hex(aes(z = read_saturation),
           bins = 30,
           size = 0.25,
           fun = function(x) mean(x, na.rm = T)) +
  scale_fill_gradientn(colours = rev(brewer.pal(9, "Reds")),
                         name = "Sequencing\nSaturation",
                         na.value = "lightgrey") +
  facet_wrap(~cell, nrow = 1) +
  labs(x = expression(paste("Abundance (log"[2], ")")),
       y = expression(paste(frac("Resampled","Original"), 
                            " (log"[2], ")"))) +
  theme(legend.position = "top",
        legend.key.width = unit(7.5, 
                                "mm"),
        plot.margin = unit(rep(0, 4), "lines"))

save_plot("per_cell_MA_seq_saturation_hexplot_without_undetected.pdf", hex_plt, 
          base_height = 4, base_aspect_ratio = 2.5)

# plot with undetected added with mean of megakaryocte read_saturation

all_megs_seq_sat <- umi_summary %>% 
  group_by(gene) %>% 
  summarize(mean_read_saturation = mean(read_saturation, 
                                   na.rm = T))

plt_dat_hex <- left_join(plt_dat, all_megs_seq_sat, by = "gene") %>% 
  mutate(read_saturation = ifelse(is.na(read_saturation),
                                               mean_read_saturation,
                                               read_saturation))
hex_plt <- ggplot(plt_dat_hex,
                      aes(mean_expression_log2,
                          log2_diff)) +
  geom_hline(aes(yintercept = 0), 
             linetype = "dashed", 
             colour = "grey") + 
  stat_summary_hex(aes(z = read_saturation),
           bins = 25,
           size = 0.25,
           fun = function(x) mean(x, na.rm = T)) +
  scale_fill_gradientn(colours = rev(brewer.pal(9, "Reds")),
                         name = "Sequencing\nSaturation",
                         na.value = "lightgrey") +
  facet_wrap(~cell, nrow = 1) +
  labs(x = expression(paste("Abundance (log"[2], ")")),
       y = expression(paste(frac("Resampled","Original"), 
                            " (log"[2], ")"))) +
  theme(legend.position = "top",
        legend.key.width = unit(7.5, 
                                "mm"),
        plot.margin = unit(rep(0, 4), "lines"))

save_plot("per_cell_MA_seq_saturation_hexplot_with_cluster_medians.pdf", hex_plt, 
          base_height = 4, base_aspect_ratio = 2.5)

```


## How much overlap with known MK markers

The 4 resampled cells have higher numbers of genes detected, but adding just these 4 cells is not sufficent to increase the number of marker genes discovered for the MK cluster. This is likely due to there only being 4 cells out of 70 that are resampled. More cells would be necessary for this task.

To determine if a subset of the new genes detected in the 4 cells are actually markers of MKs I instead compared these genes to genes identified in another PBMC experiment. 10x genomics has a scRNA-Seq dataset with 68k cells, which has > 200 MK cells (compared to our 70). I computed markers for these cells and will use this dataset as a reference for MK cell markers, in addition to the markers originally identified in the kirkpatrick expt. 

#### Shared marker genes

```{r, eval = F}
sc_objs <- readRDS("processed_data.rds")
mk_10x_markers <- read_tsv(file.path(results_dir, 
                                      "2018-06-01",
                                      "mk_markers_pbmc68k.txt")) %>% 
  filter(p_val_adj < 0.01)

new_genes <- sc_objs$mkcell_pulldown$new_genes[cells$mkcell_pulldown]


names(new_genes) <-  cell_names[names(new_genes)]

pdf("newgenes_shared_in_pulldown_cells.pdf")
  upset(fromList(new_genes), order.by = "freq")
dev.off()


all_detected_genes <- sc_objs$mkcell_pulldown$genes_detected[cells$mkcell_pulldown] %>% 
  map(., names)

names(all_detected_genes) <-  cell_names[names(all_detected_genes)]

pdf("genes_shared_in_pulldown_cells.pdf")
  upset(fromList(all_detected_genes), order.by = "freq")
dev.off()

all_detected_genes <- sc_objs$kirkpatrick$genes_detected[cells$mkcell_pulldown] %>% 
  map(., names)

names(all_detected_genes) <-  cell_names[names(all_detected_genes)]

pdf("genes_shared_in_original_cells.pdf")
  upset(fromList(all_detected_genes), order.by = "freq")
dev.off()
```
                                                    
#### compare to marker genes


```{r compare_to_markers}
sc_objs <- readRDS("processed_data.rds")

new_genes <- sc_objs$mkcell_pulldown$new_genes[cells$mkcell_pulldown]
genes_per_cell <- map(sc_objs,
                  ~.x$genes_detected[cells$mkcell_pulldown])

all_genes <- map(genes_per_cell, ~map(.x, names))


og_mk_markers <- read_tsv(file.path("original_pbmc_markers.txt")) %>% 
  filter(cluster == "Megakaryocytes",
         p_val_adj < 0.01)

summary_df <- map(all_genes,
               ~map(.x,
                    ~data_frame(n_markers = sum(.x %in%
                                             og_mk_markers$gene),
                           total_genes = length(.x))))

summary_df <- map(summary_df, bind_rows, .id = "cell") %>% 
  bind_rows(., .id = "library") %>% 
  mutate(
    cell = str_c(cell, "-1"),
    cell = cell_names[cell],
    cell = factor(cell, levels = cell_names))

plt <- ggplot(summary_df,
       aes(cell, n_markers)) +
  geom_bar(aes(fill = library), 
           stat = "identity",
       #    width = 0.9) +
           position = position_dodge(width = 0.9)) +
  scale_fill_manual(values = color_palette,
                     name = "",
                    labels = lib_names) +
  ylim(c(0, max(summary_df$n_markers))) + 
  labs(y = "# of Megakaryocyte\nmarker genes") +
  guides(fill = guide_legend(nrow = 2)) + 
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1),
        axis.title.x = element_blank(),
        legend.pos = "none")
  
save_plot("nmarkers_kirkpatrick_pbmcs.pdf", 
          plt,
          base_aspect_ratio = 1)
plt
```


#### compare to all genes detected


```{r}
og_sobj <- readRDS("original_sobj.rds")
og_sobj <- SetAllIdent(og_sobj, "cell_labels")
avg_og_expr <- AverageExpression(og_sobj)
mk_expr <- avg_og_expr[, "Megakaryocytes", drop = F]
mk_expr <- mk_expr[mk_expr$Megakaryocytes > 0, , drop = F]

summary_df <- map(all_genes,
               ~map(.x,
                    ~data_frame(n_markers = sum(.x %in%
                                             rownames(mk_expr)),
                           total_genes = length(.x))))
summary_df <- map(summary_df, bind_rows, .id = "cell") %>% 
  bind_rows(., .id = "library")

plt <- ggplot(summary_df,
       aes(cell, n_markers)) +
  geom_bar(aes(fill = library), 
           stat = "identity",
           position = "dodge") +
  scale_x_discrete(labels = cell_names) + 
  scale_fill_brewer(palette = "Set1",
                    name = "",
                    labels = lib_names) +
  labs(y = "# of Megakaryocyte\ngenes detected") +
  guides(fill = guide_legend(nrow = 2)) + 
  theme(axis.text.x = element_text(angle = 45,
                                   hjust = 1),
        axis.title.x = element_blank(),
        legend.pos = "top")
  
save_plot("ngenesdetected_kirkpatrick_pbmcs.pdf", 
          plt,
          base_aspect_ratio = 0.75)
```
  
#### compare gene expression with scatterplots
 
 Next I'll plot the original expression per cell versus the average expression in MK cells, then the resampled expression.
 
 
```{r get_expr_scat_color}
og_sobj <- readRDS("original_sobj.rds")
og_sobj <- SetAllIdent(og_sobj, "cell_labels")
average_expression <- AverageExpression(og_sobj)  
average_mk_expression <- average_expression[,
                                            "Megakaryocytes", 
                                            drop = F]

average_mk_expression <- log1p(average_mk_expression)
average_mk_expression <- average_mk_expression %>% 
  tibble::rownames_to_column("gene")
og_expr <- og_sobj@data[, cells$mkcell_pulldown] %>% 
  as.matrix(.) %>% 
  as.data.frame(.) %>% 
  tibble::rownames_to_column("gene")

average_mk_expression <- left_join(average_mk_expression, og_expr)

average_mk_expression <-  average_mk_expression %>% 
  gather(cell, expr, -gene, -Megakaryocytes) %>% 
  tbl_df()

## get cor:

cell_cor <- average_mk_expression %>% 
  group_by(cell) %>% 
  do(broom::tidy(cor(.$Megakaryocytes, .$expr))) %>% 
  ungroup() %>% 
  mutate(cell_cor = str_c("R = ", 
                          formatC(x, big.mark = ",", digits = 3)))

og_plot <- ggplot(average_mk_expression, 
         aes(expr, Megakaryocytes)) +
    geom_point(size = 0.5) + 
    geom_abline(aes(slope = 1, intercept = 0)) +
    geom_text(data = cell_cor,
              aes(x = 2,
                  y = max(average_mk_expression$Megakaryocytes) * 0.9, 
                  label = cell_cor),
              size = 3) +
    labs(x = "Original Cell Expression",
         y = "Megakaryocyte Cluster\nAverage Expression") +
    facet_wrap(~cell, nrow = 1) +
    coord_fixed() + 
    theme(legend.position = "top",
          strip.text = element_text(size = 8))


sobj <- SetAllIdent(sobj, "cell_labels")
average_rs_expression <- AverageExpression(sobj)  
average_rsmk_expression <- average_rs_expression[,
                                            "Megakaryocytes", 
                                            drop = F]

average_rsmk_expression <- log1p(average_rsmk_expression)
average_rsmk_expression <- average_rsmk_expression %>% 
  tibble::rownames_to_column("gene")
rs_expr <- sobj@data[, cells$mkcell_pulldown] %>% 
  as.matrix(.) %>% 
  as.data.frame(.) %>% 
  tibble::rownames_to_column("gene")

average_rsmk_expression <- left_join(average_rsmk_expression, rs_expr)

average_rsmk_expression <-  average_rsmk_expression %>% 
  gather(cell, expr, -gene, -Megakaryocytes) %>% 
  tbl_df()

## get cor:
rs_cell_cor <- average_rsmk_expression %>% 
  group_by(cell) %>% 
  do(broom::tidy(cor(.$Megakaryocytes, .$expr))) %>% 
  ungroup() %>% 
  mutate(cell_cor = str_c("R = ", 
                          formatC(x, big.mark = ",", digits = 3)))

rs_plt <- ggplot(average_rsmk_expression, 
         aes(expr, Megakaryocytes)) +
    geom_point(size = 0.5) + 
    geom_abline(aes(slope = 1, intercept = 0)) +
    geom_text(data = rs_cell_cor,
              aes(x = 2,
                  y = max(average_mk_expression$Megakaryocytes) * 0.9, 
                  label = cell_cor),
              size = 3) +
    facet_wrap(~cell, nrow = 1) +
    coord_fixed() + 
    labs(x = "Resampled Cell Expression",
         y = "Megakaryocyte Cluster\nAverage Expression") +
    theme(legend.position = "top",
          strip.text = element_text(size = 8))

cell_cor <- bind_rows(list(original = cell_cor, 
                           resampled = rs_cell_cor), .id = "type")

cell_cor
```

```{r plt}

plt <- plot_grid(og_plot, rs_plt, nrow = 2)

save_plot("expression_versus_cluster_average.pdf", plt, 
          nrow = 2, ncol = 1,
          base_height = 4,
          base_aspect_ratio = 2)
```


## Find markers with downsampled MK cluster

Rui had a great idea to downsample the # of cells in the MK cluster and find markers with the original cells, and the subsampled cells.

There are 68 mks in this data set. 
```{r}
subsampled_markers_og <- read_tsv("downsampled_mk_cluster_markers_no_resampling.txt") 

subsampled_markers <- read_tsv("downsampled_mk_cluster_markers_resampling.txt")

og_summary <- subsampled_markers_og %>% 
  filter(p_val_adj < 0.01) %>% 
  count(n_mks)

rs_summary <- subsampled_markers %>% 
    filter(p_val_adj < 0.01) %>% 
  count(n_mks)

summary_dat <- bind_rows(list("Original Library" = og_summary,
                              "Resampled Library" = rs_summary),
          .id = "expt")
summary_dat <- mutate(summary_dat,
                      n_mks = as.integer(n_mks) + 4) ## to account for resampled cells

plt <- ggplot(summary_dat, 
       aes(n_mks, n)) +
  geom_point(aes(color = expt), 
           stat = "identity") +
  geom_line(aes(color = expt)) +
  scale_color_manual(values = color_palette, name = "") +
  ylim(c(0, max(summary_dat$n))) + 
  labs(x = "# of Megakaryocytes in Cluster",
       y = "# of Megakaryocyte\nmarkers discovered") +
  theme(legend.pos = "none") +
  guides(fill = guide_legend(nrow = 2))

save_plot("resampling_marker_discovery_sensitivity.pdf", 
          plt, 
          base_height = 4,
          base_aspect_ratio = 1)
```


```{r tsne_with_only_4_original_cells}
sobj <- readRDS("rs_v2_sobj.rds")
mks <- sobj@meta.data[sobj@meta.data$cell_labels == "Megakaryocytes", ]
rs_mks <- mks[mks$resampled == "original cell", ]
not_mks <- sobj@meta.data[sobj@meta.data$cell_labels != "Megakaryocytes", ]
only4mks <- c(rownames(rs_mks),
  rownames(not_mks))
sub_obj <- SubsetData(og_sobj, cells.use = only4mks)

sub_obj <- RunPCA(sub_obj, pc.genes = rownames(sub_obj@data), 
               pcs.compute = 20, 
               do.print = F, seed.use = 20180525)
sub_obj <- RunTSNE(sub_obj, dims.use = 1:15, seed.use = 20180525)
sub_obj <- FindClusters(sub_obj,
                     dims.use = 1:15, 
                     k.param = 15,
                     resolution = 1.2, 
                     print.output = F, 
                     force.recalc = T,
                     random.seed = 20180525)

plot_feature(sub_obj, gene = "GNG11", pt.alpha = 1)
TSNEPlot(sub_obj, group.by = "resampled")
TSNEPlot(sub_obj, group.by = "res.1.2")
```




## Gene detection sensitivity analysis

```{r}
sobj <- readRDS("original_sobj.rds")
mks <- rownames(sobj@meta.data[sobj@meta.data$cell_labels == "Megakaryocytes", 
                               ])
rs_cells <- rownames(sobj@meta.data[sobj@meta.data$resampled, ])
not_rs_mks <- mks[!mks %in% rs_cells]
  
set.seed(42)
random_mks <- map(seq(0, length(not_rs_mks), 1),
    ~sample(not_rs_mks, .x, replace = F))

random_mks_w_rs <- map(random_mks, 
    ~c(.x, rs_cells))

not_rs_mks_genes <- sc_objs$kirkpatrick$genes_detected[not_rs_mks]

rs_rs_mks_genes <- sc_objs$mkcell_pulldown$genes_detected[rs_cells]

original_genes <- sc_objs$kirkpatrick$genes_detected
resampled_genes <- c(not_rs_mks_genes, rs_rs_mks_genes)

plt_dat <- map_dfr(random_mks_w_rs, 
    function(x){
      map_dfr(lst(original_genes, 
                  resampled_genes),
    ~length(unique(unlist(map(.x[x], 
                              names)))))},
    .id = "n_cells")

plt_dat <- gather(plt_dat, library, n_genes, -n_cells) %>% 
  mutate(n_cells = as.integer(n_cells))

plt <- ggplot(plt_dat, 
       aes(n_cells, n_genes)) +
         geom_line(aes(color = library)) +
  scale_color_manual(values = color_palette) +
  labs(x = "# of Megakaryocytes in Cluster",
       y = "# of Genes Detected\nin Cluster") +
  theme(legend.position = "none")

save_plot("resampling_gene_discovery_sensitivity.pdf",
          plt)

## repeat sampling to get mean +/- sd 

random_mks <- map(seq(0, length(not_rs_mks), 1),
    ~sample(not_rs_mks, .x, replace = F))

ncells_to_sample <- seq(0, length(not_rs_mks), 1)

run_one_round <- function(not_rs_cells,
                          rs_cells,
                          ncells, 
                          niter,
                          per_cell_gene_list
){
  nsamples <- map(1:niter, 
                  ~sample(not_rs_cells, 
                          ncells, 
                          replace = F))
  nsamples <- map(nsamples, ~c(.x, rs_cells))
  res <- map_dbl(nsamples, 
                 ~length(unique(unlist(map(per_cell_gene_list[.x], 
                                           names)))))
  data_frame(n_genes = res)
}

plt_dat <- map_dfr(ncells_to_sample, 
    function(ncells){
      res <- map_dfr(lst(original_genes, 
                  resampled_genes),
                  ~run_one_round(not_rs_cells = not_rs_mks,
                                rs_cells = rs_cells,
                                ncells = ncells,
                                niter = 100,
                                per_cell_gene_list = .x),
              .id = "library")
      res }, 
    .id = "n_cells")

plt_dat <-  mutate(plt_dat, n_cells = as.integer(n_cells)) %>% 
  group_by(n_cells, library) %>% 
  summarize(mean_genes = mean(n_genes),
            sd_genes = sd(n_genes))

plt <- ggplot(plt_dat, 
       aes(n_cells, mean_genes)) +
         geom_line(aes(color = library)) +
#  geom_linerange(aes(ymin = mean_genes - sd_genes,
#                     ymax = mean_genes + sd_genes)) + 
  scale_color_manual(values = color_palette) +
  labs(x = "# of Megakaryocytes in Cluster",
       y = "# of Genes Detected\nin Cluster") +
  theme(legend.position = "none")

save_plot("resampling_gene_discovery_sensitivity_with_replication.pdf",
          plt, base_height = 4, base_aspect_ratio = 1)


original_mk_genes <- sc_objs$kirkpatrick$genes_detected[mks]
resampled_mk_genes <- sc_objs$mkcell_pulldown$genes_detected[rs_cells]
og_genes <- sc_objs$kirkpatrick$genes_detected[mks]
rs_genes <- c(not_rs_mks_genes, resampled_mk_genes)

n_og_genes <- length(unique(unlist(map(og_genes, names))))
n_rs_genes <- length(unique(unlist(map(rs_genes, names))))

## original total genes
n_og_genes
## original total genes
n_rs_genes
```


## Correlation analysis

```{r clustify}
library(clustifyR)
library(ComplexHeatmap)
sc_objs <- readRDS("processed_data.rds")

sc_metadat <- map(sc_objs, ~.x$meta_dat) %>% 
  bind_rows(.id = "library") %>% 
  mutate(library = factor(library, levels = libs)) %>% 
  arrange(resampled)

resampled_metadat <- filter(sc_metadat,
                             library != reflib) %>% 
  mutate(library = factor(library, 
                          levels = resampled_libs))

resampled_cell_expr <- sc_objs$mkcell_pulldown$norm_umi[,
                                                        cells$mkcell_pulldown]

og_mat <- sc_objs$kirkpatrick$norm_umi
rs_mat <- sc_objs$mkcell_pulldown$norm_umi
shared_genes <- intersect(rownames(rs_mat), rownames(og_mat))

cor_mat <- cor(as.matrix(rs_mat[shared_genes, ]), 
    as.matrix(og_mat[shared_genes, ]))

Heatmap(cor_mat[, colnames(cor_mat) %in% cells$mkcell_pulldown], 
        show_column_names = F, 
        show_column_dend = F,
        show_row_names = F, 
        show_row_dend = F)

Heatmap(cor_mat[, 1:10], 
        show_column_names = F, 
        show_column_dend = F,
        show_row_names = F, 
        show_row_dend = F)
```

## KNN analysis

```{r, knns}
sobj <- readRDS("rs_sobj.rds")
## use combined data from above
data.use <- GetCellEmbeddings(object = sobj,
                              reduction.type = "pca",
                              dims.use = 1:20)

## find nearest neighboors using exact search
knn <- RANN::nn2(data.use, k = 5,
                 searchtype = 'standard',
                 eps = 0)

resampled_idxs <- knn$nn.idx[str_detect(rownames(data.use),
                                         "::resampled"), ]

nn_ids <- as_data_frame(t(apply(resampled_idxs, 1,
                      function(x)rownames(data.use)[x])))

colnames(nn_ids) <- c("query_cell", 
                      paste0("nearest neighbor ", 
                             1:(ncol(nn_ids) - 1)))


nn_ids <- mutate(nn_ids, 
                 cell_name = str_c(str_replace(query_cell, 
                                               "::resampled",
                                               ""), 
                                   "-1"),
                 cell_name = cell_names[cell_name]) %>% 
  dplyr::select(cell_name, everything())

nn_ids

clipr::write_clip(nn_ids)

# check cell labels from OG experiment
nns <- unique(unlist(nn_ids[, 3:ncol(nn_ids)]))
nns <- nns[!str_detect(nns, "::resampled")] 
sobj <- readRDS("original_sobj.rds")

table(sobj@meta.data[nns, "cell_labels"])
```

