---
title: "Pbmc megakaryocte cell pulldown"
author: "Kent Riemondy RBI"
date: '`R Sys.Date()`'
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

## Megakaryocyte targeting via LNA pulldown

Megakaryocytes represent a small proportion of the cells in a typical PBMC sample and therefore were selected as an additional test of the LNA pulldown techinique. Four megakaryocyte cells were reamplified after LNA pulldown. 

Following reamplification the 4 cell libraries were pooled and resequenced together. The raw fastqs were then processed using a Snakemake [pipeline](../../pipeline/Snakefile), to produce two processed data files:

1. A matrix with UMIs per cell (column) per gene (rows) (dge_matrix.txt)
1. A flatfile with per UMI information (umigroups.txt.gz)

This RMarkdown document will produce the following figures:

1. ...
1. ...
1. ...

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

```{r, message=F, warning=F, echo=F}
source("../../R/globals.R")
```

## Organize single cell data

First designate the libraries and the cells that were resampled. 

```{r cell_ids}
cells <- list(
  mkcell_pulldown = c("TGCGCAGCAGGTCGTC",
                      "ACTTGTTAGGACCACA",
                      "CCATTCGTCCCTGACT",
                      "TGTCCCAGTAAACACA"))

libs <- c(
  "kirkpatrick",
  "mkcell_pulldown")


lib_data_dir <- file.path(data_dir, 
                          "lna_cell",
                          "pbmc_expt")

bc_metadat <- read_tsv(file.path(lib_data_dir, 
                         "kirkpatrick", 
                         "fastq",
                         "original", 
                         "barcodes_from_10x_run.txt"),
                         col_names = c("cell_id", "barcode_10x"))

## original library to compare against
reflib <- "kirkpatrick"
## all resampled libs to plot
resampled_libs <- "mkcell_pulldown"
## reference resampled lib for resampled vs control plots
resampled_lib <- "mkcell_pulldown"

## pretty name for libraries
lib_names = c(
  kirkpatrick = "Original Library",
  mkcell_pulldown = "Resampled Library"
)

## pretty names for cells
cell_names = c(
  "TGCGCAGCAGGTCGTC" = "MK-cell-1",
  "ACTTGTTAGGACCACA" = "MK-cell-2",
  "CCATTCGTCCCTGACT" = "MK-cell-3",
  "TGTCCCAGTAAACACA" = "MK-cell-4")

```

Load and organize a table for each library of read counts per cell per gene, and a table of umi counts per cell per gene. 

```{r read_counts}

## read in umigroups flat file with read counts per umi per gene per cell
## expand out to a read count matrix
umipaths <- file.path(lib_data_dir, 
                      libs, 
                      "umis",
                      "umigroups.txt.gz")
read_dat <- map(umipaths, 
                ~umis_to_genes(.))
names(read_dat) <- libs

## read in umi_tools count table with umi counts per gene per cell
umi_dat <- map(libs, 
                ~read_tsv(file.path(lib_data_dir, 
                          .x,
                          "dgematrix",
                          "dge_matrix.txt")) %>% 
                 select(-Cell_unmatched))
names(umi_dat) <- libs

cell_obj_mdata <- map(cells, 
                      ~mutate(bc_metadat, 
                              resampled = ifelse(barcode_10x %in% .x,
                                                  TRUE,
                                                  FALSE)))

```

Next organize these tables into simple classes called `resampled-sets` to keep track of each experiment's relavant raw, processed, and meta data. 

```{r s3ify}

sc_objs <- list(umi_dat, read_dat, cell_obj_mdata)
sc_objs <- pmap(sc_objs, create_sc_obj)

rm(umi_dat)
rm(read_dat)
```

Next perform basic processing. 
  1) generate separate objects to store sparse matrices of umi and read counts. 
  2) normalize read and umi count data by total library size (sum of all read or umi counts for all cells in the experiment) and report as Reads per million or UMIs per million. 
  3) Compute per cell metrics (read and umi counts, sequencing saturation)

```{r store_matrices}
sc_objs <- map(sc_objs, generate_matrices)
sc_objs <- map(sc_objs, norm_libsize)
sc_objs <- map(sc_objs, compute_summaries)
```

Compute enrichment of reads/umis over the original library. 

```{r calc_enrichment}

sc_objs <- map(sc_objs,
    function(sub_dat){
      og_counts <- select(sc_objs[[reflib]]$meta_dat,
                          og_total_reads = total_reads,
                          og_total_umis = total_umis,
                          og_norm_total_umis = norm_total_umis,
                          og_norm_total_reads = norm_total_reads,
                          og_cDNA_duplication = cDNA_duplication,
                          barcode_10x)
      sub_dat$meta_dat <- left_join(sub_dat$meta_dat,
                         og_counts, 
                         by = "barcode_10x")
      
      sub_dat$meta_dat <- mutate(sub_dat$meta_dat,
                                 read_proportion = log2( total_reads / og_total_reads),
                                 umi_proportion = log2( total_umis / og_total_umis),
                                 norm_read_proportion = log2( norm_total_reads /
                                                                og_norm_total_reads),
                                 norm_umi_proportion = log2( norm_total_umis /
                                                               og_norm_total_umis))
      sub_dat
    })

```

Next write out matrices to disk, for eventual GEO deposit

```{r}
dir.create("count_matrices", showWarnings = F)
iwalk(sc_objs, 
     function(x, y) {
       out_fn <- file.path("count_matrices", paste0(y, "_umis.tsv"))
       write_tsv(x$umis, out_fn)
       R.utils::gzip(out_fn, overwrite = TRUE)
     })
```



## Genes detected

```{r genes}

## compute per gene or per gene/umi combo enrichment
detected_molecules <- function(sc_obj, molecule = "gene"){
  umis <- sc_obj$umi_matrix
  if (molecule == "gene"){
    n_genes <- colSums(umis > 0)
    out_mdat <- data_frame(barcode_10x = colnames(umis),
      n_genes = n_genes)
    sc_obj <- add_metadata(sc_obj, out_mdat)
    }
}
sc_objs <- map(sc_objs, ~detected_molecules(.x))
```


### Parse out new versus previously identified genes

```{r genes_new_old}

calc_gene_sensitivity <- function(sc_obj, 
                                  type = "umi"){
  
  if (type == "umi"){
    count_matrix <- sc_obj$umi_matrix
  } else {
    count_matrix <- sc_obj$read_matrix
  }
  # generate list named with barcode of each detected gene and 
  # respective read/umi count
  genes_detected <- apply(count_matrix, 2, function(x) x[x > 0])
  sc_obj$genes_detected <- genes_detected
  sc_obj
}

sc_objs <- map(sc_objs, calc_gene_sensitivity)
  
```


```{r compare_original}
sc_objs <- map(sc_objs, 
           function(x){
             og_genes <- sc_objs[[reflib]]$genes_detected
             sub_genes <- x$genes_detected
             
             # subset list of cell barcodes to the same as the og experiment
             # and also reorders the barcodes to match
             sub_genes <- sub_genes[names(og_genes)]
             
             if(length(sub_genes) != length(og_genes)){
               stop("barcode lengths not the same")
             }
             shared_genes <- map2(sub_genes, 
                                  og_genes,
                                  ~intersect(names(.x),
                                             names(.y)))
             new_genes <- map2(sub_genes,
                               og_genes,
                               ~setdiff(names(.x),
                                        names(.y)))
             
             not_recovered_genes <- map2(og_genes,
                                         sub_genes,
                                         ~setdiff(names(.x),
                                                  names(.y)))
             x$shared_genes <- shared_genes
             x$new_genes <- new_genes
             x$not_recovered_genes <- not_recovered_genes
             return(x)
             })

## add gene recovery info to meta data table
sc_objs <- map(sc_objs, 
           function(x){
             shared_genes <- map2_dfr(x$shared_genes, 
                                names(x$shared_genes),
                                function(x, y){
                                  data_frame(barcode_10x = y,
                                             shared_genes = length(x))
                                 })
             
             not_recovered_genes <- map2_dfr(x$not_recovered_genes, 
                                names(x$not_recovered_genes),
                                function(x, y){
                                  data_frame(barcode_10x = y,
                                            not_recovered_genes = length(x))
                                 })
             
             new_genes <- map2_dfr(x$new_genes, 
                                names(x$new_genes),
                                function(x, y){
                                  data_frame(barcode_10x = y,
                                            new_genes = length(x))
                                 })
             gene_mdata <- left_join(shared_genes,
                                     not_recovered_genes,
                                     by = "barcode_10x") %>% 
               left_join(., new_genes, by = "barcode_10x")
             
             x <- add_metadata(x, gene_mdata)
             x
           })

resampled_metadat <- map(sc_objs, ~.x$meta_dat) %>% 
  bind_rows(.id = "library") %>% 
   mutate(library = factor(library, 
                          levels = libs))
```


```{r save_obs}
saveRDS(sc_objs, "processed_data.rds")
```

## Correlation analysis

```{r clustify}
library(clustifyR)

resampled_cell_expr <- sc_objs$mkcell_pulldown$norm_umi[, cells$mkcell_pulldown]

a <- run_cor(sc_objs$kirkpatrick$norm_umi,
        data_frame(cell_name = colnames(sc_objs$kirkpatrick$norm_umi),
                   cluster = sample(1:10, 
               ncol(sc_objs$kirkpatrick$norm_umi),
               replace = T)),
        sc_objs$mkcell_pulldown$norm_umi,
        rownames(sc_objs$mkcell_pulldown$norm_umi),
        cluster_col = "cell_name",
        per_cell = T,
        compute_method = corr_coef)

Heatmap(a, show_column_names = F, show_column_dend = F,
        show_row_names = F, show_row_dend = F)
```

## tSNE analysis

  ### original library tSNE

```{r tsne, results='hide'}
library(Seurat)

mat <- sc_objs[[reflib]]$umi_matrix
sobj <- CreateSeuratObject(mat, min.genes = 200)
sobj <- NormalizeData(sobj)

sobj <- ScaleData(sobj)
sobj <- FindVariableGenes(sobj, do.plot = T, y.cutoff = 0.25)
sobj <- RunPCA(sobj, pc.genes = rownames(sobj@data), 
               pcs.compute = 20, 
               do.print = F, seed.use = 20180525)
sobj <- RunTSNE(sobj, dims.use = 1:15, seed.use = 20180525)
sobj <- FindClusters(sobj,
                     dims.use = 1:15, 
                     k.param = 15,
                     resolution = 1.2, 
                     print.output = F, 
                     random.seed = 20180525)

```

```{r plt_tsne}

cell_mdata <- sobj@meta.data %>% 
  tibble::rownames_to_column("cell") %>% 
  mutate(resampled = ifelse(cell %in% cells$mkcell_pulldown, 
                             T,
                             F)) %>% 
  select(cell, resampled) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("cell") 

sobj <- AddMetaData(sobj, cell_mdata, col.name = "resampled")

sobj <- SetAllIdent(sobj, "resampled")


```

```{r pbmc_markers}
immune_markers <- c(t_cells = "CD3E", 
                    cd8_t ="CD8A", 
                    cytotoxic_t = "NKG7", 
                    dendritic = "FCER1A", 
                    megakaryocyte = "PF4", 
                    b_cell = "CD79A", 
                    cd4_4 = "IL7R", 
                    monocyte = "CD14",
                    RBC = "HBB",
                    NK = "NCAM1",
                    pDC = "LILRA4",
                    monocyte = "FCGR3A")

plts <- map(immune_markers, ~plot_feature(sobj, gene = .x, 
                                          pt.alpha = 1))
plt <- plot_grid(plotlist = plts,  nrow = 4)
plt
sobj <- SetAllIdent(sobj, "res.1.2")

new_ids <- c(
  "0" = "CD4+ T-Cells",
  "1" = "CD8+ T-Cells",
  "2" = "CD4+ T-Cells",
  "3" = "CD8+ T-Cells",
  "4" = "CD14+ Monocytes",
  "5" = "CD8+ T-Cells",
  "6" = "CD14+ Monocytes",
  "7" = "FCGR3A+ Monocytes",
  "8" = "CD14+ Monocytes",
  "9" = "CD14+ Monocytes",
  "10" = "Dendritic",
  "11" = "CD4+ T-Cells",
  "12" = "B-Cells",
  "13" = "NK Cells",
  "14" = "Megakaryocytes",
  "15" = "CD8+ T-Cells",
  "16" = "Plasmacytoid dendritic"
)
old_ids <- sobj@meta.data %>% 
  rownames_to_column("cell") %>% 
  pull("res.1.2")

new_labels <- new_ids[old_ids] %>% unname()
new_df <- data.frame(row.names = rownames(sobj@meta.data),
                     cell_labels = new_labels)
sobj <- AddMetaData(sobj, new_df)

```

```{r cell_percentages}
cell_percents <- group_by(sobj@meta.data, cell_labels) %>% 
  summarize(n_cells = n()) %>% 
  mutate(total_cells = sum(n_cells), 
         percentage = 100 * (n_cells / total_cells)) %>% 
  select(-total_cells)

cell_percents
```


```{r}
saveRDS(sobj, "original_sobj.rds")
```

```{r find_mk_markers}

if (!file.exists("original_pbmc_markers.txt")){
  sobj <- SetAllIdent(sobj, "cell_labels")
  all_markers <- FindAllMarkers(sobj)
  write_tsv(all_markers, "original_pbmc_markers.txt")
}

all_markers <- read_tsv("original_pbmc_markers.txt")
cell_mdata <- sobj@meta.data %>% 
  tibble::rownames_to_column("cell") 

```



### original library tSNE supplemented with resampled barcodes


```{r resampled_tsn, results='hide'}

mat <- sc_objs[[reflib]]$umi_matrix

resampled_ids <- sc_objs[[resampled_libs]]$meta_dat %>% 
  filter(resampled) %>% 
  pull(barcode_10x)

resampled_mat <- sc_objs[[resampled_libs]]$umi_matrix[,
                                                      cells$mkcell_pulldown]
colnames(resampled_mat) <- str_c(colnames(resampled_mat), 
                                  "::", "resampled")

mat <- as.data.frame(as.matrix(mat)) %>% rownames_to_column("gene")
resampled_mat <- as.data.frame(as.matrix(resampled_mat)) %>% rownames_to_column("gene")

combined_mats <- left_join(mat, resampled_mat, by = c("gene")) 
combined_mats <- as.data.frame(combined_mats) %>% 
  column_to_rownames("gene") %>% 
  as.matrix() %>% 
  as(., "sparseMatrix")   

combined_mats[is.na(combined_mats)] <- 0

sobj <- CreateSeuratObject(combined_mats, min.genes = 200)

new_ids <- sobj@meta.data %>% 
  rownames_to_column("cell") %>% 
  mutate(resampled = ifelse(str_detect(cell, "resampled"),
                             "resampled",
                             "not resampled"))

resampled_cell_ids <- new_ids[new_ids$resampled == "resampled", 
                              "cell"] %>% 
  str_replace("::resampled", "")
 
new_ids <- mutate(new_ids, 
                  resampled = ifelse(cell %in% resampled_cell_ids, 
                                      "original cell",
                                      resampled)) %>% 
  select(cell, resampled) %>% 
  as.data.frame(.) %>% 
  column_to_rownames("cell")

sobj <- AddMetaData(sobj, new_ids)
sobj <- NormalizeData(sobj)
sobj <- ScaleData(sobj)
sobj <- FindVariableGenes(sobj, do.plot = F, y.cutoff = 0.25)
sobj <- RunPCA(sobj, pc.genes = rownames(sobj@data), 
               pcs.compute = 20, 
               do.print = F, seed.use = 20180605)
sobj <- RunTSNE(sobj, dims.use = 1:15, seed.use = 20180605)
sobj <- FindClusters(sobj,
                     dims.use = 1:15, 
                     resolution = 1.2, 
                     print.output = F, 
                     random.seed = 20180605)
```

```{r}
saveRDS(sobj, "rs_sobj.rds")
```

## kNN analysis

Find the k-nearest neighbors in PCA space


```{r, knns}

## use combined data from above
data.use <- GetCellEmbeddings(object = sobj,
                              reduction.type = "pca",
                              dims.use = 1:20)

## find nearest neighboors using exact search
knn <- RANN::nn2(data.use, k = 5,
                 searchtype = 'standard',
                 eps = 0)

resampled_idxs <- knn$nn.idx[str_detect(rownames(data.use),
                                         "::resampled"), ]

nn_ids <- as_data_frame(t(apply(resampled_idxs, 1,
                      function(x)rownames(data.use)[x])))

colnames(nn_ids) <- c("query_cell", 
                      paste0("nearest neighbor ", 
                             1:(ncol(nn_ids) - 1)))

nn_ids
```







## markers from supplemented sample

Calculate markers for MKs after merging resampling expression into original cells. 

```{r merge}

mat <- sc_objs[[reflib]]$umi_matrix

resampled_mat <- sc_objs[[resampled_lib]]$umi_matrix[,
                                            cells[[resampled_lib]]]

not_resampled_cell_ids <- colnames(mat)[!colnames(mat) %in% cells[[resampled_lib]]]
no_resampled_mat <- mat[, not_resampled_cell_ids]

## add additional genes (original)
new_genes <- setdiff(rownames(resampled_mat), rownames(no_resampled_mat))
zero_mat <- matrix(0L, 
                   ncol = ncol(no_resampled_mat), 
                   nrow = length(new_genes),
                   dimnames = list(new_genes, colnames(no_resampled_mat)))
no_resampled_mat <- rbind(no_resampled_mat, zero_mat)

## add additional genes (resampled)
new_genes <- setdiff(rownames(no_resampled_mat), rownames(resampled_mat))
zero_mat <- matrix(0L, 
                   ncol = ncol(resampled_mat), 
                   nrow = length(new_genes),
                   dimnames = list(new_genes, colnames(resampled_mat)))
resampled_mat <- rbind(resampled_mat, 
      zero_mat)
## match original matrix roworder
resampled_mat <- resampled_mat[rownames(no_resampled_mat), ]

combined_mat <- cbind(no_resampled_mat, resampled_mat)

sobj <- CreateSeuratObject(combined_mat, min.genes = 200)

new_ids <- sobj@meta.data %>% 
  rownames_to_column("cell") %>% 
  mutate(resampled = ifelse(cell %in% cells[[resampled_lib]],
                             "resampled",
                             "not resampled"))

resampled_cell_ids <- new_ids[new_ids$resampled == "resampled", 
                               "cell"] %>% 
  str_replace("::resampled", "")
 
new_ids <- mutate(new_ids, 
                  resampled = ifelse(cell %in% resampled_cell_ids, 
                                      "original cell",
                                      resampled)) %>% 
  select(cell, resampled) %>% 
  as.data.frame(.) %>% 
  column_to_rownames("cell")

sobj <- AddMetaData(sobj, new_ids)

cell_ids <- select(cell_mdata, cell, cell_labels) %>% 
  as.data.frame() %>% 
  tibble::column_to_rownames("cell")

sobj <- AddMetaData(sobj, cell_ids)
sobj <- NormalizeData(sobj)
sobj <- ScaleData(sobj)
sobj <- SetAllIdent(sobj,
                    "cell_labels") 
new_markers <- FindMarkers(sobj, 
                           ident.1 = "Megakaryocytes")

new_mk_markers <- tibble::rownames_to_column(new_markers, "gene") %>% 
  filter(p_val_adj < 0.01) %>% 
  tbl_df()

mk_markers <- read_tsv("original_pbmc_markers.txt") %>% 
  filter(cluster == "Megakaryocytes", p_val_adj < 0.01)

shared_mk_markers <- inner_join(new_mk_markers,
                                mk_markers, by = "gene",
                                suffix = c("_new", "_old"))

saveRDS(sobj, "rs_v2_sobj.rds")
```

## Find markers with downsampled MK cluster

Rui had a great idea to downsample the # of cells in the MK cluster and find markers with the original cells, and the subsampled cells.

There are 68 mks in this data set. 
```{r downsample}
og_sobj <- readRDS("original_sobj.rds")
sobj <- readRDS("rs_v2_sobj.rds")

og_sobj <- SetAllIdent(og_sobj, "cell_labels")
sobj <- SetAllIdent(sobj, "cell_labels")

resampled_file <- file.exists("downsampled_mk_cluster_markers_resampling.txt")
not_resampled_file <- file.exists("downsampled_mk_cluster_markers_no_resampling.txt")

if(!(resampled_file & not_resampled_file)){
  mks <- sobj@meta.data[sobj@meta.data$cell_labels == "Megakaryocytes", ]
  rs_mks <- mks[mks$resampled == "original cell", ]
  not_re_mks <- mks[mks$resampled == "not resampled", ]
  not_mks <- sobj@meta.data[sobj@meta.data$cell_labels != "Megakaryocytes", ]
  
  n_mks_to_test <- seq(0, nrow(not_re_mks), by = 5)
  
  set.seed(42)
  not_re_mk_sampled <- map(n_mks_to_test, 
                           ~sample_n(not_re_mks, .x))
  
  sampled_mks <- map(not_re_mk_sampled, 
                     ~c(rownames(.x), 
                        rownames(rs_mks)))
  
  all_cells_minus_not_sampled_mks <- map(sampled_mks,
                                         ~c(.x, rownames(not_mks)))
  
  subsampled_markers <- map(all_cells_minus_not_sampled_mks, 
                            function(x){
                              tmp_dat <- SubsetData(sobj, cells.use = x)
                              markers <- FindMarkers(tmp_dat, 
                                                     "Megakaryocytes",
                                                     only.pos = T)
                              markers
                            })
  
  
  og_rs_mks <- og_sobj@meta.data[og_sobj@meta.data$resampled, ]
  
  sampled_mks <- map(not_re_mk_sampled, 
                     ~c(rownames(.x), 
                        rownames(og_rs_mks)))
  
  all_cells_minus_not_sampled_mks <- map(sampled_mks,
                                         ~c(.x, rownames(not_mks)))
  
  subsampled_markers_og <- map(all_cells_minus_not_sampled_mks,
                               function(x){
                                 tmp_dat <- SubsetData(og_sobj, cells.use = x)
                                 markers <- FindMarkers(tmp_dat, 
                                                        "Megakaryocytes",
                                                        only_pos = T)
                                 markers
                               })
  subsampled_markers <- map(subsampled_markers, 
                            ~tibble::rownames_to_column(.x, "gene")) 
  names(subsampled_markers) <- n_mks_to_test
  subsampled_markers <- bind_rows(subsampled_markers, .id = "n_mks")
  write_tsv(subsampled_markers,
            "downsampled_mk_cluster_markers_resampling.txt")
  
  subsampled_markers_og <- map(subsampled_markers_og, 
                               ~tibble::rownames_to_column(.x, "gene")) 
  
  names(subsampled_markers_og) <- n_mks_to_test
  subsampled_markers_og <- bind_rows(subsampled_markers_og, .id = "n_mks")
  write_tsv(subsampled_markers_og,
            "downsampled_mk_cluster_markers_no_resampling.txt")
}
```
